{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style> div#notebook {\n",
       " font-family: sans-serif;\n",
       " font-size: 13pt;\n",
       " line-height: 170%;\n",
       " color: #cdd2e9;\n",
       " -webkit-font-smoothing: antialiased !important;\n",
       " padding-top: 25px !important;\n",
       "}\n",
       "body,\n",
       "div.body {\n",
       " font-family: sans-serif;\n",
       " font-size: 13pt;\n",
       " color: #a2b0c7;\n",
       " background-color: #1a2028;\n",
       " background: #1a2028;\n",
       " -webkit-font-smoothing: antialiased !important;\n",
       "}\n",
       "body.notebook_app {\n",
       " padding: 0;\n",
       " background-color: #1a2028;\n",
       " background: #1a2028;\n",
       " padding-right: 0px !important;\n",
       " overflow-y: hidden;\n",
       "}\n",
       "a {\n",
       " font-family: sans-serif;\n",
       " color: #a2b0c7;\n",
       " -webkit-font-smoothing: antialiased !important;\n",
       "}\n",
       "a:hover,\n",
       "a:focus {\n",
       " color: #dbe1ea;\n",
       " -webkit-font-smoothing: antialiased !important;\n",
       "}\n",
       "div#maintoolbar {\n",
       " position: absolute;\n",
       " width: 90%;\n",
       " margin-left: -10%;\n",
       " padding-right: 8%;\n",
       " float: left;\n",
       " background: transparent !important;\n",
       "}\n",
       "#maintoolbar {\n",
       " margin-bottom: -3px;\n",
       " margin-top: 0px;\n",
       " border: 0px;\n",
       " min-height: 27px;\n",
       " padding-top: 2px;\n",
       " padding-bottom: 0px;\n",
       "}\n",
       "#maintoolbar .container {\n",
       " width: 75%;\n",
       " margin-right: auto;\n",
       " margin-left: auto;\n",
       "}\n",
       ".list_header,\n",
       "div#notebook_list_header.row.list_header {\n",
       " font-size: 14pt;\n",
       " color: #dbe1ea;\n",
       " background-color: transparent;\n",
       " height: 35px;\n",
       "}\n",
       "i.fa.fa-folder {\n",
       " display: inline-block;\n",
       " font: normal normal normal 14px \"FontAwesome\";\n",
       " font-family: \"FontAwesome\" !important;\n",
       " text-rendering: auto;\n",
       " -webkit-font-smoothing: antialiased;\n",
       " font-size: 18px;\n",
       " -moz-osx-font-smoothing: grayscale;\n",
       "}\n",
       "#running .panel-group .panel .panel-heading {\n",
       " font-size: 14pt;\n",
       " color: #a2b0c7;\n",
       " padding: 8px 8px;\n",
       " background: #252b35;\n",
       " background-color: #252b35;\n",
       "}\n",
       "#running .panel-group .panel .panel-heading a {\n",
       " font-size: 14pt;\n",
       " color: #a2b0c7;\n",
       "}\n",
       "#running .panel-group .panel .panel-heading a:focus,\n",
       "#running .panel-group .panel .panel-heading a:hover {\n",
       " font-size: 14pt;\n",
       " color: #a2b0c7;\n",
       "}\n",
       "#running .panel-group .panel .panel-body .list_container .list_item {\n",
       " background: #2d3846;\n",
       " background-color: #2d3846;\n",
       " padding: 2px;\n",
       " border-bottom: 2px solid rgba(75,95,118,.30);\n",
       "}\n",
       "#running .panel-group .panel .panel-body .list_container .list_item:hover {\n",
       " background: #2d3846;\n",
       " background-color: #2d3846;\n",
       "}\n",
       "#running .panel-group .panel .panel-body {\n",
       " padding: 2px;\n",
       "}\n",
       "button#refresh_running_list {\n",
       " border: none !important;\n",
       "}\n",
       "button#refresh_cluster_list {\n",
       " border: none !important;\n",
       "}\n",
       "div.running_list_info.toolbar_info {\n",
       " font-size: 15px;\n",
       " padding: 4px 0 4px 0;\n",
       " margin-top: 5px;\n",
       " margin-bottom: 8px;\n",
       " height: 24px;\n",
       " line-height: 24px;\n",
       " text-shadow: none;\n",
       "}\n",
       ".list_placeholder {\n",
       " font-weight: normal;\n",
       "}\n",
       "#tree-selector {\n",
       " padding: 0px;\n",
       " border-color: transparent;\n",
       "}\n",
       "#project_name > ul > li > a > i.fa.fa-home {\n",
       " color: #0b98c8;\n",
       " font-size: 17pt;\n",
       " display: inline-block;\n",
       " position: static;\n",
       " padding: 0px 0px;\n",
       " font-weight: normal;\n",
       " text-align: center;\n",
       " vertical-align: text-top;\n",
       "}\n",
       ".fa-folder:before {\n",
       " color: #4c8be2;\n",
       "}\n",
       ".fa-arrow-up:before {\n",
       " font-size: 14px;\n",
       "}\n",
       ".fa-arrow-down:before {\n",
       " font-size: 14px;\n",
       "}\n",
       "span#last-modified.btn.btn-xs.btn-default.sort-action:hover .fa,\n",
       "span#sort-name.btn.btn-xs.btn-default.sort-action:hover .fa {\n",
       " color: #009cd1;\n",
       "}\n",
       ".folder_icon:before {\n",
       " display: inline-block;\n",
       " font: normal normal normal 14px/1 FontAwesome;\n",
       " font-size: inherit;\n",
       " text-rendering: auto;\n",
       " -webkit-font-smoothing: antialiased;\n",
       " -moz-osx-font-smoothing: grayscale;\n",
       " content: \"\\f07b\";\n",
       " color: #4c8be2;\n",
       "}\n",
       ".notebook_icon:before {\n",
       " display: inline-block;\n",
       " font: normal normal normal 14px/1 FontAwesome;\n",
       " font-size: inherit;\n",
       " text-rendering: auto;\n",
       " -webkit-font-smoothing: antialiased;\n",
       " -moz-osx-font-smoothing: grayscale;\n",
       " content: \"\\f02d\";\n",
       " position: relative;\n",
       " color: #48a667 !important;\n",
       " top: 0px;\n",
       "}\n",
       ".file_icon:before {\n",
       " display: inline-block;\n",
       " font: normal normal normal 14px/1 FontAwesome;\n",
       " font-size: inherit;\n",
       " text-rendering: auto;\n",
       " -webkit-font-smoothing: antialiased;\n",
       " -moz-osx-font-smoothing: grayscale;\n",
       " content: \"\\f15b\";\n",
       " position: relative;\n",
       " top: 0px;\n",
       " color: #92a2bd !important;\n",
       "}\n",
       "#project_name a {\n",
       " display: inline-flex;\n",
       " padding-left: 7px;\n",
       " margin-left: -2px;\n",
       " text-align: -webkit-auto;\n",
       " vertical-align: baseline;\n",
       " font-size: 18px;\n",
       "}\n",
       "div#notebook_toolbar div.dynamic-instructions {\n",
       " font-family: sans-serif;\n",
       " font-size: 17px;\n",
       " color: #546386;\n",
       "}\n",
       "span#login_widget > .button,\n",
       "#logout {\n",
       " font-family: \"Proxima Nova\", sans-serif;\n",
       " color: #a2b0c7;\n",
       " background: transparent;\n",
       " background-color: transparent;\n",
       " border: 2px solid #252e3a;\n",
       " font-weight: normal;\n",
       " box-shadow: none;\n",
       " text-shadow: none;\n",
       " border-radius: 3px;\n",
       " margin-right: 10px;\n",
       " padding: 2px 7px;\n",
       "}\n",
       "span#login_widget > .button:hover,\n",
       "#logout:hover {\n",
       " color: #009cd1;\n",
       " background-color: transparent;\n",
       " background: transparent;\n",
       " border: 2px solid #009cd1;\n",
       " background-image: none;\n",
       " box-shadow: none !important;\n",
       " border-radius: 3px;\n",
       "}\n",
       "span#login_widget > .button:focus,\n",
       "#logout:focus,\n",
       "span#login_widget > .button.focus,\n",
       "#logout.focus,\n",
       "span#login_widget > .button:active,\n",
       "#logout:active,\n",
       "span#login_widget > .button.active,\n",
       "#logout.active,\n",
       ".open > .dropdown-togglespan#login_widget > .button,\n",
       ".open > .dropdown-toggle#logout {\n",
       " color: #fefefe;\n",
       " background-color: #a2b0c7;\n",
       " background: #a2b0c7;\n",
       " border-color: #a2b0c7;\n",
       " background-image: none;\n",
       " box-shadow: none !important;\n",
       " border-radius: 2px;\n",
       "}\n",
       "body > #header #header-container {\n",
       " padding-bottom: 0px;\n",
       " padding-top: 4px;\n",
       " box-sizing: border-box;\n",
       " -moz-box-sizing: border-box;\n",
       " -webkit-box-sizing: border-box;\n",
       "}\n",
       "body > #header {\n",
       " background: #1a2028;\n",
       " background-color: #1a2028;\n",
       " position: relative;\n",
       " z-index: 100;\n",
       "}\n",
       ".list_container {\n",
       " font-size: 13pt;\n",
       " color: #a2b0c7;\n",
       " border: none;\n",
       " text-shadow: none !important;\n",
       "}\n",
       ".list_container > div {\n",
       " border-bottom: 1px solid rgba(75,95,118,.30);\n",
       " font-size: 13pt;\n",
       "}\n",
       ".list_header > div,\n",
       ".list_item > div {\n",
       " padding-top: 6px;\n",
       " padding-bottom: 2px;\n",
       " padding-left: 0px;\n",
       "}\n",
       ".list_header > div .item_link,\n",
       ".list_item > div .item_link {\n",
       " margin-left: -1px;\n",
       " vertical-align: middle;\n",
       " line-height: 22px;\n",
       " font-size: 13pt;\n",
       "}\n",
       ".item_icon {\n",
       " color: #4c8be2;\n",
       " font-size: 13pt;\n",
       " vertical-align: middle;\n",
       "}\n",
       ".list_item input:not([type=\"checkbox\"]) {\n",
       " padding-right: 0px;\n",
       " height: 1.75em;\n",
       " width: 25%;\n",
       " margin: 0px 0 0;\n",
       " margin-top: 0px;\n",
       "}\n",
       ".list_header > div .item_link,\n",
       ".list_item > div .item_link {\n",
       " margin-left: -1px;\n",
       " vertical-align: middle;\n",
       " line-height: 1.5em;\n",
       " font-size: 12pt;\n",
       " display: inline-table;\n",
       " position: static;\n",
       "}\n",
       "#button-select-all {\n",
       " height: 34px;\n",
       " min-width: 55px;\n",
       " z-index: 0;\n",
       " border: none !important;\n",
       " padding-top: 0px;\n",
       " padding-bottom: 0px;\n",
       " margin-bottom: 0px;\n",
       " margin-top: 0px;\n",
       " left: -3px;\n",
       " border-radius: 0px !important;\n",
       "}\n",
       "#button-select-all:focus,\n",
       "#button-select-all:active:focus,\n",
       "#button-select-all.active:focus,\n",
       "#button-select-all.focus,\n",
       "#button-select-all:active.focus,\n",
       "#button-select-all.active.focus {\n",
       " background-color: #252e3a !important;\n",
       " background: #252e3a !important;\n",
       "}\n",
       "button#tree-selector-btn {\n",
       " height: 34px;\n",
       " font-size: 12.0pt;\n",
       " border: none;\n",
       " left: 0px;\n",
       " border-radius: 0px !important;\n",
       "}\n",
       "input#select-all.pull-left.tree-selector {\n",
       " margin-left: 7px;\n",
       " margin-right: 2px;\n",
       " margin-top: 2px;\n",
       " top: 4px;\n",
       "}\n",
       "input[type=\"radio\"],\n",
       "input[type=\"checkbox\"] {\n",
       " margin-top: 1px;\n",
       " line-height: normal;\n",
       "}\n",
       ".delete-button {\n",
       " border: none !important;\n",
       "}\n",
       "i.fa.fa-trash {\n",
       " font-size: 13.5pt;\n",
       "}\n",
       ".list_container a {\n",
       " font-size: 16px;\n",
       " color: #a2b0c7;\n",
       " border: none;\n",
       " text-shadow: none !important;\n",
       " font-weight: normal;\n",
       " font-style: normal;\n",
       "}\n",
       "div.list_container a:hover {\n",
       " color: #dbe1ea;\n",
       "}\n",
       ".list_header > div input,\n",
       ".list_item > div input {\n",
       " margin-right: 7px;\n",
       " margin-left: 12px;\n",
       " vertical-align: baseline;\n",
       " line-height: 22px;\n",
       " position: relative;\n",
       " top: -1px;\n",
       "}\n",
       "div.list_item:hover {\n",
       " background-color: rgba(75,95,118,.10);\n",
       "}\n",
       ".breadcrumb > li {\n",
       " font-size: 12.0pt;\n",
       " color: #a2b0c7;\n",
       " border: none;\n",
       " text-shadow: none !important;\n",
       "}\n",
       ".breadcrumb > li + li:before {\n",
       " content: \"/\\00a0\";\n",
       " padding: 0px;\n",
       " color: #a2b0c7;\n",
       " font-size: 18px;\n",
       "}\n",
       "#project_name > .breadcrumb {\n",
       " padding: 0px;\n",
       " margin-bottom: 0px;\n",
       " background-color: transparent;\n",
       " font-weight: normal;\n",
       " margin-top: -2px;\n",
       "}\n",
       "ul#tabs a {\n",
       " font-family: sans-serif;\n",
       " font-size: 13.5pt;\n",
       " font-weight: normal;\n",
       " font-style: normal;\n",
       " text-shadow: none !important;\n",
       "}\n",
       ".nav-tabs {\n",
       " font-family: sans-serif;\n",
       " font-size: 13.5pt;\n",
       " font-weight: normal;\n",
       " font-style: normal;\n",
       " background-color: transparent;\n",
       " border-color: transparent;\n",
       " text-shadow: none !important;\n",
       " border: 2px solid transparent;\n",
       "}\n",
       ".nav-tabs > li > a:active,\n",
       ".nav-tabs > li > a:focus,\n",
       ".nav-tabs > li > a:hover,\n",
       ".nav-tabs > li.active > a,\n",
       ".nav-tabs > li.active > a:focus,\n",
       ".nav-tabs > li.active > a:hover,\n",
       ".nav-tabs > li.active > a,\n",
       ".nav-tabs > li.active > a:hover,\n",
       ".nav-tabs > li.active > a:focus {\n",
       " color: #009cd1;\n",
       " background-color: transparent;\n",
       " border-color: transparent;\n",
       " border-bottom: 2px solid transparent;\n",
       "}\n",
       ".nav > li.disabled > a,\n",
       ".nav > li.disabled > a:hover {\n",
       " color: #546386;\n",
       "}\n",
       ".nav-tabs > li > a:before {\n",
       " content: \"\";\n",
       " position: absolute;\n",
       " width: 100%;\n",
       " height: 2px;\n",
       " bottom: -2px;\n",
       " left: 0;\n",
       " background-color: #009cd1;\n",
       " visibility: hidden;\n",
       " -webkit-transform: perspective(0)scaleX(0);\n",
       " transform: perspective(0)scaleX(0);\n",
       " -webkit-transition: ease 220ms;\n",
       " transition: ease 220ms;\n",
       " -webkit-font-smoothing: antialiased !important;\n",
       "}\n",
       ".nav-tabs > li > a:hover:before {\n",
       " visibility: visible;\n",
       " -webkit-transform: perspective(1)scaleX(1);\n",
       " transform: perspective(1)scaleX(1);\n",
       "}\n",
       ".nav-tabs > li.active > a:before {\n",
       " content: \"\";\n",
       " position: absolute;\n",
       " width: 100%;\n",
       " height: 2px;\n",
       " bottom: -2px;\n",
       " left: 0;\n",
       " background-color: #009cd1;\n",
       " visibility: visible;\n",
       " -webkit-transform: perspective(1)scaleX(1);\n",
       " transform: perspective(1)scaleX(1);\n",
       " -webkit-font-smoothing: subpixel-antialiased !important;\n",
       "}\n",
       "div#notebook {\n",
       " font-family: sans-serif;\n",
       " font-size: 13pt;\n",
       " padding-top: 4px;\n",
       "}\n",
       ".notebook_app {\n",
       " background-color: #1a2028;\n",
       "}\n",
       "#notebook-container {\n",
       " padding: 13px 2px;\n",
       " background-color: #1a2028;\n",
       " min-height: 0px;\n",
       " box-shadow: none;\n",
       " width: 980px;\n",
       " margin-right: auto;\n",
       " margin-left: auto;\n",
       "}\n",
       "div#ipython-main-app.container {\n",
       " width: 980px;\n",
       " margin-right: auto;\n",
       " margin-left: auto;\n",
       " margin-right: auto;\n",
       " margin-left: auto;\n",
       "}\n",
       ".container {\n",
       " width: 980px;\n",
       " margin-right: auto;\n",
       " margin-left: auto;\n",
       "}\n",
       "div#menubar-container {\n",
       " width: 100%;\n",
       " width: 980px;\n",
       "}\n",
       "div#header-container {\n",
       " width: 980px;\n",
       "}\n",
       ".notebook_app #header,\n",
       ".edit_app #header {\n",
       " box-shadow: none !important;\n",
       " background-color: #1a2028;\n",
       " border-bottom: 2px solid rgba(75,95,118,.30);\n",
       "}\n",
       "#header,\n",
       ".edit_app #header {\n",
       " font-family: sans-serif;\n",
       " font-size: 13pt;\n",
       " box-shadow: none;\n",
       " background-color: #1a2028;\n",
       "}\n",
       "#header .header-bar,\n",
       ".edit_app #header .header-bar {\n",
       " background: #1a2028;\n",
       " background-color: #1a2028;\n",
       "}\n",
       "body > #header .header-bar {\n",
       " width: 100%;\n",
       " background: #1a2028;\n",
       "}\n",
       "span.checkpoint_status,\n",
       "span.autosave_status {\n",
       " font-size: small;\n",
       " display: none;\n",
       "}\n",
       "#menubar,\n",
       "div#menubar {\n",
       " background-color: #1a2028;\n",
       " padding-top: 0px !important;\n",
       "}\n",
       "#menubar .navbar,\n",
       ".navbar-default {\n",
       " background-color: #1a2028;\n",
       " margin-bottom: 0px;\n",
       " margin-top: 0px;\n",
       "}\n",
       ".navbar {\n",
       " border: none;\n",
       "}\n",
       "div.navbar-text,\n",
       ".navbar-text,\n",
       ".navbar-text.indicator_area,\n",
       "p.navbar-text.indicator_area {\n",
       " margin-top: 8px !important;\n",
       " margin-bottom: 0px;\n",
       " color: #0b98c8;\n",
       "}\n",
       ".navbar-default {\n",
       " font-family: sans-serif;\n",
       " font-size: 13pt;\n",
       " background-color: #1a2028;\n",
       " border-color: #323b48;\n",
       " line-height: 1.5em;\n",
       " padding-bottom: 0px;\n",
       "}\n",
       ".navbar-default .navbar-nav > li > a {\n",
       " font-family: sans-serif;\n",
       " font-size: 13pt;\n",
       " color: #a2b0c7;\n",
       " display: block;\n",
       " line-height: 1.5em;\n",
       " padding-top: 14px;\n",
       " padding-bottom: 11px;\n",
       "}\n",
       ".navbar-default .navbar-nav > li > a:hover,\n",
       ".navbar-default .navbar-nav > li > a:focus {\n",
       " color: #dbe1ea !important;\n",
       " background-color: rgba(75,95,118,.30) !important;\n",
       " border-color: #323b48 !important;\n",
       " line-height: 1.5em;\n",
       " transition: 80ms ease;\n",
       "}\n",
       ".navbar-default .navbar-nav > .open > a,\n",
       ".navbar-default .navbar-nav > .open > a:hover,\n",
       ".navbar-default .navbar-nav > .open > a:focus {\n",
       " color: #fefefe;\n",
       " background-color: #36404e;\n",
       " border-color: #36404e;\n",
       " line-height: 1.5em;\n",
       "}\n",
       ".navbar-nav > li > .dropdown-menu {\n",
       " margin-top: 0px;\n",
       "}\n",
       ".navbar-nav {\n",
       " margin: 0;\n",
       "}\n",
       "div.notification_widget.info,\n",
       ".notification_widget.info,\n",
       ".notification_widget:active:hover,\n",
       ".notification_widget.active:hover,\n",
       ".open > .dropdown-toggle.notification_widget:hover,\n",
       ".notification_widget:active:focus,\n",
       ".notification_widget.active:focus,\n",
       ".open > .dropdown-toggle.notification_widget:focus,\n",
       ".notification_widget:active.focus,\n",
       ".notification_widget.active.focus,\n",
       ".open > .dropdown-toggle.notification_widget.focus,\n",
       "div#notification_notebook.notification_widget.btn.btn-xs.navbar-btn,\n",
       "div#notification_notebook.notification_widget.btn.btn-xs.navbar-btn:hover,\n",
       "div#notification_notebook.notification_widget.btn.btn-xs.navbar-btn:focus {\n",
       " color: #899ab8 !important;\n",
       " background-color: transparent !important;\n",
       " border-color: transparent !important;\n",
       " padding-bottom: 0px !important;\n",
       " margin-bottom: 0px !important;\n",
       " font-size: 9pt !important;\n",
       " z-index: 0;\n",
       "}\n",
       "div#notification_notebook.notification_widget.btn.btn-xs.navbar-btn {\n",
       " font-size: 9pt !important;\n",
       " z-index: 0;\n",
       "}\n",
       ".notification_widget {\n",
       " color: #4c8be2;\n",
       " z-index: -500;\n",
       " font-size: 9pt;\n",
       " background: transparent;\n",
       " background-color: transparent;\n",
       " margin-right: 3px;\n",
       " border: none;\n",
       "}\n",
       ".notification_widget,\n",
       "div.notification_widget {\n",
       " margin-right: 0px;\n",
       " margin-left: 0px;\n",
       " padding-right: 0px;\n",
       " vertical-align: text-top !important;\n",
       " margin-top: 6px !important;\n",
       " background: transparent !important;\n",
       " background-color: transparent !important;\n",
       " font-size: 9pt !important;\n",
       " border: none;\n",
       "}\n",
       ".navbar-btn.btn-xs:hover {\n",
       " border: none !important;\n",
       " background: transparent !important;\n",
       " background-color: transparent !important;\n",
       " color: #a2b0c7 !important;\n",
       "}\n",
       "div.notification_widget.info,\n",
       ".notification_widget.info {\n",
       " display: none !important;\n",
       "}\n",
       ".edit_mode .modal_indicator:before {\n",
       " display: none;\n",
       "}\n",
       ".command_mode .modal_indicator:before {\n",
       " display: none;\n",
       "}\n",
       ".item_icon {\n",
       " color: #4c8be2;\n",
       "}\n",
       ".item_buttons .kernel-name {\n",
       " font-size: 13pt;\n",
       " color: #4c8be2;\n",
       "}\n",
       ".running_notebook_icon:before {\n",
       " color: #48a667 !important;\n",
       " font: normal normal normal 15px/1 FontAwesome;\n",
       " font-size: 15px;\n",
       " text-rendering: auto;\n",
       " -webkit-font-smoothing: antialiased;\n",
       " -moz-osx-font-smoothing: grayscale;\n",
       " content: \"\\f10c\";\n",
       " vertical-align: middle;\n",
       " position: static;\n",
       " display: inherit;\n",
       "}\n",
       ".item_buttons .running-indicator {\n",
       " padding-top: 4px;\n",
       " color: #48a667;\n",
       " font-family: sans-serif;\n",
       " text-rendering: auto;\n",
       " -webkit-font-smoothing: antialiased;\n",
       "}\n",
       "#notification_trusted {\n",
       " font-family: sans-serif;\n",
       " border: none;\n",
       " background: transparent;\n",
       " background-color: transparent;\n",
       " margin-bottom: 0px !important;\n",
       " vertical-align: bottom !important;\n",
       " color: #546386 !important;\n",
       " cursor: default !important;\n",
       "}\n",
       "#notification_area,\n",
       "div.notification_area {\n",
       " float: right !important;\n",
       " position: static;\n",
       " cursor: pointer;\n",
       " padding-top: 6px;\n",
       " padding-right: 4px;\n",
       "}\n",
       "div#notification_notebook.notification_widget.btn.btn-xs.navbar-btn {\n",
       " font-size: 9pt !important;\n",
       " z-index: 0;\n",
       " margin-top: -5px !important;\n",
       "}\n",
       "#modal_indicator {\n",
       " float: right !important;\n",
       " color: #4c8be2;\n",
       " background: #1a2028;\n",
       " background-color: #1a2028;\n",
       " margin-top: 8px !important;\n",
       " margin-left: 0px;\n",
       "}\n",
       "#kernel_indicator {\n",
       " float: right !important;\n",
       " color: #0b98c8;\n",
       " background: #1a2028;\n",
       " background-color: #1a2028;\n",
       " border-left: 2px solid #0b98c8;\n",
       " padding-top: 0px;\n",
       " padding-bottom: 4px;\n",
       " margin-top: 10px !important;\n",
       " margin-left: -2px;\n",
       " padding-left: 5px !important;\n",
       "}\n",
       "#kernel_indicator .kernel_indicator_name {\n",
       " font-size: 17px;\n",
       " color: #0b98c8;\n",
       " background: #1a2028;\n",
       " background-color: #1a2028;\n",
       " padding-left: 5px;\n",
       " padding-right: 5px;\n",
       " margin-top: 4px;\n",
       " vertical-align: text-top;\n",
       " padding-bottom: 0px;\n",
       "}\n",
       ".kernel_idle_icon:before {\n",
       " display: inline-block;\n",
       " font: normal normal normal 22px/1 FontAwesome;\n",
       " font-size: 22px;\n",
       " text-rendering: auto;\n",
       " -webkit-font-smoothing: antialiased;\n",
       " cursor: pointer;\n",
       " margin-left: 0px !important;\n",
       " opacity: 0.7;\n",
       " vertical-align: bottom;\n",
       " margin-top: 1px;\n",
       " content: \"\\f1db\";\n",
       "}\n",
       ".kernel_busy_icon:before {\n",
       " display: inline-block;\n",
       " font: normal normal normal 22px/1 FontAwesome;\n",
       " font-size: 22px;\n",
       " -webkit-animation: pulsate 2s infinite ease-out;\n",
       " animation: pulsate 2s infinite ease-out;\n",
       " text-rendering: auto;\n",
       " -webkit-font-smoothing: antialiased;\n",
       " cursor: pointer;\n",
       " margin-left: 0px !important;\n",
       " vertical-align: bottom;\n",
       " margin-top: 1px;\n",
       " content: \"\\f111\";\n",
       "}\n",
       "@-webkit-keyframes pulsate {\n",
       " 0% {\n",
       "  -webkit-transform: scale(1.0,1.0);\n",
       "  opacity: 0.8;\n",
       " }\n",
       " 8% {\n",
       "  -webkit-transform: scale(1.0,1.0);\n",
       "  opacity: 0.8;\n",
       " }\n",
       " 50% {\n",
       "  -webkit-transform: scale(0.75,0.75);\n",
       "  opacity: 0.3;\n",
       " }\n",
       " 92% {\n",
       "  -webkit-transform: scale(1.0,1.0);\n",
       "  opacity: 0.8;\n",
       " }\n",
       " 100% {\n",
       "  -webkit-transform: scale(1.0,1.0);\n",
       "  opacity: 0.8;\n",
       " }\n",
       "}\n",
       "div.notification_widget.info,\n",
       ".notification_widget.info,\n",
       ".notification_widget:active:hover,\n",
       ".notification_widget.active:hover,\n",
       ".open > .dropdown-toggle.notification_widget:hover,\n",
       ".notification_widget:active:focus,\n",
       ".notification_widget.active:focus,\n",
       ".open > .dropdown-toggle.notification_widget:focus,\n",
       ".notification_widget:active.focus,\n",
       ".notification_widget.active.focus,\n",
       ".open > .dropdown-toggle.notification_widget.focus,\n",
       "div#notification_notebook.notification_widget.btn.btn-xs.navbar-btn,\n",
       "div#notification_notebook.notification_widget.btn.btn-xs.navbar-btn:hover,\n",
       "div#notification_notebook.notification_widget.btn.btn-xs.navbar-btn:focus {\n",
       " color: #899ab8;\n",
       " background-color: #1a2028;\n",
       " border-color: #1a2028;\n",
       "}\n",
       "#notification_area,\n",
       "div.notification_area {\n",
       " float: right !important;\n",
       " position: static;\n",
       "}\n",
       ".notification_widget,\n",
       "div.notification_widget {\n",
       " margin-right: 0px;\n",
       " margin-left: 0px;\n",
       " padding-right: 0px;\n",
       " vertical-align: text-top !important;\n",
       " margin-top: 6px !important;\n",
       " z-index: 1000;\n",
       "}\n",
       "#kernel_logo_widget,\n",
       "#kernel_logo_widget .current_kernel_logo {\n",
       " display: none;\n",
       "}\n",
       "div#ipython_notebook {\n",
       " display: none;\n",
       "}\n",
       "i.fa.fa-icon {\n",
       " -webkit-font-smoothing: antialiased;\n",
       " -moz-osx-font-smoothing: grayscale;\n",
       " text-rendering: auto;\n",
       "}\n",
       ".fa {\n",
       " display: inline-block;\n",
       " font: normal normal normal 10pt/1 \"FontAwesome\", sans-serif;\n",
       " text-rendering: auto;\n",
       " -webkit-font-smoothing: antialiased;\n",
       " -moz-osx-font-smoothing: grayscale;\n",
       "}\n",
       ".dropdown-menu {\n",
       " font-family: sans-serif;\n",
       " font-size: 13pt;\n",
       " box-shadow: none;\n",
       " padding: 0px;\n",
       " text-align: left;\n",
       " border: none;\n",
       " background-color: #36404e;\n",
       " background: #36404e;\n",
       " line-height: 1;\n",
       "}\n",
       ".dropdown-menu:hover {\n",
       " font-family: sans-serif;\n",
       " font-size: 13pt;\n",
       " box-shadow: none;\n",
       " padding: 0px;\n",
       " text-align: left;\n",
       " border: none;\n",
       " background-color: #36404e;\n",
       " box-shadow: none;\n",
       " line-height: 1;\n",
       "}\n",
       ".dropdown-menu > li > a {\n",
       " font-family: sans-serif;\n",
       " font-size: 12.0pt;\n",
       " display: block;\n",
       " padding: 10px 20px 9px 10px;\n",
       " color: #a2b0c7;\n",
       " background-color: #36404e;\n",
       " background: #36404e;\n",
       "}\n",
       ".dropdown-menu > li > a:hover,\n",
       ".dropdown-menu > li > a:focus {\n",
       " color: #dbe1ea;\n",
       " background-color: #323b48;\n",
       " background: #323b48;\n",
       " border-color: #323b48;\n",
       " transition: 200ms ease;\n",
       "}\n",
       ".dropdown-menu .divider {\n",
       " height: 1px;\n",
       " margin: 0px 0px;\n",
       " overflow: hidden;\n",
       " background-color: rgba(75,95,118,.55);\n",
       "}\n",
       ".dropdown-submenu > .dropdown-menu {\n",
       " display: none;\n",
       " top: 2px !important;\n",
       " left: 100%;\n",
       " margin-top: -2px;\n",
       " margin-left: 0px;\n",
       " padding-top: 0px;\n",
       " transition: 200ms ease;\n",
       "}\n",
       ".dropdown-menu > .disabled > a,\n",
       ".dropdown-menu > .disabled > a:hover,\n",
       ".dropdown-menu > .disabled > a:focus {\n",
       " font-family: sans-serif;\n",
       " font-size: 12.0pt;\n",
       " font-weight: normal;\n",
       " color: #546386;\n",
       " padding: none;\n",
       " display: block;\n",
       " clear: both;\n",
       " white-space: nowrap;\n",
       "}\n",
       ".dropdown-submenu > a:after {\n",
       " color: #a2b0c7;\n",
       " margin-right: -16px;\n",
       " margin-top: 0px;\n",
       " display: inline-block;\n",
       "}\n",
       ".dropdown-submenu:hover > a:after,\n",
       ".dropdown-submenu:active > a:after,\n",
       ".dropdown-submenu:focus > a:after,\n",
       ".dropdown-submenu:visited > a:after {\n",
       " color: #0b98c8;\n",
       " margin-right: -16px;\n",
       " display: inline-block !important;\n",
       "}\n",
       "div.kse-dropdown > .dropdown-menu,\n",
       ".kse-dropdown > .dropdown-menu {\n",
       " min-width: 0;\n",
       " top: 94%;\n",
       "}\n",
       ".btn,\n",
       ".btn-default {\n",
       " font-family: sans-serif;\n",
       " color: #a2b0c7;\n",
       " background: #252e3a;\n",
       " background-color: #252e3a;\n",
       " border: 2px solid #252e3a;\n",
       " font-weight: normal;\n",
       " box-shadow: none;\n",
       " text-shadow: none;\n",
       " border-radius: 3px;\n",
       " font-size: initial;\n",
       "}\n",
       ".btn:hover,\n",
       ".btn:active:hover,\n",
       ".btn.active:hover,\n",
       ".btn-default:hover,\n",
       ".open > .dropdown-toggle.btn-default:hover,\n",
       ".open > .dropdown-toggle.btn:hover {\n",
       " color: #009cd1;\n",
       " border: 2px solid #293340;\n",
       " background-color: #293340;\n",
       " background: #293340;\n",
       " background-image: none;\n",
       " box-shadow: none !important;\n",
       " border-radius: 3px;\n",
       "}\n",
       ".btn:active,\n",
       ".btn.active,\n",
       ".btn:active:focus,\n",
       ".btn.active:focus,\n",
       ".btn:active.focus,\n",
       ".btn.active.focus,\n",
       ".btn-default:focus,\n",
       ".btn-default.focus,\n",
       ".btn-default:active,\n",
       ".btn-default.active,\n",
       ".btn-default:active:hover,\n",
       ".btn-default.active:hover,\n",
       ".btn-default:active:focus,\n",
       ".btn-default.active:focus,\n",
       ".btn-default:active.focus,\n",
       ".btn-default.active.focus,\n",
       ".open > .dropdown-toggle.btn:focus,\n",
       ".open > .dropdown-toggle.btn.focus,\n",
       ".open > .dropdown-toggle.btn-default:hover,\n",
       ".open > .dropdown-toggle.btn-default:focus,\n",
       ".open > .dropdown-toggle.btn-default.hover,\n",
       ".open > .dropdown-toggle.btn-default.focus {\n",
       " color: #009cd1;\n",
       " border: 2px solid #293340;\n",
       " background-color: #293340 !important;\n",
       " background: #293340 !important;\n",
       " background-image: none;\n",
       " box-shadow: none !important;\n",
       " border-radius: 3px;\n",
       "}\n",
       ".btn-default:active:hover,\n",
       ".btn-default.active:hover,\n",
       ".btn-default:active:focus,\n",
       ".btn-default.active:focus,\n",
       ".btn-default:active.focus,\n",
       ".btn-default.active.focus {\n",
       " color: #009cd1 !important;\n",
       " background-color: #252e3a;\n",
       " border-color: #33517c !important;\n",
       " transition: 2000ms ease;\n",
       "}\n",
       ".btn:focus,\n",
       ".btn.focus,\n",
       ".btn:active:focus,\n",
       ".btn.active:focus,\n",
       ".btn:active,\n",
       ".btn.active,\n",
       ".btn:active.focus,\n",
       ".btn.active.focus {\n",
       " color: #009cd1 !important;\n",
       " outline: none !important;\n",
       " outline-width: 0px !important;\n",
       " background: #33517c !important;\n",
       " background-color: #33517c !important;\n",
       " border-color: #33517c !important;\n",
       " transition: 200ms ease !important;\n",
       "}\n",
       ".item_buttons > .btn,\n",
       ".item_buttons > .btn-group,\n",
       ".item_buttons > .input-group {\n",
       " font-size: 13pt;\n",
       " background: transparent;\n",
       " background-color: transparent;\n",
       " border: 0px solid #252b35;\n",
       " border-bottom: 2px solid transparent;\n",
       " margin-left: 5px;\n",
       " padding-top: 4px !important;\n",
       "}\n",
       ".item_buttons > .btn:hover,\n",
       ".item_buttons > .btn-group:hover,\n",
       ".item_buttons > .input-group:hover,\n",
       ".item_buttons > .btn.active,\n",
       ".item_buttons > .btn-group.active,\n",
       ".item_buttons > .input-group.active,\n",
       ".item_buttons > .btn.focus {\n",
       " margin-left: 5px;\n",
       " background: #21262f;\n",
       " padding-top: 4px !important;\n",
       " background-color: transparent;\n",
       " border: 0px solid transparent;\n",
       " border-bottom: 2px solid #0b98c8;\n",
       " border-radius: 0px;\n",
       " transition: none;\n",
       "}\n",
       ".item_buttons {\n",
       " line-height: 1.5em !important;\n",
       "}\n",
       ".item_buttons .btn {\n",
       " min-width: 11ex;\n",
       "}\n",
       ".btn-group > .btn:first-child {\n",
       " margin-left: 3px;\n",
       "}\n",
       ".btn-group > .btn-mini,\n",
       ".btn-sm,\n",
       ".btn-group-sm > .btn,\n",
       ".btn-xs,\n",
       ".btn-group-xs > .btn,\n",
       ".alternate_upload .btn-upload,\n",
       ".btn-group,\n",
       ".btn-group-vertical {\n",
       " font-size: inherit;\n",
       " font-weight: normal;\n",
       " height: inherit;\n",
       " line-height: inherit;\n",
       "}\n",
       ".btn-xs,\n",
       ".btn-group-xs > .btn {\n",
       " font-size: initial !important;\n",
       " background-image: none;\n",
       " font-weight: normal;\n",
       " text-shadow: none;\n",
       " display: inline-table;\n",
       " padding: 2px 5px;\n",
       " line-height: 1.45;\n",
       "}\n",
       ".btn-group > .btn:first-child {\n",
       " margin-left: 3px;\n",
       "}\n",
       "div#new-buttons > button,\n",
       "#new-buttons > button,\n",
       "div#refresh_notebook_list,\n",
       "#refresh_notebook_list {\n",
       " background: transparent;\n",
       " background-color: transparent;\n",
       " border: none;\n",
       "}\n",
       "div#new-buttons > button:hover,\n",
       "#new-buttons > button:hover,\n",
       "div#refresh_notebook_list,\n",
       "#refresh_notebook_list,\n",
       "div.alternate_upload .btn-upload,\n",
       ".alternate_upload .btn-upload,\n",
       "div.dynamic-buttons > button,\n",
       ".dynamic-buttons > button,\n",
       ".dynamic-buttons > button:focus,\n",
       ".dynamic-buttons > button:active:focus,\n",
       ".dynamic-buttons > button.active:focus,\n",
       ".dynamic-buttons > button.focus,\n",
       ".dynamic-buttons > button:active.focus,\n",
       ".dynamic-buttons > button.active.focus,\n",
       "#new-buttons > button:focus,\n",
       "#new-buttons > button:active:focus,\n",
       "#new-buttons > button.active:focus,\n",
       "#new-buttons > button.focus,\n",
       "#new-buttons > button:active.focus,\n",
       "#new-buttons > button.active.focus,\n",
       ".alternate_upload .btn-upload:focus,\n",
       ".alternate_upload .btn-upload:active:focus,\n",
       ".alternate_upload .btn-upload.active:focus,\n",
       ".alternate_upload .btn-upload.focus,\n",
       ".alternate_upload .btn-upload:active.focus,\n",
       ".alternate_upload .btn-upload.active.focus {\n",
       " background: transparent !important;\n",
       " background-color: transparent !important;\n",
       " border: none !important;\n",
       "}\n",
       ".alternate_upload input.fileinput {\n",
       " text-align: center;\n",
       " vertical-align: bottom;\n",
       " margin-left: -.5ex;\n",
       " display: inline-table;\n",
       " border: solid 0px #252e3a;\n",
       " margin-bottom: -1ex;\n",
       "}\n",
       ".alternate_upload .btn-upload {\n",
       " display: inline-table;\n",
       " background: transparent;\n",
       " border: none;\n",
       "}\n",
       ".btn-group .btn + .btn,\n",
       ".btn-group .btn + .btn-group,\n",
       ".btn-group .btn-group + .btn,\n",
       ".btn-group .btn-group + .btn-group {\n",
       " margin-left: -2px;\n",
       "}\n",
       ".btn-group > .btn:first-child:not(:last-child):not(.dropdown-toggle) {\n",
       " border-bottom-right-radius: 0;\n",
       " border-top-right-radius: 0;\n",
       " z-index: 2;\n",
       "}\n",
       ".dropdown-header {\n",
       " font-family: sans-serif !important;\n",
       " font-size: 13pt !important;\n",
       " color: #0b98c8 !important;\n",
       " border-bottom: none !important;\n",
       " padding: 0px !important;\n",
       " margin: 6px 6px 0px !important;\n",
       "}\n",
       "span#last-modified.btn.btn-xs.btn-default.sort-action,\n",
       "span#sort-name.btn.btn-xs.btn-default.sort-action,\n",
       "span#file-size.btn.btn-xs.btn-default.sort-action {\n",
       " font-family: sans-serif;\n",
       " font-size: 16px;\n",
       " background-color: transparent;\n",
       " background: transparent;\n",
       " border: none;\n",
       " color: #a2b0c7;\n",
       " padding-bottom: 0px;\n",
       " margin-bottom: 0px;\n",
       " vertical-align: sub;\n",
       "}\n",
       "span#last-modified.btn.btn-xs.btn-default.sort-action {\n",
       " margin-left: 19px;\n",
       "}\n",
       "button.close {\n",
       " border: 0px none;\n",
       " font-family: sans-serif;\n",
       " font-size: 20pt;\n",
       " font-weight: normal;\n",
       "}\n",
       ".dynamic-buttons {\n",
       " padding-top: 0px;\n",
       " display: inline-block;\n",
       "}\n",
       ".close {\n",
       " color: #dc6972;\n",
       " opacity: .5;\n",
       " text-shadow: none;\n",
       " font-weight: normal;\n",
       "}\n",
       ".close:hover {\n",
       " color: #dc6972;\n",
       " opacity: 1;\n",
       " font-weight: normal;\n",
       "}\n",
       "div.nbext-enable-btns .btn[disabled],\n",
       "div.nbext-enable-btns .btn[disabled]:hover,\n",
       ".btn-default.disabled,\n",
       ".btn-default[disabled],\n",
       ".btn-default.disabled:hover,\n",
       ".btn-default[disabled]:hover,\n",
       "fieldset[disabled] .btn-default:hover,\n",
       ".btn-default.disabled:focus,\n",
       ".btn-default[disabled]:focus,\n",
       "fieldset[disabled] .btn-default:focus,\n",
       ".btn-default.disabled.focus,\n",
       ".btn-default[disabled].focus,\n",
       "fieldset[disabled] .btn-default.focus {\n",
       " color: #92a2bd;\n",
       " background: #232c37;\n",
       " background-color: #232c37;\n",
       " border-color: #232c37;\n",
       " transition: 200ms ease;\n",
       "}\n",
       ".input-group-addon {\n",
       " padding: 2px 5px;\n",
       " font-size: 13pt;\n",
       " font-weight: normal;\n",
       " height: auto;\n",
       " color: #a2b0c7;\n",
       " text-align: center;\n",
       " background-color: transparent;\n",
       " border: 2px solid transparent !important;\n",
       " text-transform: capitalize;\n",
       "}\n",
       "a.btn.btn-default.input-group-addon:hover {\n",
       " background: transparent !important;\n",
       " background-color: transparent !important;\n",
       "}\n",
       ".btn-group > .btn + .dropdown-toggle {\n",
       " padding-left: 8px;\n",
       " padding-right: 8px;\n",
       " height: 100%;\n",
       "}\n",
       ".btn-group > .btn + .dropdown-toggle:hover {\n",
       " background: #293340 !important;\n",
       "}\n",
       ".input-group-btn {\n",
       " position: relative;\n",
       " font-size: inherit;\n",
       " white-space: nowrap;\n",
       " background: #252b35;\n",
       " background-color: #252b35;\n",
       " border: none;\n",
       "}\n",
       ".input-group-btn:hover {\n",
       " background: #21262f;\n",
       " background-color: #21262f;\n",
       " border: none;\n",
       "}\n",
       ".input-group-btn:first-child > .btn,\n",
       ".input-group-btn:first-child > .btn-group {\n",
       " background: #252b35;\n",
       " background-color: #252b35;\n",
       " border: none;\n",
       " margin-left: 2px;\n",
       " margin-right: -1px;\n",
       " font-size: inherit;\n",
       "}\n",
       ".input-group-btn:first-child > .btn:hover,\n",
       ".input-group-btn:first-child > .btn-group:hover {\n",
       " background: #293340;\n",
       " background-color: #293340;\n",
       " border: none;\n",
       " font-size: inherit;\n",
       " transition: 200ms ease;\n",
       "}\n",
       "div.modal .btn-group > .btn:first-child {\n",
       " background: #252b35;\n",
       " background-color: #252b35;\n",
       " border: 1px solid #232932;\n",
       " margin-top: 0px !important;\n",
       " margin-left: 0px;\n",
       " margin-bottom: 2px;\n",
       "}\n",
       "div.modal .btn-group > .btn:first-child:hover {\n",
       " background: #21262f;\n",
       " background-color: #21262f;\n",
       " border: 1px solid #21262f;\n",
       " transition: 200ms ease;\n",
       "}\n",
       "div.modal > button,\n",
       "div.modal-footer > button {\n",
       " background: #252b35;\n",
       " background-color: #252b35;\n",
       " border-color: #252b35;\n",
       "}\n",
       "div.modal > button:hover,\n",
       "div.modal-footer > button:hover {\n",
       " background: #21262f;\n",
       " background-color: #21262f;\n",
       " border-color: #21262f;\n",
       " transition: 200ms ease;\n",
       "}\n",
       ".modal-content {\n",
       " font-family: sans-serif;\n",
       " font-size: 12.0pt;\n",
       " position: relative;\n",
       " background: #252b35;\n",
       " background-color: #252b35;\n",
       " border: none;\n",
       " border-radius: 1px;\n",
       " background-clip: padding-box;\n",
       " outline: none;\n",
       "}\n",
       ".modal-header {\n",
       " font-family: sans-serif;\n",
       " font-size: 13pt;\n",
       " color: #a2b0c7;\n",
       " background: #252b35;\n",
       " background-color: #252b35;\n",
       " border-color: rgba(75,95,118,.30);\n",
       " padding: 12px;\n",
       " min-height: 16.4286px;\n",
       "}\n",
       ".modal-content h4 {\n",
       " font-family: sans-serif;\n",
       " font-size: 16pt;\n",
       " color: #a2b0c7;\n",
       " padding: 5px;\n",
       "}\n",
       ".modal-body {\n",
       " background-color: #2d3846;\n",
       " position: relative;\n",
       " padding: 15px;\n",
       "}\n",
       ".modal-footer {\n",
       " padding: 8px;\n",
       " text-align: right;\n",
       " background-color: #2d3846;\n",
       " border-top: none;\n",
       "}\n",
       ".alert-info {\n",
       " background-color: #323f50;\n",
       " border-color: rgba(75,95,118,.30);\n",
       " color: #a2b0c7;\n",
       "}\n",
       ".modal-header .close {\n",
       " margin-top: -5px;\n",
       " font-size: 25pt;\n",
       "}\n",
       ".modal-backdrop,\n",
       ".modal-backdrop.in {\n",
       " opacity: 0.85;\n",
       " background-color: notebook-bg;\n",
       "}\n",
       "div.panel,\n",
       "div.panel-default,\n",
       ".panel,\n",
       ".panel-default {\n",
       " font-family: sans-serif;\n",
       " font-size: 13pt;\n",
       " background-color: #2d3846;\n",
       " color: #a2b0c7;\n",
       " margin-bottom: 14px;\n",
       " border: 0;\n",
       " box-shadow: none;\n",
       "}\n",
       "div.panel > .panel-heading,\n",
       "div.panel-default > .panel-heading {\n",
       " font-size: 14pt;\n",
       " color: #a2b0c7;\n",
       " background: #252b35;\n",
       " background-color: #252b35;\n",
       " border: 0;\n",
       "}\n",
       ".modal .modal-dialog {\n",
       " min-width: 950px;\n",
       " margin: 50px auto;\n",
       "}\n",
       "div.container-fluid {\n",
       " margin-right: auto;\n",
       " margin-left: auto;\n",
       " padding-left: 0px;\n",
       " padding-right: 5px;\n",
       "}\n",
       "div.form-control,\n",
       ".form-control {\n",
       " font-family: sans-serif;\n",
       " font-size: initial;\n",
       " color: #a2b0c7;\n",
       " background-color: #252b35;\n",
       " border: 1px solid #252e3a !important;\n",
       " margin-left: 2px;\n",
       " box-shadow: none;\n",
       " transition: border-color 0.15s ease-in-out 0s, box-shadow 0.15s ease-in-out 0s;\n",
       "}\n",
       ".form-control-static {\n",
       " min-height: inherit;\n",
       " height: inherit;\n",
       "}\n",
       ".form-group.list-group-item {\n",
       " color: #a2b0c7;\n",
       " background-color: #2d3846;\n",
       " border-color: rgba(75,95,118,.30);\n",
       " margin-bottom: 0px;\n",
       "}\n",
       ".form-group .input-group {\n",
       " float: left;\n",
       "}\n",
       "input,\n",
       "button,\n",
       "select,\n",
       "textarea {\n",
       " background-color: #252b35;\n",
       " font-weight: normal;\n",
       " border: 1px solid rgba(75,95,118,.30);\n",
       "}\n",
       "select.form-control.select-xs {\n",
       " height: 33px;\n",
       " font-size: 13pt;\n",
       "}\n",
       ".toolbar select,\n",
       ".toolbar label {\n",
       " width: auto;\n",
       " vertical-align: middle;\n",
       " margin-right: 0px;\n",
       " margin-bottom: 0px;\n",
       " display: inline;\n",
       " font-size: 92%;\n",
       " margin-left: 10px;\n",
       " padding: 0px;\n",
       " background: #252e3a !important;\n",
       " background-color: #252e3a !important;\n",
       " border: 2px solid #212934 !important;\n",
       "}\n",
       ".form-control:focus {\n",
       " border-color: #0b98c8;\n",
       " outline: 2px solid rgba(0,156,209,.5);\n",
       " -webkit-box-shadow: none;\n",
       "}\n",
       "::-webkit-input-placeholder {\n",
       " color: #546386;\n",
       "}\n",
       "::-moz-placeholder {\n",
       " color: #546386;\n",
       "}\n",
       ":-ms-input-placeholder {\n",
       " color: #546386;\n",
       "}\n",
       ":-moz-placeholder {\n",
       " color: #546386;\n",
       "}\n",
       "[dir=\"ltr\"] #find-and-replace .input-group-btn + .form-control {\n",
       " border: 2px solid rgba(75,95,118,.30) !important;\n",
       "}\n",
       "[dir=\"ltr\"] #find-and-replace .input-group-btn + .form-control:focus {\n",
       " border-color: #0b98c8;\n",
       " outline: 2px solid rgba(0,156,209,.5);\n",
       " -webkit-box-shadow: none;\n",
       " box-shadow: none;\n",
       "}\n",
       "div.output.output_scroll {\n",
       " box-shadow: none;\n",
       "}\n",
       "::-webkit-scrollbar {\n",
       " width: 11px;\n",
       " max-height: 9px;\n",
       " background-color: #292d3a;\n",
       " border-radius: 3px;\n",
       " border: none;\n",
       "}\n",
       "::-webkit-scrollbar-track {\n",
       " background: #292d3a;\n",
       " border: none;\n",
       " width: 11px;\n",
       " max-height: 9px;\n",
       "}\n",
       "::-webkit-scrollbar-thumb {\n",
       " border-radius: 2px;\n",
       " border: none;\n",
       " background: #3f4555;\n",
       " background-clip: content-box;\n",
       " width: 11px;\n",
       "}\n",
       "HTML,\n",
       "body,\n",
       "div,\n",
       "dl,\n",
       "dt,\n",
       "dd,\n",
       "ul,\n",
       "ol,\n",
       "li,\n",
       "h1,\n",
       "h2,\n",
       "h3,\n",
       "h4,\n",
       "h5,\n",
       "h6,\n",
       "pre,\n",
       "code,\n",
       "form,\n",
       "fieldset,\n",
       "legend,\n",
       "input,\n",
       "button,\n",
       "textarea,\n",
       "p,\n",
       "blockquote,\n",
       "th,\n",
       "td,\n",
       "span,\n",
       "a {\n",
       " text-rendering: geometricPrecision;\n",
       " -webkit-font-smoothing: subpixel-antialiased;\n",
       " font-weight: 400;\n",
       "}\n",
       "div.input_area {\n",
       " background-color: #293340;\n",
       " background: #293340;\n",
       " padding-right: 1.2em;\n",
       " border: 0px;\n",
       " border-radius: 0px;\n",
       " border-top-right-radius: 4px;\n",
       " border-bottom-right-radius: 4px;\n",
       "}\n",
       "div.cell {\n",
       " padding: 0px;\n",
       " background: #293340;\n",
       " background-color: #293340;\n",
       " border: medium solid #1a2028;\n",
       " border-radius: 4px;\n",
       " top: 0;\n",
       "}\n",
       "div.cell.selected {\n",
       " background: #293340;\n",
       " background-color: #293340;\n",
       " border: medium solid #1a2028;\n",
       " padding: 0px;\n",
       " border-radius: 5px;\n",
       "}\n",
       ".edit_mode div.cell.selected {\n",
       " padding: 0px;\n",
       " background: #293340;\n",
       " background-color: #293340;\n",
       " border: medium solid #1a2028;\n",
       " border-radius: 5px;\n",
       "}\n",
       "div.cell.edit_mode {\n",
       " padding: 0px;\n",
       " background: #293340;\n",
       " background-color: #293340;\n",
       "}\n",
       "div.CodeMirror-sizer {\n",
       " margin-left: 0px;\n",
       " margin-bottom: -21px;\n",
       " border-right-width: 16px;\n",
       " min-height: 37px;\n",
       " padding-right: 0px;\n",
       " padding-bottom: 0px;\n",
       " margin-top: 0px;\n",
       "}\n",
       "div.cell.selected:before,\n",
       ".edit_mode div.cell.selected:before,\n",
       "div.cell.selected:before,\n",
       "div.cell.selected.jupyter-soft-selected:before {\n",
       " background: #293340 !important;\n",
       " border: none;\n",
       " border-radius: 3px;\n",
       " position: absolute;\n",
       " display: block;\n",
       " top: 0px;\n",
       " left: 0px;\n",
       " width: 0px;\n",
       " height: 100%;\n",
       "}\n",
       "div.cell.text_cell.selected::before,\n",
       ".edit_mode div.cell.text_cell.selected:before,\n",
       "div.cell.text_cell.selected:before,\n",
       "div.cell.text_cell.selected.jupyter-soft-selected:before {\n",
       " background: #293340 !important;\n",
       " background-color: #293340 !important;\n",
       " border-color: #0b98c8 !important;\n",
       "}\n",
       "div.cell.code_cell .input {\n",
       " border-left: 5px solid #293340 !important;\n",
       " border-radius: 3px;\n",
       " border-bottom-left-radius: 3px;\n",
       " border-top-left-radius: 3px;\n",
       "}\n",
       "div.cell.code_cell.selected .input {\n",
       " border-left: 5px solid #008ebf !important;\n",
       " border-radius: 3px;\n",
       "}\n",
       ".edit_mode div.cell.code_cell.selected .input {\n",
       " border-left: 5px solid #005573 !important;\n",
       " border-radius: 3px;\n",
       "}\n",
       ".edit_mode div.cell.selected:before {\n",
       " height: 100%;\n",
       " border-left: 5px solid #005573 !important;\n",
       " border-radius: 3px;\n",
       "}\n",
       "div.cell.jupyter-soft-selected,\n",
       "div.cell.selected.jupyter-soft-selected {\n",
       " border-left-color: #005573 !important;\n",
       " border-left-width: 0px !important;\n",
       " padding-left: 7px !important;\n",
       " border-right-color: #005573 !important;\n",
       " border-right-width: 0px !important;\n",
       " background: #005573 !important;\n",
       " border-radius: 6px !important;\n",
       "}\n",
       "div.cell.selected.jupyter-soft-selected .input {\n",
       " border-left: 5px solid #293340 !important;\n",
       "}\n",
       "div.cell.selected.jupyter-soft-selected {\n",
       " border-left-color: #008ebf;\n",
       " border-color: #1a2028;\n",
       " padding-left: 7px;\n",
       " border-radius: 6px;\n",
       "}\n",
       "div.cell.code_cell.selected .input {\n",
       " border-left: none;\n",
       " border-radius: 3px;\n",
       "}\n",
       "div.cell.selected.jupyter-soft-selected .prompt,\n",
       "div.cell.text_cell.selected.jupyter-soft-selected .prompt {\n",
       " top: 0;\n",
       " border-left: #293340 !important;\n",
       " border-radius: 2px;\n",
       "}\n",
       "div.cell.text_cell.selected.jupyter-soft-selected .input_prompt {\n",
       " border-left: none !important;\n",
       "}\n",
       "div.cell.text_cell.jupyter-soft-selected,\n",
       "div.cell.text_cell.selected.jupyter-soft-selected {\n",
       " border-left-color: #005573 !important;\n",
       " border-left-width: 0px !important;\n",
       " padding-left: 26px !important;\n",
       " border-right-color: #005573 !important;\n",
       " border-right-width: 0px !important;\n",
       " background: #005573 !important;\n",
       " border-radius: 5px !important;\n",
       "}\n",
       "div.cell.jupyter-soft-selected .input,\n",
       "div.cell.selected.jupyter-soft-selected .input {\n",
       " border-left-color: #005573 !important;\n",
       "}\n",
       "div.prompt,\n",
       ".prompt {\n",
       " font-family: monospace, monospace;\n",
       " font-size: 9pt !important;\n",
       " font-weight: normal;\n",
       " color: #546386;\n",
       " line-height: 170%;\n",
       " padding: 0px;\n",
       " padding-top: 4px;\n",
       " padding-left: 0px;\n",
       " padding-right: 1px;\n",
       " text-align: right !important;\n",
       " min-width: 11.5ex !important;\n",
       " width: 11.5ex !important;\n",
       "}\n",
       "div.prompt.input_prompt {\n",
       " font-size: 9pt !important;\n",
       " background-color: #293340;\n",
       " border-top: 0px;\n",
       " border-top-right-radius: 0px;\n",
       " border-bottom-left-radius: 0px;\n",
       " border-bottom-right-radius: 0px;\n",
       " padding-right: 3px;\n",
       " min-width: 11.5ex;\n",
       " width: 11.5ex !important;\n",
       "}\n",
       "div.cell.code_cell .input_prompt {\n",
       " border-right: 2px solid rgba(0,156,209,.5);\n",
       "}\n",
       "div.cell.selected .prompt {\n",
       " top: 0;\n",
       "}\n",
       ".edit_mode div.cell.selected .prompt {\n",
       " top: 0;\n",
       "}\n",
       ".edit_mode div.cell.selected .prompt {\n",
       " top: 0;\n",
       "}\n",
       ".run_this_cell {\n",
       " visibility: hidden;\n",
       " color: transparent;\n",
       " padding-top: 0px;\n",
       " padding-bottom: 0px;\n",
       " padding-left: 3px;\n",
       " padding-right: 12px;\n",
       " width: 1.5ex;\n",
       " width: 0ex;\n",
       " background: transparent;\n",
       " background-color: transparent;\n",
       "}\n",
       "div.code_cell:hover div.input .run_this_cell {\n",
       " visibility: visible;\n",
       "}\n",
       "div.cell.code_cell.rendered.selected .run_this_cell:hover {\n",
       " background-color: #212934;\n",
       " background: #212934;\n",
       " color: #008ebf !important;\n",
       "}\n",
       "div.cell.code_cell.rendered.unselected .run_this_cell:hover {\n",
       " background-color: #212934;\n",
       " background: #212934;\n",
       " color: #008ebf !important;\n",
       "}\n",
       "i.fa-step-forward.fa {\n",
       " display: inline-block;\n",
       " font: normal normal normal 9px \"FontAwesome\";\n",
       "}\n",
       ".fa-step-forward:before {\n",
       " content: \"\\f04b\";\n",
       "}\n",
       "div.cell.selected.jupyter-soft-selected .run_this_cell,\n",
       "div.cell.selected.jupyter-soft-selected .run_this_cell:hover,\n",
       "div.cell.unselected.jupyter-soft-selected .run_this_cell:hover,\n",
       "div.cell.code_cell.rendered.selected.jupyter-soft-selected .run_this_cell:hover,\n",
       "div.cell.code_cell.rendered.unselected.jupyter-soft-selected .run_this_cell:hover {\n",
       " background-color: #005573 !important;\n",
       " background: #005573 !important;\n",
       " color: #005573 !important;\n",
       "}\n",
       "div.output_wrapper {\n",
       " background-color: #323a48;\n",
       " border: 0px;\n",
       " left: 0px;\n",
       " margin-bottom: 0em;\n",
       " margin-top: 0em;\n",
       " border-top-right-radius: 0px;\n",
       " border-top-left-radius: 0px;\n",
       "}\n",
       "div.output_subarea.output_text.output_stream.output_stdout,\n",
       "div.output_subarea.output_text {\n",
       " font-family: monospace, monospace;\n",
       " font-size: 8.5pt !important;\n",
       " line-height: 150% !important;\n",
       " background-color: #323a48;\n",
       " color: #b4bcde;\n",
       " border-top-right-radius: 0px;\n",
       " border-top-left-radius: 0px;\n",
       " margin-left: 11.5px;\n",
       "}\n",
       "div.output_area pre {\n",
       " font-family: monospace, monospace;\n",
       " font-size: 8.5pt !important;\n",
       " line-height: 151% !important;\n",
       " color: #b4bcde;\n",
       " border-top-right-radius: 0px;\n",
       " border-top-left-radius: 0px;\n",
       "}\n",
       "div.output_area {\n",
       " display: -webkit-box;\n",
       "}\n",
       "div.output_html {\n",
       " font-family: monospace, monospace;\n",
       " font-size: 8.5pt;\n",
       " color: #e2e5f2;\n",
       " background-color: #323a48;\n",
       " background: #323a48;\n",
       "}\n",
       "div.output_subarea {\n",
       " overflow-x: auto;\n",
       " padding: 1.2em !important;\n",
       " -webkit-box-flex: 1;\n",
       " -moz-box-flex: 1;\n",
       " box-flex: 1;\n",
       " flex: 1;\n",
       "}\n",
       "div.btn.btn-default.output_collapsed {\n",
       " background: #1b1f26;\n",
       " background-color: #1b1f26;\n",
       " border-color: #1b1f26;\n",
       "}\n",
       "div.btn.btn-default.output_collapsed:hover {\n",
       " background: #161a20;\n",
       " background-color: #161a20;\n",
       " border-color: #161a20;\n",
       "}\n",
       "div.prompt.output_prompt {\n",
       " font-family: monospace, monospace;\n",
       " font-weight: bold !important;\n",
       " background-color: #323a48;\n",
       " color: transparent;\n",
       " border-bottom-left-radius: 4px;\n",
       " border-top-right-radius: 0px;\n",
       " border-top-left-radius: 0px;\n",
       " border-bottom-right-radius: 0px;\n",
       " min-width: 11.5ex !important;\n",
       " width: 11.5ex !important;\n",
       " border-right: 2px solid transparent;\n",
       "}\n",
       "div.out_prompt_overlay.prompt {\n",
       " font-family: monospace, monospace;\n",
       " font-weight: bold !important;\n",
       " background-color: #323a48;\n",
       " border-bottom-left-radius: 2px;\n",
       " border-top-right-radius: 0px;\n",
       " border-top-left-radius: 0px;\n",
       " border-bottom-right-radius: 0px;\n",
       " min-width: 11.5ex !important;\n",
       " width: 11.5ex !important;\n",
       " border-right: 2px solid transparent;\n",
       " color: transparent;\n",
       "}\n",
       "div.out_prompt_overlay.prompt:hover {\n",
       " background-color: #374556;\n",
       " box-shadow: none !important;\n",
       " border: none;\n",
       " border-bottom-left-radius: 2px;\n",
       " -webkit-border-: 2px;\n",
       " -moz-border-radius: 2px;\n",
       " border-top-right-radius: 0px;\n",
       " border-top-left-radius: 0px;\n",
       " min-width: 11.5ex !important;\n",
       " width: 11.5ex !important;\n",
       " border-right: 2px solid #374556 !important;\n",
       "}\n",
       "div.cell.code_cell .output_prompt {\n",
       " border-right: 2px solid transparent;\n",
       " color: transparent;\n",
       "}\n",
       "div.cell.selected .output_prompt,\n",
       "div.cell.selected .out_prompt_overlay.prompt {\n",
       " border-left: 5px solid #005573;\n",
       " border-right: 2px solid #323a48;\n",
       " border-radius: 0px !important;\n",
       "}\n",
       ".edit_mode div.cell.selected .output_prompt,\n",
       ".edit_mode div.cell.selected .out_prompt_overlay.prompt {\n",
       " border-left: 5px solid #005573;\n",
       " border-right: 2px solid #323a48;\n",
       " border-radius: 0px !important;\n",
       "}\n",
       "div.text_cell,\n",
       "div.text_cell_render pre,\n",
       "div.text_cell_render {\n",
       " font-family: sans-serif;\n",
       " font-size: 13pt;\n",
       " line-height: 130% !important;\n",
       " color: #b0bdd7;\n",
       " background: #293340;\n",
       " background-color: #293340;\n",
       " border-radius: 0px;\n",
       "}\n",
       "div .text_cell_render {\n",
       " padding: 0.4em 0.4em 0.4em 0.4em;\n",
       "}\n",
       "div.cell.text_cell .CodeMirror-lines {\n",
       " padding-top: .7em !important;\n",
       " padding-bottom: .4em !important;\n",
       " padding-left: .5em !important;\n",
       " padding-right: .5em !important;\n",
       " margin-top: .4em;\n",
       " margin-bottom: .3em;\n",
       "}\n",
       "div.cell.text_cell.unrendered div.input_area,\n",
       "div.cell.text_cell.rendered div.input_area {\n",
       " background-color: #293340;\n",
       " background: #293340;\n",
       " border: 0px;\n",
       " border-radius: 2px;\n",
       "}\n",
       "div.cell.text_cell .CodeMirror,\n",
       "div.cell.text_cell .CodeMirror pre {\n",
       " line-height: 170% !important;\n",
       "}\n",
       "div.cell.text_cell.rendered.selected {\n",
       " font-family: sans-serif;\n",
       " line-height: 170% !important;\n",
       " background: #293340;\n",
       " background-color: #293340;\n",
       " border-radius: 0px;\n",
       "}\n",
       "div.cell.text_cell.unrendered.selected {\n",
       " font-family: sans-serif;\n",
       " line-height: 170% !important;\n",
       " background: #293340;\n",
       " background-color: #293340;\n",
       " border-radius: 0px;\n",
       "}\n",
       "div.cell.text_cell.selected {\n",
       " font-family: sans-serif;\n",
       " line-height: 170% !important;\n",
       " background: #293340;\n",
       " background-color: #293340;\n",
       " border-radius: 0px;\n",
       "}\n",
       ".edit_mode div.cell.text_cell.selected {\n",
       " font-family: sans-serif;\n",
       " line-height: 170% !important;\n",
       " background: #293340;\n",
       " background-color: #293340;\n",
       " border-radius: 0px;\n",
       "}\n",
       "div.text_cell.unrendered,\n",
       "div.text_cell.unrendered.selected,\n",
       "div.edit_mode div.text_cell.unrendered {\n",
       " font-family: sans-serif;\n",
       " line-height: 170% !important;\n",
       " background: #293340;\n",
       " background-color: #293340;\n",
       " border-radius: 0px;\n",
       "}\n",
       "div.cell.text_cell .prompt {\n",
       " border-right: 0;\n",
       " min-width: 11.5ex !important;\n",
       " width: 11.5ex !important;\n",
       "}\n",
       "div.cell.text_cell.rendered .prompt {\n",
       " font-family: monospace, monospace;\n",
       " font-size: 9.5pt !important;\n",
       " font-weight: normal;\n",
       " color: #546386 !important;\n",
       " text-align: right !important;\n",
       " min-width: 14.5ex !important;\n",
       " width: 14.5ex !important;\n",
       " background-color: #293340;\n",
       " border-right: 2px solid rgba(0,156,209,.5);\n",
       " border-left: 4px solid #293340;\n",
       "}\n",
       "div.cell.text_cell.unrendered .prompt {\n",
       " font-family: monospace, monospace;\n",
       " font-size: 9.5pt !important;\n",
       " font-weight: normal;\n",
       " color: #546386 !important;\n",
       " text-align: right !important;\n",
       " min-width: 14.5ex !important;\n",
       " width: 14.5ex !important;\n",
       " border-right: 2px solid rgba(0,156,209,.5);\n",
       " border-left: 4px solid #293340;\n",
       " background-color: #293340;\n",
       "}\n",
       "div.cell.text_cell.rendered .prompt {\n",
       " border-right: 2px solid rgba(0,156,209,.5);\n",
       "}\n",
       "div.cell.text_cell.rendered.selected .prompt {\n",
       " top: 0;\n",
       " border-left: 4px solid #0b98c8;\n",
       " border-right: 2px solid rgba(0,156,209,.5);\n",
       "}\n",
       "div.text_cell.unrendered.selected .prompt,\n",
       "div.text_cell.rendered.selected .prompt {\n",
       " top: 0;\n",
       " background: #293340;\n",
       " border-left: 4px solid #005573;\n",
       " border-right: 2px solid rgba(0,156,209,.5);\n",
       "}\n",
       "div.rendered_html code {\n",
       " font-family: monospace, monospace;\n",
       " font-size: 11pt;\n",
       " padding-top: 3px;\n",
       " padding-left: 2px;\n",
       " color: #cdd2e9;\n",
       " background: #252e3a;\n",
       " background-color: #252e3a;\n",
       "}\n",
       "pre,\n",
       "code,\n",
       "kbd,\n",
       "samp {\n",
       " white-space: pre-wrap;\n",
       "}\n",
       ".well code,\n",
       "code {\n",
       " font-family: monospace, monospace;\n",
       " font-size: 11pt !important;\n",
       " line-height: 170% !important;\n",
       " color: #b0bdd7;\n",
       " background: #252e3a;\n",
       " background-color: #252e3a;\n",
       " border-color: #252e3a;\n",
       "}\n",
       "kbd {\n",
       " padding: 1px;\n",
       " font-size: 11pt;\n",
       " font-weight: 800;\n",
       " color: #cdd2e9;\n",
       " background-color: transparent !important;\n",
       " border: 0;\n",
       " box-shadow: none;\n",
       "}\n",
       "pre {\n",
       " display: block;\n",
       " padding: 8.5px;\n",
       " margin: 0 0 9px;\n",
       " font-size: 12.0pt;\n",
       " line-height: 1.42857143;\n",
       " color: #cdd2e9;\n",
       " background-color: #252e3a;\n",
       " border: 1px solid #252e3a;\n",
       " border-radius: 2px;\n",
       "}\n",
       "div.rendered_html {\n",
       " color: #b0bdd7;\n",
       "}\n",
       ".rendered_html * + ul {\n",
       " margin-top: .4em;\n",
       " margin-bottom: .3em;\n",
       "}\n",
       ".rendered_html * + p {\n",
       " margin-top: .5em;\n",
       " margin-bottom: .5em;\n",
       "}\n",
       "div.rendered_html pre {\n",
       " font-family: monospace, monospace;\n",
       " font-size: 11pt !important;\n",
       " line-height: 170% !important;\n",
       " color: #b0bdd7 !important;\n",
       " background: #252e3a;\n",
       " background-color: #252e3a;\n",
       " max-width: 80%;\n",
       " border-radius: 0px;\n",
       " border-left: 3px solid #252e3a;\n",
       " max-width: 80%;\n",
       " border-radius: 0px;\n",
       " padding-left: 5px;\n",
       " margin-left: 6px;\n",
       "}\n",
       "div.text_cell_render pre,\n",
       "div.text_cell_render code {\n",
       " font-family: monospace, monospace;\n",
       " font-size: 11pt !important;\n",
       " line-height: 170% !important;\n",
       " color: #b0bdd7;\n",
       " background: #1a2028;\n",
       " background-color: #1a2028;\n",
       " max-width: 80%;\n",
       " border-radius: 0px;\n",
       " border-left: none;\n",
       "}\n",
       "div.text_cell_render pre {\n",
       " border-left: 3px solid rgba(0,156,209,.5) !important;\n",
       " max-width: 80%;\n",
       " border-radius: 0px;\n",
       " padding-left: 5px;\n",
       " margin-left: 6px;\n",
       "}\n",
       "div.text_cell_render h1,\n",
       "div.rendered_html h1,\n",
       "div.text_cell_render h2,\n",
       "div.rendered_html h2,\n",
       "div.text_cell_render h3,\n",
       "div.rendered_html h3,\n",
       "div.text_cell_render h4,\n",
       "div.rendered_html h4,\n",
       "div.text_cell_render h5,\n",
       "div.rendered_html h5 {\n",
       " font-family: sans-serif;\n",
       " margin: 0.4em .2em .3em .2em !important;\n",
       "}\n",
       ".rendered_html h1:first-child,\n",
       ".rendered_html h2:first-child,\n",
       ".rendered_html h3:first-child,\n",
       ".rendered_html h4:first-child,\n",
       ".rendered_html h5:first-child,\n",
       ".rendered_html h6:first-child {\n",
       " margin-top: 0.2em !important;\n",
       " margin-bottom: 0.2em !important;\n",
       "}\n",
       ".rendered_html h1,\n",
       ".text_cell_render h1 {\n",
       " color: #0b98c8 !important;\n",
       " font-size: 200%;\n",
       " text-align: left;\n",
       " font-style: normal;\n",
       " font-weight: normal;\n",
       "}\n",
       ".rendered_html h2,\n",
       ".text_cell_render h2 {\n",
       " color: #0b98c8 !important;\n",
       " font-size: 170%;\n",
       " font-style: normal;\n",
       " font-weight: normal;\n",
       "}\n",
       ".rendered_html h3,\n",
       ".text_cell_render h3 {\n",
       " color: #0b98c8 !important;\n",
       " font-size: 140%;\n",
       " font-style: normal;\n",
       " font-weight: normal;\n",
       "}\n",
       ".rendered_html h4,\n",
       ".text_cell_render h4 {\n",
       " color: #0b98c8 !important;\n",
       " font-size: 110%;\n",
       " font-style: normal;\n",
       " font-weight: normal;\n",
       "}\n",
       ".rendered_html h5,\n",
       ".text_cell_render h5 {\n",
       " color: #0b98c8 !important;\n",
       " font-size: 100%;\n",
       " font-style: normal;\n",
       " font-weight: normal;\n",
       "}\n",
       "hr {\n",
       " margin-top: 8px;\n",
       " margin-bottom: 10px;\n",
       " border: 0;\n",
       " border-top: 1px solid #0b98c8;\n",
       "}\n",
       ".rendered_html hr {\n",
       " color: #0b98c8;\n",
       " background-color: #0b98c8;\n",
       " margin-right: 2em;\n",
       "}\n",
       "#complete > select > option:hover {\n",
       " background: #323b48;\n",
       " background-color: #323b48;\n",
       "}\n",
       "div#_vivaldi-spatnav-focus-indicator._vivaldi-spatnav-focus-indicator {\n",
       " position: absolute;\n",
       " z-index: 9999999999;\n",
       " top: 0px;\n",
       " left: 0px;\n",
       " box-shadow: none;\n",
       " pointer-events: none;\n",
       " border-radius: 2px;\n",
       "}\n",
       ".rendered_html tr,\n",
       ".rendered_html th,\n",
       ".rendered_html td {\n",
       " text-align: left;\n",
       " vertical-align: middle;\n",
       " padding: 0.42em 0.47em;\n",
       " line-height: normal;\n",
       " white-space: normal;\n",
       " max-width: none;\n",
       " border: none;\n",
       "}\n",
       ".rendered_html td {\n",
       " font-family: sans-serif !important;\n",
       " font-size: 9.3pt;\n",
       "}\n",
       ".rendered_html table {\n",
       " font-family: sans-serif !important;\n",
       " margin-left: 8px;\n",
       " margin-right: auto;\n",
       " border: none;\n",
       " border-collapse: collapse;\n",
       " border-spacing: 0;\n",
       " color: #e2e5f2;\n",
       " table-layout: fixed;\n",
       "}\n",
       ".rendered_html thead {\n",
       " font-family: sans-serif !important;\n",
       " font-size: 10.3pt !important;\n",
       " background: #27313d;\n",
       " color: #bbc2e1;\n",
       " border-bottom: 1px solid #27313d;\n",
       " vertical-align: bottom;\n",
       "}\n",
       ".rendered_html tbody tr:nth-child(odd) {\n",
       " background: #3f495a;\n",
       "}\n",
       ".rendered_html tbody tr {\n",
       " background: #394251;\n",
       "}\n",
       ".rendered_html tbody tr:hover:nth-child(odd) {\n",
       " background: #3d4757;\n",
       "}\n",
       ".rendered_html tbody tr:hover {\n",
       " background: #373f4e;\n",
       "}\n",
       ".rendered_html * + table {\n",
       " margin-top: .05em;\n",
       "}\n",
       "div.widget-area {\n",
       " background-color: #323a48;\n",
       " background: #323a48;\n",
       " color: #b4bcde;\n",
       "}\n",
       "div.widget-area a {\n",
       " font-family: sans-serif;\n",
       " font-size: 12.0pt;\n",
       " font-weight: normal;\n",
       " font-style: normal;\n",
       " color: #a2b0c7;\n",
       " text-shadow: none !important;\n",
       "}\n",
       "div.widget-area a:hover,\n",
       "div.widget-area a:focus {\n",
       " font-family: sans-serif;\n",
       " font-size: 12.0pt;\n",
       " font-weight: normal;\n",
       " font-style: normal;\n",
       " color: #dbe1ea;\n",
       " background: rgba(75,95,118,.30);\n",
       " background-color: rgba(75,95,118,.30);\n",
       " border-color: transparent;\n",
       " background-image: none;\n",
       " text-shadow: none !important;\n",
       "}\n",
       "div.widget_item.btn-group > button.btn.btn-default.widget-combo-btn,\n",
       "div.widget_item.btn-group > button.btn.btn-default.widget-combo-btn:hover {\n",
       " background: #232932;\n",
       " background-color: #232932;\n",
       " border: 2px solid #232932 !important;\n",
       " font-size: inherit;\n",
       " z-index: 0;\n",
       "}\n",
       "div.jupyter-widgets.widget-hprogress.widget-hbox {\n",
       " display: inline-table !important;\n",
       " width: 38% !important;\n",
       " margin-left: 10px;\n",
       "}\n",
       "div.jupyter-widgets.widget-hprogress.widget-hbox .widget-label,\n",
       "div.widget-hbox .widget-label,\n",
       ".widget-hbox .widget-label,\n",
       ".widget-inline-hbox .widget-label,\n",
       "div.widget-label {\n",
       " text-align: -webkit-auto !important;\n",
       " margin-left: 15px !important;\n",
       " max-width: 240px !important;\n",
       " min-width: 100px !important;\n",
       " vertical-align: text-top !important;\n",
       " color: #b4bcde !important;\n",
       " font-size: 14px !important;\n",
       "}\n",
       ".widget-hprogress .progress {\n",
       " flex-grow: 1;\n",
       " height: 20px;\n",
       " margin-top: auto;\n",
       " margin-left: 12px;\n",
       " margin-bottom: auto;\n",
       " width: 300px;\n",
       "}\n",
       ".progress {\n",
       " overflow: hidden;\n",
       " height: 22px;\n",
       " margin-bottom: 10px;\n",
       " padding-left: 10px;\n",
       " background-color: #4a5569 !important;\n",
       " border-radius: 2px;\n",
       " -webkit-box-shadow: none;\n",
       " box-shadow: none;\n",
       " z-index: 10;\n",
       "}\n",
       ".progress-bar-danger {\n",
       " background-color: #e74c3c !important;\n",
       "}\n",
       ".progress-bar-info {\n",
       " background-color: #3498db !important;\n",
       "}\n",
       ".progress-bar-warning {\n",
       " background-color: #ff914d !important;\n",
       "}\n",
       ".progress-bar-success {\n",
       " background-color: #83a83b !important;\n",
       "}\n",
       ".widget-select select {\n",
       " margin-left: 12px;\n",
       "}\n",
       ".rendered_html :link {\n",
       " font-family: sans-serif;\n",
       " font-size: 100%;\n",
       " color: #0b98c8;\n",
       " text-decoration: underline;\n",
       "}\n",
       ".rendered_html :visited,\n",
       ".rendered_html :visited:active,\n",
       ".rendered_html :visited:focus {\n",
       " color: #12a3d6;\n",
       "}\n",
       ".rendered_html :visited:hover,\n",
       ".rendered_html :link:hover {\n",
       " font-family: sans-serif;\n",
       " font-size: 100%;\n",
       " color: #0080aa;\n",
       "}\n",
       "div.cell.text_cell a.anchor-link:link {\n",
       " font-size: inherit;\n",
       " text-decoration: none;\n",
       " padding: 0px 20px;\n",
       " visibility: none;\n",
       " color: rgba(0,0,0,.32);\n",
       "}\n",
       "div.cell.text_cell a.anchor-link:link:hover {\n",
       " font-size: inherit;\n",
       " color: #0dc1ff;\n",
       "}\n",
       ".navbar-text {\n",
       " margin-top: 4px;\n",
       " margin-bottom: 0px;\n",
       "}\n",
       "#clusters > a {\n",
       " color: #51c0ef;\n",
       " text-decoration: underline;\n",
       " cursor: auto;\n",
       "}\n",
       "#clusters > a:hover {\n",
       " color: #4c8be2;\n",
       " text-decoration: underline;\n",
       " cursor: auto;\n",
       "}\n",
       "#nbextensions-configurator-container > div.row.container-fluid.nbext-selector > h3 {\n",
       " font-size: 17px;\n",
       " margin-top: 5px;\n",
       " margin-bottom: 8px;\n",
       " height: 24px;\n",
       " padding: 4px 0 4px 0;\n",
       "}\n",
       "div#nbextensions-configurator-container.container,\n",
       "#nbextensions-configurator-container.container {\n",
       " width: 100%;\n",
       " margin-right: auto;\n",
       " margin-left: auto;\n",
       "}\n",
       "div.nbext-selector > nav > .nav > li > a {\n",
       " font-family: sans-serif;\n",
       " font-size: 10.5pt;\n",
       " padding: 2px 5px;\n",
       "}\n",
       "div.nbext-selector > nav > .nav > li > a:hover {\n",
       " background: transparent;\n",
       "}\n",
       "div.nbext-selector > nav > .nav > li:hover {\n",
       " background-color: rgba(75,95,118,.30) !important;\n",
       " background: rgba(75,95,118,.30) !important;\n",
       "}\n",
       "div.nbext-selector > nav > .nav > li.active:hover {\n",
       " background: transparent !important;\n",
       " background-color: transparent !important;\n",
       "}\n",
       ".nav-pills > li.active > a,\n",
       ".nav-pills > li.active > a:active,\n",
       ".nav-pills > li.active > a:hover,\n",
       ".nav-pills > li.active > a:focus {\n",
       " color: #fefefe;\n",
       " background-color: rgba(75,95,118,.30) !important;\n",
       " background: rgba(75,95,118,.30) !important;\n",
       " -webkit-backface-visibility: hidden;\n",
       " -webkit-font-smoothing: subpixel-antialiased !important;\n",
       "}\n",
       "div.nbext-readme > .nbext-readme-contents > .rendered_html {\n",
       " font-family: sans-serif;\n",
       " font-size: 11.5pt;\n",
       " line-height: 145%;\n",
       " padding: 1em 1em;\n",
       " color: #b0bdd7;\n",
       " background-color: #293340;\n",
       " -webkit-box-shadow: none;\n",
       " -moz-box-shadow: none;\n",
       " box-shadow: none;\n",
       "}\n",
       ".nbext-icon,\n",
       ".nbext-desc,\n",
       ".nbext-compat-div,\n",
       ".nbext-enable-btns,\n",
       ".nbext-params {\n",
       " margin-bottom: 8px;\n",
       " font-size: 11.5pt;\n",
       "}\n",
       "div.nbext-readme > .nbext-readme-contents {\n",
       " padding: 0;\n",
       " overflow-y: hidden;\n",
       "}\n",
       "div.nbext-readme > .nbext-readme-contents:not(:empty) {\n",
       " margin-top: 0.5em;\n",
       " margin-bottom: 2em;\n",
       " border: none;\n",
       " border-top-color: rgba(0,156,209,.3);\n",
       "}\n",
       ".nbext-showhide-incompat {\n",
       " padding-bottom: 0.5em;\n",
       " color: #92a2bd;\n",
       " font-size: 10.5pt;\n",
       "}\n",
       ".nbext-filter-menu.dropdown-menu > li > a:hover,\n",
       ".nbext-filter-menu.dropdown-menu > li > a:focus,\n",
       ".nbext-filter-menu.dropdown-menu > li > a.ui-state-focus {\n",
       " color: #dbe1ea !important;\n",
       " background-color: #323b48 !important;\n",
       " background: #323b48 !important;\n",
       " border-color: #323b48 !important;\n",
       "}\n",
       ".nbext-filter-input-wrap > .nbext-filter-input-subwrap,\n",
       ".nbext-filter-input-wrap > .nbext-filter-input-subwrap > input {\n",
       " border: none;\n",
       " outline: none;\n",
       " background-color: transparent;\n",
       " padding: 0;\n",
       " vertical-align: middle;\n",
       " margin-top: -2px;\n",
       "}\n",
       "span.rendered_html code {\n",
       " background-color: transparent;\n",
       " color: #a2b0c7;\n",
       "}\n",
       "#nbextensions-configurator-container > div.row.container-fluid.nbext-selector {\n",
       " padding-left: 0px;\n",
       " padding-right: 0px;\n",
       "}\n",
       ".nbext-filter-menu {\n",
       " max-height: 55vh !important;\n",
       " overflow-y: auto;\n",
       " outline: none;\n",
       " border: none;\n",
       "}\n",
       ".nbext-filter-menu:hover {\n",
       " border: none;\n",
       "}\n",
       ".alert-warning {\n",
       " background-color: #2d3846;\n",
       " border-color: #2d3846;\n",
       " color: #a2b0c7;\n",
       "}\n",
       ".notification_widget.danger {\n",
       " color: #ffffff;\n",
       " background-color: #e74c3c;\n",
       " border-color: #e74c3c;\n",
       " padding-right: 5px;\n",
       "}\n",
       "#nbextensions-configurator-container > div.nbext-buttons.tree-buttons.no-padding.pull-right > span > button {\n",
       " border: none !important;\n",
       "}\n",
       "button#refresh_running_list {\n",
       " border: none !important;\n",
       "}\n",
       "mark,\n",
       ".mark {\n",
       " background-color: #293340;\n",
       " color: #b0bdd7;\n",
       " padding: .15em;\n",
       "}\n",
       "a.text-warning,\n",
       "a.text-warning:hover {\n",
       " color: #546386;\n",
       "}\n",
       "a.text-warning.bg-warning {\n",
       " background-color: #1a2028;\n",
       "}\n",
       "span.bg-success.text-success {\n",
       " background-color: transparent;\n",
       " color: #48a667;\n",
       "}\n",
       "span.bg-danger.text-danger {\n",
       " background-color: #1a2028;\n",
       " color: #dc6972;\n",
       "}\n",
       ".has-success .input-group-addon {\n",
       " color: #48a667;\n",
       " border-color: transparent;\n",
       " background: inherit;\n",
       " background-color: rgba(83,180,115,.10);\n",
       "}\n",
       ".has-success .form-control {\n",
       " border-color: #48a667;\n",
       " -webkit-box-shadow: inset 0 1px 1px rgba(0,0,0,0.025);\n",
       " box-shadow: inset 0 1px 1px rgba(0,0,0,0.025);\n",
       "}\n",
       ".has-error .input-group-addon {\n",
       " color: #dc6972;\n",
       " border-color: transparent;\n",
       " background: inherit;\n",
       " background-color: rgba(192,57,67,.10);\n",
       "}\n",
       ".has-error .form-control {\n",
       " border-color: #dc6972;\n",
       " -webkit-box-shadow: inset 0 1px 1px rgba(0,0,0,0.025);\n",
       " box-shadow: inset 0 1px 1px rgba(0,0,0,0.025);\n",
       "}\n",
       ".kse-input-group-pretty > kbd {\n",
       " font-family: monospace, monospace;\n",
       " color: #a2b0c7;\n",
       " font-weight: normal;\n",
       " background: transparent;\n",
       "}\n",
       ".kse-input-group-pretty > kbd {\n",
       " font-family: monospace, monospace;\n",
       " color: #a2b0c7;\n",
       " font-weight: normal;\n",
       " background: transparent;\n",
       "}\n",
       "div.nbext-enable-btns .btn[disabled],\n",
       "div.nbext-enable-btns .btn[disabled]:hover,\n",
       ".btn-default.disabled,\n",
       ".btn-default[disabled] {\n",
       " background: #232c37;\n",
       " background-color: #232c37;\n",
       " color: #98a8c1;\n",
       "}\n",
       "label#Keyword-Filter {\n",
       " display: none;\n",
       "}\n",
       ".input-group .nbext-list-btn-add,\n",
       ".input-group-btn:last-child > .btn-group > .btn {\n",
       " background: #252b35;\n",
       " background-color: #252b35;\n",
       " border-color: #252b35;\n",
       " border: 2px solid #252b35;\n",
       "}\n",
       ".input-group .nbext-list-btn-add:hover,\n",
       ".input-group-btn:last-child > .btn-group > .btn:hover {\n",
       " background: #21262f;\n",
       " background-color: #21262f;\n",
       " border-color: #21262f;\n",
       " border: 2px solid #21262f;\n",
       "}\n",
       "#notebook-container > div.cell.code_cell.rendered.selected > div.widget-area > div.widget-subarea > div > div.widget_item.btn-group > button.btn.btn-default.dropdown-toggle.widget-combo-carrot-btn {\n",
       " background: #252b35;\n",
       " background-color: #252b35;\n",
       " border-color: #252b35;\n",
       "}\n",
       "#notebook-container > div.cell.code_cell.rendered.selected > div.widget-area > div.widget-subarea > div > div.widget_item.btn-group > button.btn.btn-default.dropdown-toggle.widget-combo-carrot-btn:hover {\n",
       " background: #21262f;\n",
       " background-color: #21262f;\n",
       " border-color: #21262f;\n",
       "}\n",
       ".ui-widget-content {\n",
       " background: #252e3a;\n",
       " background-color: #252e3a;\n",
       " border: 2px solid #252e3a;\n",
       " color: #a2b0c7;\n",
       "}\n",
       "div.collapsible_headings_toggle {\n",
       " color: rgba(75,95,118,.55) !important;\n",
       "}\n",
       "div.collapsible_headings_toggle:hover {\n",
       " color: #0b98c8 !important;\n",
       "}\n",
       ".collapsible_headings_toggle .h1,\n",
       ".collapsible_headings_toggle .h2,\n",
       ".collapsible_headings_toggle .h3,\n",
       ".collapsible_headings_toggle .h4,\n",
       ".collapsible_headings_toggle .h5,\n",
       ".collapsible_headings_toggle .h6 {\n",
       " margin: 0.3em .4em 0em 0em !important;\n",
       " line-height: 1.2 !important;\n",
       "}\n",
       "div.collapsible_headings_toggle .fa-caret-down:before,\n",
       "div.collapsible_headings_toggle .fa-caret-right:before {\n",
       " font-size: xx-large;\n",
       " transition: transform 1000ms;\n",
       " transform: none !important;\n",
       "}\n",
       ".collapsible_headings_collapsed.collapsible_headings_ellipsis .rendered_html h1:after,\n",
       ".collapsible_headings_collapsed.collapsible_headings_ellipsis .rendered_html h2:after,\n",
       ".collapsible_headings_collapsed.collapsible_headings_ellipsis .rendered_html h3:after,\n",
       ".collapsible_headings_collapsed.collapsible_headings_ellipsis .rendered_html h4:after,\n",
       ".collapsible_headings_collapsed.collapsible_headings_ellipsis .rendered_html h5:after,\n",
       ".collapsible_headings_collapsed.collapsible_headings_ellipsis .rendered_html h6:after {\n",
       " position: absolute;\n",
       " right: 0;\n",
       " bottom: 20% !important;\n",
       " content: \"[\\002026]\";\n",
       " color: rgba(75,95,118,.55) !important;\n",
       " padding: 0.5em 0em 0em 0em !important;\n",
       "}\n",
       ".collapsible_headings_ellipsis .rendered_html h1,\n",
       ".collapsible_headings_ellipsis .rendered_html h2,\n",
       ".collapsible_headings_ellipsis .rendered_html h3,\n",
       ".collapsible_headings_ellipsis .rendered_html h4,\n",
       ".collapsible_headings_ellipsis .rendered_html h5,\n",
       ".collapsible_headings_ellipsis .rendered_html h6,\n",
       ".collapsible_headings_toggle .fa {\n",
       " transition: transform 1000ms !important;\n",
       " -webkit-transform: inherit !important;\n",
       " -moz-transform: inherit !important;\n",
       " -ms-transform: inherit !important;\n",
       " -o-transform: inherit !important;\n",
       " transform: inherit !important;\n",
       " padding-right: 0px !important;\n",
       "}\n",
       "#toc-wrapper {\n",
       " z-index: 90;\n",
       " position: fixed !important;\n",
       " display: flex;\n",
       " flex-direction: column;\n",
       " overflow: hidden;\n",
       " padding: 10px;\n",
       " border-style: solid;\n",
       " border-width: thin;\n",
       " border-right-width: medium !important;\n",
       " background-color: #1a2028 !important;\n",
       "}\n",
       "#toc-wrapper.ui-draggable.ui-resizable.sidebar-wrapper {\n",
       " border-color: rgba(75,95,118,.30) !important;\n",
       "}\n",
       "#toc a,\n",
       "#navigate_menu a,\n",
       ".toc {\n",
       " color: #a2b0c7 !important;\n",
       " font-size: 11pt !important;\n",
       "}\n",
       "#toc li > span:hover {\n",
       " background-color: #323b48 !important;\n",
       "}\n",
       "#toc a:hover,\n",
       "#navigate_menu a:hover,\n",
       ".toc {\n",
       " color: #fefefe !important;\n",
       " font-size: 11pt !important;\n",
       "}\n",
       "#toc-wrapper .toc-item-num {\n",
       " color: #0b98c8 !important;\n",
       " font-size: 11pt !important;\n",
       "}\n",
       "input.raw_input {\n",
       " font-family: monospace, monospace;\n",
       " font-size: 11pt !important;\n",
       " color: #cdd2e9;\n",
       " background-color: #252e3a;\n",
       " border-color: #232c37;\n",
       " background: #232c37;\n",
       " width: auto;\n",
       " vertical-align: baseline;\n",
       " padding: 0em 0.25em;\n",
       " margin: 0em 0.25em;\n",
       " -webkit-box-shadow: none;\n",
       " box-shadow: none;\n",
       "}\n",
       "audio,\n",
       "video {\n",
       " display: inline;\n",
       " vertical-align: middle;\n",
       " align-content: center;\n",
       " margin-left: 20%;\n",
       "}\n",
       ".cmd-palette .modal-body {\n",
       " padding: 0px;\n",
       " margin: 0px;\n",
       "}\n",
       ".cmd-palette form {\n",
       " background: #293547;\n",
       " background-color: #293547;\n",
       "}\n",
       ".typeahead-field input:last-child,\n",
       ".typeahead-hint {\n",
       " background: #293547;\n",
       " background-color: #293547;\n",
       " z-index: 1;\n",
       "}\n",
       ".typeahead-field input {\n",
       " font-family: sans-serif;\n",
       " color: #cdd2e9;\n",
       " border: none;\n",
       " font-size: 28pt;\n",
       " display: inline-block;\n",
       " line-height: inherit;\n",
       " padding: 3px 10px;\n",
       " height: 70px;\n",
       "}\n",
       ".typeahead-select {\n",
       " background-color: #293547;\n",
       "}\n",
       "body > div.modal.cmd-palette.typeahead-field {\n",
       " display: table;\n",
       " border-collapse: separate;\n",
       " background-color: #2b3850;\n",
       "}\n",
       ".typeahead-container button {\n",
       " font-family: sans-serif;\n",
       " font-size: 28pt;\n",
       " background-color: #252b35;\n",
       " border: none;\n",
       " display: inline-block;\n",
       " line-height: inherit;\n",
       " padding: 3px 10px;\n",
       " height: 70px;\n",
       "}\n",
       ".typeahead-search-icon {\n",
       " min-width: 40px;\n",
       " min-height: 55px;\n",
       " display: block;\n",
       " vertical-align: middle;\n",
       " text-align: center;\n",
       "}\n",
       ".typeahead-container button:focus,\n",
       ".typeahead-container button:hover {\n",
       " color: #dbe1ea;\n",
       " background-color: #21262f;\n",
       " border-color: #293340;\n",
       "}\n",
       ".typeahead-list > li.typeahead-group.active > a,\n",
       ".typeahead-list > li.typeahead-group > a,\n",
       ".typeahead-list > li.typeahead-group > a:focus,\n",
       ".typeahead-list > li.typeahead-group > a:hover {\n",
       " display: none;\n",
       "}\n",
       ".typeahead-dropdown > li > a,\n",
       ".typeahead-list > li > a {\n",
       " color: #a2b0c7;\n",
       " text-decoration: none;\n",
       "}\n",
       ".typeahead-dropdown,\n",
       ".typeahead-list {\n",
       " font-family: sans-serif;\n",
       " font-size: 13pt;\n",
       " color: #a2b0c7;\n",
       " background-color: #202937;\n",
       " border: none;\n",
       " background-clip: padding-box;\n",
       " margin-top: 0px;\n",
       " padding: 3px 2px 3px 0px;\n",
       " line-height: 1.7;\n",
       "}\n",
       ".typeahead-dropdown > li.active > a,\n",
       ".typeahead-dropdown > li > a:focus,\n",
       ".typeahead-dropdown > li > a:hover,\n",
       ".typeahead-list > li.active > a,\n",
       ".typeahead-list > li > a:focus,\n",
       ".typeahead-list > li > a:hover {\n",
       " color: #dbe1ea;\n",
       " background-color: #2b3850;\n",
       " border-color: #2b3850;\n",
       "}\n",
       ".command-shortcut:before {\n",
       " content: \"(command)\";\n",
       " padding-right: 3px;\n",
       " color: #546386;\n",
       "}\n",
       ".edit-shortcut:before {\n",
       " content: \"(edit)\";\n",
       " padding-right: 3px;\n",
       " color: #546386;\n",
       "}\n",
       "ul.typeahead-list i {\n",
       " margin-left: 1px;\n",
       " width: 18px;\n",
       " margin-right: 10px;\n",
       "}\n",
       "ul.typeahead-list {\n",
       " max-height: 50vh;\n",
       " overflow: auto;\n",
       "}\n",
       ".typeahead-list > li {\n",
       " position: relative;\n",
       " border: none;\n",
       "}\n",
       "div.input.typeahead-hint,\n",
       "input.typeahead-hint,\n",
       "body > div.modal.cmd-palette.in > div > div > div > form > div > div.typeahead-field > span.typeahead-query > input.typeahead-hint {\n",
       " color: #546386 !important;\n",
       " background-color: transparent;\n",
       " padding: 3px 10px;\n",
       "}\n",
       ".typeahead-dropdown > li > a,\n",
       ".typeahead-list > li > a {\n",
       " display: block;\n",
       " padding: 5px;\n",
       " clear: both;\n",
       " font-weight: 400;\n",
       " line-height: 1.7;\n",
       " border: 1px solid #202937;\n",
       " border-bottom-color: rgba(75,95,118,.55);\n",
       "}\n",
       "body > div.modal.cmd-palette.in > div {\n",
       " min-width: 750px;\n",
       " margin: 150px auto;\n",
       "}\n",
       ".typeahead-container strong {\n",
       " font-weight: bolder;\n",
       " color: #0b98c8;\n",
       "}\n",
       "#find-and-replace #replace-preview .match,\n",
       "#find-and-replace #replace-preview .insert {\n",
       " color: #ffffff;\n",
       " background-color: #008ebf;\n",
       " border-color: #008ebf;\n",
       " border-style: solid;\n",
       " border-width: 1px;\n",
       " border-radius: 0px;\n",
       "}\n",
       "#find-and-replace #replace-preview .replace .match {\n",
       " background-color: #dc6972;\n",
       " border-color: #dc6972;\n",
       " border-radius: 0px;\n",
       "}\n",
       "#find-and-replace #replace-preview .replace .insert {\n",
       " background-color: #48a667;\n",
       " border-color: #48a667;\n",
       " border-radius: 0px;\n",
       "}\n",
       ".jupyter-dashboard-menu-item.selected::before {\n",
       " font-family: 'FontAwesome' !important;\n",
       " content: '\\f00c' !important;\n",
       " position: absolute !important;\n",
       " color: #0b98c8 !important;\n",
       " left: 0px !important;\n",
       " top: 13px !important;\n",
       " font-size: 12px !important;\n",
       "}\n",
       ".shortcut_key,\n",
       "span.shortcut_key {\n",
       " display: inline-block;\n",
       " width: 16ex;\n",
       " text-align: right;\n",
       " font-family: monospace;\n",
       "}\n",
       ".jupyter-keybindings {\n",
       " padding: 1px;\n",
       " line-height: 24px;\n",
       " border-bottom: 1px solid rgba(75,95,118,.30);\n",
       "}\n",
       ".jupyter-keybindings i {\n",
       " background: #252e3a;\n",
       " font-size: small;\n",
       " padding: 5px;\n",
       " margin-left: 7px;\n",
       "}\n",
       "div#short-key-bindings-intro.well,\n",
       ".well {\n",
       " background-color: #252b35;\n",
       " border: 1px solid #252b35;\n",
       " color: #a2b0c7;\n",
       " border-radius: 2px;\n",
       " -webkit-box-shadow: none;\n",
       " box-shadow: none;\n",
       "}\n",
       "#texteditor-backdrop {\n",
       " background: #1a2028;\n",
       " background-color: #1a2028;\n",
       "}\n",
       "#texteditor-backdrop #texteditor-container .CodeMirror-gutter,\n",
       "#texteditor-backdrop #texteditor-container .CodeMirror-gutters {\n",
       " background: #334050;\n",
       " background-color: #334050;\n",
       " color: #546386;\n",
       "}\n",
       ".edit_app #menubar .navbar {\n",
       " margin-bottom: 0px;\n",
       "}\n",
       "#texteditor-backdrop #texteditor-container {\n",
       " padding: 0px;\n",
       " background-color: #293340;\n",
       " box-shadow: none;\n",
       "}\n",
       ".terminal-app {\n",
       " background: #1a2028;\n",
       "}\n",
       ".terminal-app > #header {\n",
       " background: #1a2028;\n",
       "}\n",
       ".terminal-app .terminal {\n",
       " font-family: monospace, monospace;\n",
       " font-size: 11pt;\n",
       " line-height: 170%;\n",
       " color: #cdd2e9;\n",
       " background: #293340;\n",
       " padding: 0.4em;\n",
       " border-radius: 2px;\n",
       " -webkit-box-shadow: none;\n",
       " box-shadow: none;\n",
       "}\n",
       ".terminal .xterm-viewport {\n",
       " background-color: #293340;\n",
       " color: #cdd2e9;\n",
       " overflow-y: auto;\n",
       "}\n",
       ".terminal .xterm-color-0 {\n",
       " color: #0b98c8;\n",
       "}\n",
       ".terminal .xterm-color-1 {\n",
       " color: #e17e85;\n",
       "}\n",
       ".terminal .xterm-color-2 {\n",
       " color: #4cb2ff;\n",
       "}\n",
       ".terminal .xterm-color-3 {\n",
       " color: #e17e85;\n",
       "}\n",
       ".terminal .xterm-color-4 {\n",
       " color: #51c0ef;\n",
       "}\n",
       ".terminal .xterm-color-5 {\n",
       " color: #61ba86;\n",
       "}\n",
       ".terminal .xterm-color-6 {\n",
       " color: #be86e3;\n",
       "}\n",
       ".terminal .xterm-color-7 {\n",
       " color: #ffec8e;\n",
       "}\n",
       ".terminal .xterm-color-8 {\n",
       " color: #51c0ef;\n",
       "}\n",
       ".terminal .xterm-color-9 {\n",
       " color: #61ba86;\n",
       "}\n",
       ".terminal .xterm-color-10 {\n",
       " color: #e17e85;\n",
       "}\n",
       ".terminal .xterm-color-14 {\n",
       " color: #be86e3;\n",
       "}\n",
       ".terminal .xterm-bg-color-15 {\n",
       " background-color: #293340;\n",
       "}\n",
       ".terminal:not(.xterm-cursor-style-underline):not(.xterm-cursor-style-bar) .terminal-cursor {\n",
       " background-color: #0b98c8;\n",
       " color: #293340;\n",
       "}\n",
       ".terminal:not(.focus) .terminal-cursor {\n",
       " outline: 1px solid #0b98c8;\n",
       " outline-offset: -1px;\n",
       "}\n",
       ".celltoolbar {\n",
       " font-size: 100%;\n",
       " padding-top: 3px;\n",
       " border-color: transparent;\n",
       " border-bottom: thin solid rgba(0,156,209,.3);\n",
       " background: transparent;\n",
       "}\n",
       ".cell-tag,\n",
       ".tags-input input,\n",
       ".tags-input button {\n",
       " color: #a2b0c7;\n",
       " background-color: #1a2028;\n",
       " background-image: none;\n",
       " border: 1px solid #a2b0c7;\n",
       " border-radius: 1px;\n",
       " box-shadow: none;\n",
       " width: inherit;\n",
       " font-size: inherit;\n",
       " height: 22px;\n",
       " line-height: 22px;\n",
       "}\n",
       "#notebook-container > div.cell.code_cell.rendered.selected > div.input > div.inner_cell > div.ctb_hideshow.ctb_show > div > div > button,\n",
       "#notebook-container > div.input > div.inner_cell > div.ctb_hideshow.ctb_show > div > div > button {\n",
       " font-size: 10pt;\n",
       " color: #a2b0c7;\n",
       " background-color: #1a2028;\n",
       " background-image: none;\n",
       " border: 1px solid #a2b0c7;\n",
       " border-radius: 1px;\n",
       " box-shadow: none;\n",
       " width: inherit;\n",
       " font-size: inherit;\n",
       " height: 22px;\n",
       " line-height: 22px;\n",
       "}\n",
       "div#pager #pager-contents {\n",
       " background: #1a2028 !important;\n",
       " background-color: #1a2028 !important;\n",
       "}\n",
       "div#pager pre {\n",
       " color: #cdd2e9 !important;\n",
       " background: #293340 !important;\n",
       " background-color: #293340 !important;\n",
       " padding: 0.4em;\n",
       "}\n",
       "div#pager .ui-resizable-handle {\n",
       " top: 0px;\n",
       " height: 8px;\n",
       " background: #0b98c8 !important;\n",
       " border-top: 1px solid #0b98c8;\n",
       " border-bottom: 1px solid #0b98c8;\n",
       "}\n",
       "div.CodeMirror,\n",
       "div.CodeMirror pre {\n",
       " font-family: monospace, monospace;\n",
       " font-size: 11pt;\n",
       " line-height: 170%;\n",
       " color: #cdd2e9;\n",
       "}\n",
       "div.CodeMirror-lines {\n",
       " padding-bottom: .9em;\n",
       " padding-left: .5em;\n",
       " padding-right: 1.5em;\n",
       " padding-top: .7em;\n",
       "}\n",
       "span.ansiblack,\n",
       ".ansi-black-fg {\n",
       " color: #2b303b;\n",
       "}\n",
       "span.ansiblue,\n",
       ".ansi-blue-fg,\n",
       ".ansi-blue-intense-fg {\n",
       " color: #61afef;\n",
       "}\n",
       "span.ansigray,\n",
       ".ansi-gray-fg,\n",
       ".ansi-gray-intense-fg {\n",
       " color: #899ab8;\n",
       "}\n",
       "span.ansigreen,\n",
       ".ansi-green-fg {\n",
       " color: #8fca9a;\n",
       "}\n",
       ".ansi-green-intense-fg {\n",
       " color: #899ab8;\n",
       "}\n",
       "span.ansipurple,\n",
       ".ansi-purple-fg,\n",
       ".ansi-purple-intense-fg {\n",
       " color: #b399ef;\n",
       "}\n",
       "span.ansicyan,\n",
       ".ansi-cyan-fg,\n",
       ".ansi-cyan-intense-fg {\n",
       " color: #b399ef;\n",
       "}\n",
       "span.ansiyellow,\n",
       ".ansi-yellow-fg,\n",
       ".ansi-yellow-intense-fg {\n",
       " color: #ffec8e;\n",
       "}\n",
       "span.ansired,\n",
       ".ansi-red-fg,\n",
       ".ansi-red-intense-fg {\n",
       " color: #e07a7a;\n",
       "}\n",
       "div.output-stderr {\n",
       " background-color: #e07a7a;\n",
       "}\n",
       "div.output-stderr pre {\n",
       " color: #d0d4e6;\n",
       "}\n",
       "div.js-error {\n",
       " color: #e07a7a;\n",
       "}\n",
       ".ipython_tooltip {\n",
       " font-family: monospace, monospace;\n",
       " font-size: 11pt;\n",
       " line-height: 170%;\n",
       " border: 2px solid #252c36;\n",
       " background: #363f4e;\n",
       " background-color: #363f4e;\n",
       " border-radius: 2px;\n",
       " overflow-x: visible;\n",
       " overflow-y: visible;\n",
       " box-shadow: none;\n",
       " position: absolute;\n",
       " z-index: 1000;\n",
       "}\n",
       ".ipython_tooltip .tooltiptext pre {\n",
       " font-family: monospace, monospace;\n",
       " font-size: 11pt;\n",
       " line-height: 170%;\n",
       " background: #363f4e;\n",
       " background-color: #363f4e;\n",
       " color: #cdd2e9;\n",
       " overflow-x: visible;\n",
       " overflow-y: visible;\n",
       " max-width: 900px;\n",
       "}\n",
       "div#tooltip.ipython_tooltip {\n",
       " overflow-x: wrap;\n",
       " overflow-y: visible;\n",
       " max-width: 800px;\n",
       "}\n",
       "div.tooltiptext.bigtooltip {\n",
       " overflow-x: visible;\n",
       " overflow-y: scroll;\n",
       " height: 400px;\n",
       " max-width: 800px;\n",
       "}\n",
       ".cm-s-ipython.CodeMirror {\n",
       " font-family: monospace, monospace;\n",
       " font-size: 11pt;\n",
       " background: #293340;\n",
       " color: #cdd2e9;\n",
       " border-radius: 2px;\n",
       " font-style: normal;\n",
       " font-weight: normal;\n",
       "}\n",
       ".cm-s-ipython div.CodeMirror-selected {\n",
       " background: #334050;\n",
       "}\n",
       ".CodeMirror-gutters {\n",
       " border: none;\n",
       " border-right: 1px solid #334050 !important;\n",
       " background-color: #334050 !important;\n",
       " background: #334050 !important;\n",
       " border-radius: 0px;\n",
       " white-space: nowrap;\n",
       "}\n",
       ".cm-s-ipython .CodeMirror-gutters {\n",
       " background: #334050;\n",
       " border: none;\n",
       " border-radius: 0px;\n",
       " width: 36px;\n",
       "}\n",
       ".cm-s-ipython .CodeMirror-linenumber {\n",
       " color: #546386;\n",
       "}\n",
       ".CodeMirror-sizer {\n",
       " margin-left: 40px;\n",
       "}\n",
       ".CodeMirror-linenumber,\n",
       "div.CodeMirror-linenumber,\n",
       ".CodeMirror-gutter.CodeMirror-linenumberdiv.CodeMirror-gutter.CodeMirror-linenumber {\n",
       " padding-right: 1px;\n",
       " margin-left: 0px;\n",
       " margin: 0px;\n",
       " width: 26px !important;\n",
       " padding: 0px;\n",
       " text-align: right;\n",
       "}\n",
       ".CodeMirror-linenumber {\n",
       " color: #546386;\n",
       "}\n",
       ".cm-s-ipython .CodeMirror-cursor {\n",
       " border-left: 2px solid #0095ff !important;\n",
       "}\n",
       ".cm-s-ipython span.cm-comment {\n",
       " color: #667fb1;\n",
       " font-style: italic;\n",
       "}\n",
       ".cm-s-ipython span.cm-atom {\n",
       " color: #be86e3;\n",
       "}\n",
       ".cm-s-ipython span.cm-number {\n",
       " color: #51c0ef;\n",
       "}\n",
       ".cm-s-ipython span.cm-property {\n",
       " color: #cdd2e9;\n",
       "}\n",
       ".cm-s-ipython span.cm-attribute {\n",
       " color: #cdd2e9;\n",
       "}\n",
       ".cm-s-ipython span.cm-keyword {\n",
       " color: #4cb2ff;\n",
       " font-weight: normal;\n",
       "}\n",
       ".cm-s-ipython span.cm-string {\n",
       " color: #61ba86;\n",
       "}\n",
       ".cm-s-ipython span.cm-meta {\n",
       " color: #ffec8e;\n",
       "}\n",
       ".cm-s-ipython span.cm-operator {\n",
       " color: #00b4ff;\n",
       "}\n",
       ".cm-s-ipython span.cm-builtin {\n",
       " color: #e17e85;\n",
       "}\n",
       ".cm-s-ipython span.cm-variable {\n",
       " color: #cdd2e9;\n",
       "}\n",
       ".cm-s-ipython span.cm-variable-2 {\n",
       " color: #e17e85;\n",
       "}\n",
       ".cm-s-ipython span.cm-variable-3 {\n",
       " color: #ffec8e;\n",
       "}\n",
       ".cm-s-ipython span.cm-def {\n",
       " color: #ffec8e;\n",
       " font-weight: normal;\n",
       "}\n",
       ".cm-s-ipython span.cm-error {\n",
       " background: rgba(191,97,106,.4);\n",
       "}\n",
       ".cm-s-ipython span.cm-tag {\n",
       " color: #be86e3;\n",
       "}\n",
       ".cm-s-ipython span.cm-link {\n",
       " color: #51c0ef;\n",
       "}\n",
       ".cm-s-ipython span.cm-storage {\n",
       " color: #be86e3;\n",
       "}\n",
       ".cm-s-ipython span.cm-entity {\n",
       " color: #be86e3;\n",
       "}\n",
       ".cm-s-ipython span.cm-quote {\n",
       " color: #61ba86;\n",
       "}\n",
       "div.CodeMirror span.CodeMirror-matchingbracket {\n",
       " color: #ffffff;\n",
       " font-weight: bold;\n",
       " background-color: #4c8be2;\n",
       "}\n",
       "div.CodeMirror span.CodeMirror-nonmatchingbracket {\n",
       " color: #ffffff;\n",
       " font-weight: bold;\n",
       " background: rgba(191,97,106,.4) !important;\n",
       "}\n",
       ".cm-header-1 {\n",
       " font-size: 215%;\n",
       "}\n",
       ".cm-header-2 {\n",
       " font-size: 180%;\n",
       "}\n",
       ".cm-header-3 {\n",
       " font-size: 150%;\n",
       "}\n",
       ".cm-header-4 {\n",
       " font-size: 120%;\n",
       "}\n",
       ".cm-header-5 {\n",
       " font-size: 100%;\n",
       "}\n",
       ".cm-s-default .cm-hr {\n",
       " color: #00b4ff;\n",
       "}\n",
       "div.cell.text_cell .cm-s-default .cm-header {\n",
       " font-family: sans-serif;\n",
       " font-weight: normal;\n",
       " color: #0b98c8 !important;\n",
       " margin-top: 0.3em !important;\n",
       " margin-bottom: 0.3em !important;\n",
       "}\n",
       "div.cell.text_cell .cm-s-default span.cm-variable-2 {\n",
       " color: #b0bdd7 !important;\n",
       "}\n",
       "div.cell.text_cell .cm-s-default span.cm-variable-3 {\n",
       " color: #ffec8e !important;\n",
       "}\n",
       ".cm-s-default span.cm-comment {\n",
       " color: #667fb1 !important;\n",
       "}\n",
       ".cm-s-default .cm-tag {\n",
       " color: #8fb36a;\n",
       "}\n",
       ".cm-s-default .cm-builtin {\n",
       " color: #e17e85;\n",
       "}\n",
       ".cm-s-default .cm-string {\n",
       " color: #61ba86;\n",
       "}\n",
       ".cm-s-default .cm-keyword {\n",
       " color: #4cb2ff;\n",
       "}\n",
       ".cm-s-default .cm-number {\n",
       " color: #51c0ef;\n",
       "}\n",
       ".cm-s-default .cm-error {\n",
       " color: #be86e3;\n",
       "}\n",
       ".cm-s-default .cm-link {\n",
       " color: #51c0ef;\n",
       "}\n",
       ".cm-s-default .cm-atom {\n",
       " color: #51c0ef;\n",
       "}\n",
       ".cm-s-default .cm-def {\n",
       " color: #ffec8e;\n",
       "}\n",
       ".CodeMirror-cursor {\n",
       " border-left: 2px solid #0095ff !important;\n",
       " border-right: none;\n",
       " width: 0;\n",
       "}\n",
       ".cm-s-default div.CodeMirror-selected {\n",
       " background: #334050;\n",
       "}\n",
       ".cm-s-default .cm-selected {\n",
       " background: #334050;\n",
       "}\n",
       ".MathJax_Display,\n",
       ".MathJax {\n",
       " border: 0 !important;\n",
       " font-size: 100% !important;\n",
       " text-align: center !important;\n",
       " margin: 0em !important;\n",
       " line-height: 2.25 !important;\n",
       "}\n",
       ".MathJax:focus,\n",
       "body :focus .MathJax {\n",
       " display: inline-block !important;\n",
       "}\n",
       ".MathJax:focus,\n",
       "body :focus .MathJax {\n",
       " display: inline-block !important;\n",
       "}\n",
       ".completions {\n",
       " position: absolute;\n",
       " z-index: 110;\n",
       " overflow: hidden;\n",
       " border: medium solid rgba(0,156,209,.5);\n",
       " box-shadow: none;\n",
       " line-height: 1;\n",
       "}\n",
       ".completions select {\n",
       " background: #293340;\n",
       " background-color: #293340;\n",
       " outline: none;\n",
       " border: none;\n",
       " padding: 0px;\n",
       " margin: 0px;\n",
       " margin-left: 2px;\n",
       " overflow: auto;\n",
       " font-family: monospace, monospace;\n",
       " font-size: 11pt;\n",
       " color: #cdd2e9;\n",
       " width: auto;\n",
       "}\n",
       "div#maintoolbar {\n",
       " display: none !important;\n",
       "}\n",
       "#header-container {\n",
       " display: none !important;\n",
       "}\n",
       "\n",
       "<script>\n",
       "    MathJax.Hub.Config({\n",
       "        \"HTML-CSS\": {\n",
       "            /*preferredFont: \"TeX\",*/\n",
       "            /*availableFonts: [\"TeX\", \"STIX\"],*/\n",
       "            styles: {\n",
       "                scale: 100,\n",
       "                \".MathJax_Display\": {\n",
       "                    \"font-size\": \"100%\",\n",
       "                }\n",
       "            }\n",
       "        }\n",
       "    });\n",
       "</script>\n",
       "     </style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from jupyterthemes import get_themes\n",
    "import jupyterthemes as jt\n",
    "from jupyterthemes.stylefx import set_nb_theme\n",
    "set_nb_theme('chesterish')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np \n",
    "import os\n",
    "import json\n",
    "from pprint import pprint\n",
    "from copy import deepcopy\n",
    "from tqdm.notebook import tqdm "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_data = pd.read_csv('../all_sources_metadata_2020-03-13.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sha</th>\n",
       "      <th>source_x</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>license</th>\n",
       "      <th>abstract</th>\n",
       "      <th>publish_time</th>\n",
       "      <th>authors</th>\n",
       "      <th>journal</th>\n",
       "      <th>Microsoft Academic Paper ID</th>\n",
       "      <th>WHO #Covidence</th>\n",
       "      <th>has_full_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>c630ebcdf30652f0422c3ec12a00b50241dc9bd9</td>\n",
       "      <td>CZI</td>\n",
       "      <td>Angiotensin-converting enzyme 2 (ACE2) as a SARS-CoV-2 receptor: molecular mechanisms and potential therapeutic target</td>\n",
       "      <td>10.1007/s00134-020-05985-9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32125455.0</td>\n",
       "      <td>cc-by-nc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020</td>\n",
       "      <td>Zhang, Haibo; Penninger, Josef M.; Li, Yimin; Zhong, Nanshan; Slutsky, Arthur S.</td>\n",
       "      <td>Intensive Care Med</td>\n",
       "      <td>2.002765e+09</td>\n",
       "      <td>#3252</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>53eccda7977a31e3d0f565c884da036b1e85438e</td>\n",
       "      <td>CZI</td>\n",
       "      <td>Comparative genetic analysis of the novel coronavirus (2019-nCoV/SARS-CoV-2) receptor ACE2 in different populations</td>\n",
       "      <td>10.1038/s41421-020-0147-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cc-by</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020</td>\n",
       "      <td>Cao, Yanan; Li, Lin; Feng, Zhimin; Wan, Shengqing; Huang, Peide; Sun, Xiaohui; Wen, Fang; Huang, Xuanlin; Ning, Guang; Wang, Weiqing</td>\n",
       "      <td>Cell Discovery</td>\n",
       "      <td>3.003431e+09</td>\n",
       "      <td>#1861</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>210a892deb1c61577f6fba58505fd65356ce6636</td>\n",
       "      <td>CZI</td>\n",
       "      <td>Incubation Period and Other Epidemiological Characteristics of 2019 Novel Coronavirus Infections with Right Truncation: A Statistical Analysis of Publicly Available Case Data</td>\n",
       "      <td>10.3390/jcm9020538</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cc-by</td>\n",
       "      <td>The geographic spread of 2019 novel coronavirus (COVID-19) infections from the epicenter of Wuhan, China, has provided an opportunity to study the natural history of the recently emerged virus. Us...</td>\n",
       "      <td>2020</td>\n",
       "      <td>Linton, M. Natalie; Kobayashi, Tetsuro; Yang, Yichi; Hayashi, Katsuma; Akhmetzhanov, R. Andrei; Jung, Sung-mok; Yuan, Baoyin; Kinoshita, Ryo; Nishiura, Hiroshi</td>\n",
       "      <td>Journal of Clinical Medicine</td>\n",
       "      <td>3.006065e+09</td>\n",
       "      <td>#1043</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>e3b40cc8e0e137c416b4a2273a4dca94ae8178cc</td>\n",
       "      <td>CZI</td>\n",
       "      <td>Characteristics of and Public Health Responses to the Coronavirus Disease 2019 Outbreak in China</td>\n",
       "      <td>10.3390/jcm9020575</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32093211.0</td>\n",
       "      <td>cc-by</td>\n",
       "      <td>In December 2019, cases of unidentified pneumonia with a history of exposure in the Huanan Seafood Market were reported in Wuhan, Hubei Province. A novel coronavirus, SARS-CoV-2, was identified to...</td>\n",
       "      <td>2020</td>\n",
       "      <td>Deng, Sheng-Qun; Peng, Hong-Juan</td>\n",
       "      <td>J Clin Med</td>\n",
       "      <td>1.776631e+08</td>\n",
       "      <td>#1999</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>92c2c9839304b4f2bc1276d41b1aa885d8b364fd</td>\n",
       "      <td>CZI</td>\n",
       "      <td>Imaging changes in severe COVID-19 pneumonia</td>\n",
       "      <td>10.1007/s00134-020-05976-w</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32125453.0</td>\n",
       "      <td>cc-by-nc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020</td>\n",
       "      <td>Zhang, Wei</td>\n",
       "      <td>Intensive Care Med</td>\n",
       "      <td>3.006643e+09</td>\n",
       "      <td>#3242</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        sha source_x  \\\n",
       "0  c630ebcdf30652f0422c3ec12a00b50241dc9bd9      CZI   \n",
       "1  53eccda7977a31e3d0f565c884da036b1e85438e      CZI   \n",
       "2  210a892deb1c61577f6fba58505fd65356ce6636      CZI   \n",
       "3  e3b40cc8e0e137c416b4a2273a4dca94ae8178cc      CZI   \n",
       "4  92c2c9839304b4f2bc1276d41b1aa885d8b364fd      CZI   \n",
       "\n",
       "                                                                                                                                                                            title  \\\n",
       "0                                                          Angiotensin-converting enzyme 2 (ACE2) as a SARS-CoV-2 receptor: molecular mechanisms and potential therapeutic target   \n",
       "1                                                             Comparative genetic analysis of the novel coronavirus (2019-nCoV/SARS-CoV-2) receptor ACE2 in different populations   \n",
       "2  Incubation Period and Other Epidemiological Characteristics of 2019 Novel Coronavirus Infections with Right Truncation: A Statistical Analysis of Publicly Available Case Data   \n",
       "3                                                                                Characteristics of and Public Health Responses to the Coronavirus Disease 2019 Outbreak in China   \n",
       "4                                                                                                                                    Imaging changes in severe COVID-19 pneumonia   \n",
       "\n",
       "                          doi pmcid   pubmed_id   license  \\\n",
       "0  10.1007/s00134-020-05985-9   NaN  32125455.0  cc-by-nc   \n",
       "1   10.1038/s41421-020-0147-1   NaN         NaN     cc-by   \n",
       "2          10.3390/jcm9020538   NaN         NaN     cc-by   \n",
       "3          10.3390/jcm9020575   NaN  32093211.0     cc-by   \n",
       "4  10.1007/s00134-020-05976-w   NaN  32125453.0  cc-by-nc   \n",
       "\n",
       "                                                                                                                                                                                                  abstract  \\\n",
       "0                                                                                                                                                                                                      NaN   \n",
       "1                                                                                                                                                                                                      NaN   \n",
       "2  The geographic spread of 2019 novel coronavirus (COVID-19) infections from the epicenter of Wuhan, China, has provided an opportunity to study the natural history of the recently emerged virus. Us...   \n",
       "3  In December 2019, cases of unidentified pneumonia with a history of exposure in the Huanan Seafood Market were reported in Wuhan, Hubei Province. A novel coronavirus, SARS-CoV-2, was identified to...   \n",
       "4                                                                                                                                                                                                      NaN   \n",
       "\n",
       "  publish_time  \\\n",
       "0         2020   \n",
       "1         2020   \n",
       "2         2020   \n",
       "3         2020   \n",
       "4         2020   \n",
       "\n",
       "                                                                                                                                                           authors  \\\n",
       "0                                                                                 Zhang, Haibo; Penninger, Josef M.; Li, Yimin; Zhong, Nanshan; Slutsky, Arthur S.   \n",
       "1                             Cao, Yanan; Li, Lin; Feng, Zhimin; Wan, Shengqing; Huang, Peide; Sun, Xiaohui; Wen, Fang; Huang, Xuanlin; Ning, Guang; Wang, Weiqing   \n",
       "2  Linton, M. Natalie; Kobayashi, Tetsuro; Yang, Yichi; Hayashi, Katsuma; Akhmetzhanov, R. Andrei; Jung, Sung-mok; Yuan, Baoyin; Kinoshita, Ryo; Nishiura, Hiroshi   \n",
       "3                                                                                                                                 Deng, Sheng-Qun; Peng, Hong-Juan   \n",
       "4                                                                                                                                                       Zhang, Wei   \n",
       "\n",
       "                        journal  Microsoft Academic Paper ID WHO #Covidence  \\\n",
       "0            Intensive Care Med                 2.002765e+09          #3252   \n",
       "1                Cell Discovery                 3.003431e+09          #1861   \n",
       "2  Journal of Clinical Medicine                 3.006065e+09          #1043   \n",
       "3                    J Clin Med                 1.776631e+08          #1999   \n",
       "4            Intensive Care Med                 3.006643e+09          #3242   \n",
       "\n",
       "  has_full_text  \n",
       "0          True  \n",
       "1          True  \n",
       "2          True  \n",
       "3          True  \n",
       "4         False  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 29500 entries, 0 to 29499\n",
      "Data columns (total 14 columns):\n",
      "sha                            17420 non-null object\n",
      "source_x                       29500 non-null object\n",
      "title                          29130 non-null object\n",
      "doi                            26357 non-null object\n",
      "pmcid                          27337 non-null object\n",
      "pubmed_id                      16730 non-null float64\n",
      "license                        17692 non-null object\n",
      "abstract                       26553 non-null object\n",
      "publish_time                   18248 non-null object\n",
      "authors                        28554 non-null object\n",
      "journal                        17791 non-null object\n",
      "Microsoft Academic Paper ID    1134 non-null float64\n",
      "WHO #Covidence                 1236 non-null object\n",
      "has_full_text                  17420 non-null object\n",
      "dtypes: float64(2), object(12)\n",
      "memory usage: 3.2+ MB\n"
     ]
    }
   ],
   "source": [
    "meta_data.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sha</th>\n",
       "      <th>source_x</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>license</th>\n",
       "      <th>abstract</th>\n",
       "      <th>publish_time</th>\n",
       "      <th>authors</th>\n",
       "      <th>journal</th>\n",
       "      <th>Microsoft Academic Paper ID</th>\n",
       "      <th>WHO #Covidence</th>\n",
       "      <th>has_full_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>count</td>\n",
       "      <td>17420</td>\n",
       "      <td>29500</td>\n",
       "      <td>29130</td>\n",
       "      <td>26357</td>\n",
       "      <td>27337</td>\n",
       "      <td>1.673000e+04</td>\n",
       "      <td>17692</td>\n",
       "      <td>26553</td>\n",
       "      <td>18248</td>\n",
       "      <td>28554</td>\n",
       "      <td>17791</td>\n",
       "      <td>1.134000e+03</td>\n",
       "      <td>1236</td>\n",
       "      <td>17420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>unique</td>\n",
       "      <td>17398</td>\n",
       "      <td>4</td>\n",
       "      <td>24654</td>\n",
       "      <td>22203</td>\n",
       "      <td>23222</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15</td>\n",
       "      <td>22453</td>\n",
       "      <td>4499</td>\n",
       "      <td>23860</td>\n",
       "      <td>1732</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1223</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>top</td>\n",
       "      <td>72a5640aa0c307fbe171ca7ad55d3fda48b53988</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Articles of Significant Interest Selected from This Issue by the Editors</td>\n",
       "      <td>http://dx.doi.org/10.1371/journal.pcbi.1006483</td>\n",
       "      <td>PMC6224041</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CC BY</td>\n",
       "      <td>The automated comparison of protein-ligand binding sites provides useful insights into yet unexplored site similarities. Various stages of computational and chemical biology research can benefit f...</td>\n",
       "      <td>2020</td>\n",
       "      <td>['Ehrt, Christiane', 'Brinkjost, Tobias', 'Koch, Oliver']</td>\n",
       "      <td>PLoS One</td>\n",
       "      <td>NaN</td>\n",
       "      <td>#2333</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>freq</td>\n",
       "      <td>4</td>\n",
       "      <td>27337</td>\n",
       "      <td>67</td>\n",
       "      <td>35</td>\n",
       "      <td>35</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11575</td>\n",
       "      <td>35</td>\n",
       "      <td>1148</td>\n",
       "      <td>35</td>\n",
       "      <td>2204</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>13219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mean</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.626759e+07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.697968e+09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>std</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.697935e+06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.873586e+08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>min</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.717300e+04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.912630e+07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>25%</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.363481e+07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.366820e+09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>50%</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.704658e+07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.002534e+09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>75%</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.995128e+07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.005536e+09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>max</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.347697e+08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.006672e+09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             sha source_x  \\\n",
       "count                                      17420    29500   \n",
       "unique                                     17398        4   \n",
       "top     72a5640aa0c307fbe171ca7ad55d3fda48b53988      PMC   \n",
       "freq                                           4    27337   \n",
       "mean                                         NaN      NaN   \n",
       "std                                          NaN      NaN   \n",
       "min                                          NaN      NaN   \n",
       "25%                                          NaN      NaN   \n",
       "50%                                          NaN      NaN   \n",
       "75%                                          NaN      NaN   \n",
       "max                                          NaN      NaN   \n",
       "\n",
       "                                                                           title  \\\n",
       "count                                                                      29130   \n",
       "unique                                                                     24654   \n",
       "top     Articles of Significant Interest Selected from This Issue by the Editors   \n",
       "freq                                                                          67   \n",
       "mean                                                                         NaN   \n",
       "std                                                                          NaN   \n",
       "min                                                                          NaN   \n",
       "25%                                                                          NaN   \n",
       "50%                                                                          NaN   \n",
       "75%                                                                          NaN   \n",
       "max                                                                          NaN   \n",
       "\n",
       "                                                   doi       pmcid  \\\n",
       "count                                            26357       27337   \n",
       "unique                                           22203       23222   \n",
       "top     http://dx.doi.org/10.1371/journal.pcbi.1006483  PMC6224041   \n",
       "freq                                                35          35   \n",
       "mean                                               NaN         NaN   \n",
       "std                                                NaN         NaN   \n",
       "min                                                NaN         NaN   \n",
       "25%                                                NaN         NaN   \n",
       "50%                                                NaN         NaN   \n",
       "75%                                                NaN         NaN   \n",
       "max                                                NaN         NaN   \n",
       "\n",
       "           pubmed_id license  \\\n",
       "count   1.673000e+04   17692   \n",
       "unique           NaN      15   \n",
       "top              NaN   CC BY   \n",
       "freq             NaN   11575   \n",
       "mean    2.626759e+07     NaN   \n",
       "std     4.697935e+06     NaN   \n",
       "min     6.717300e+04     NaN   \n",
       "25%     2.363481e+07     NaN   \n",
       "50%     2.704658e+07     NaN   \n",
       "75%     2.995128e+07     NaN   \n",
       "max     1.347697e+08     NaN   \n",
       "\n",
       "                                                                                                                                                                                                       abstract  \\\n",
       "count                                                                                                                                                                                                     26553   \n",
       "unique                                                                                                                                                                                                    22453   \n",
       "top     The automated comparison of protein-ligand binding sites provides useful insights into yet unexplored site similarities. Various stages of computational and chemical biology research can benefit f...   \n",
       "freq                                                                                                                                                                                                         35   \n",
       "mean                                                                                                                                                                                                        NaN   \n",
       "std                                                                                                                                                                                                         NaN   \n",
       "min                                                                                                                                                                                                         NaN   \n",
       "25%                                                                                                                                                                                                         NaN   \n",
       "50%                                                                                                                                                                                                         NaN   \n",
       "75%                                                                                                                                                                                                         NaN   \n",
       "max                                                                                                                                                                                                         NaN   \n",
       "\n",
       "       publish_time  \\\n",
       "count         18248   \n",
       "unique         4499   \n",
       "top            2020   \n",
       "freq           1148   \n",
       "mean            NaN   \n",
       "std             NaN   \n",
       "min             NaN   \n",
       "25%             NaN   \n",
       "50%             NaN   \n",
       "75%             NaN   \n",
       "max             NaN   \n",
       "\n",
       "                                                          authors   journal  \\\n",
       "count                                                       28554     17791   \n",
       "unique                                                      23860      1732   \n",
       "top     ['Ehrt, Christiane', 'Brinkjost, Tobias', 'Koch, Oliver']  PLoS One   \n",
       "freq                                                           35      2204   \n",
       "mean                                                          NaN       NaN   \n",
       "std                                                           NaN       NaN   \n",
       "min                                                           NaN       NaN   \n",
       "25%                                                           NaN       NaN   \n",
       "50%                                                           NaN       NaN   \n",
       "75%                                                           NaN       NaN   \n",
       "max                                                           NaN       NaN   \n",
       "\n",
       "        Microsoft Academic Paper ID WHO #Covidence has_full_text  \n",
       "count                  1.134000e+03           1236         17420  \n",
       "unique                          NaN           1223             2  \n",
       "top                             NaN          #2333          True  \n",
       "freq                            NaN              2         13219  \n",
       "mean                   2.697968e+09            NaN           NaN  \n",
       "std                    4.873586e+08            NaN           NaN  \n",
       "min                    3.912630e+07            NaN           NaN  \n",
       "25%                    2.366820e+09            NaN           NaN  \n",
       "50%                    3.002534e+09            NaN           NaN  \n",
       "75%                    3.005536e+09            NaN           NaN  \n",
       "max                    3.006672e+09            NaN           NaN  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_data.describe(include='all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 24655 entries, 0 to 29138\n",
      "Data columns (total 14 columns):\n",
      "sha                            12819 non-null object\n",
      "source_x                       24655 non-null object\n",
      "title                          24654 non-null object\n",
      "doi                            21620 non-null object\n",
      "pmcid                          22923 non-null object\n",
      "pubmed_id                      12561 non-null float64\n",
      "license                        13089 non-null object\n",
      "abstract                       22365 non-null object\n",
      "publish_time                   14008 non-null object\n",
      "authors                        24274 non-null object\n",
      "journal                        13595 non-null object\n",
      "Microsoft Academic Paper ID    1122 non-null float64\n",
      "WHO #Covidence                 1224 non-null object\n",
      "has_full_text                  12819 non-null object\n",
      "dtypes: float64(2), object(12)\n",
      "memory usage: 2.8+ MB\n"
     ]
    }
   ],
   "source": [
    "meta_data.drop_duplicates(subset=['title'], inplace=True)\n",
    "meta_data.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def doi_url(d): \n",
    "    if d.startswith('http://'):\n",
    "        return d\n",
    "    elif d.startswith('doi.org'):\n",
    "        return f'http://{d}'\n",
    "    else: \n",
    "        return f'http://doi.org/{d}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sha</th>\n",
       "      <th>source_x</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>license</th>\n",
       "      <th>abstract</th>\n",
       "      <th>publish_time</th>\n",
       "      <th>authors</th>\n",
       "      <th>journal</th>\n",
       "      <th>Microsoft Academic Paper ID</th>\n",
       "      <th>WHO #Covidence</th>\n",
       "      <th>has_full_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>c630ebcdf30652f0422c3ec12a00b50241dc9bd9</td>\n",
       "      <td>CZI</td>\n",
       "      <td>Angiotensin-converting enzyme 2 (ACE2) as a SARS-CoV-2 receptor: molecular mechanisms and potential therapeutic target</td>\n",
       "      <td>http://doi.org/10.1007/s00134-020-05985-9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32125455.0</td>\n",
       "      <td>cc-by-nc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020</td>\n",
       "      <td>Zhang, Haibo; Penninger, Josef M.; Li, Yimin; Zhong, Nanshan; Slutsky, Arthur S.</td>\n",
       "      <td>Intensive Care Med</td>\n",
       "      <td>2.002765e+09</td>\n",
       "      <td>#3252</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>53eccda7977a31e3d0f565c884da036b1e85438e</td>\n",
       "      <td>CZI</td>\n",
       "      <td>Comparative genetic analysis of the novel coronavirus (2019-nCoV/SARS-CoV-2) receptor ACE2 in different populations</td>\n",
       "      <td>http://doi.org/10.1038/s41421-020-0147-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cc-by</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020</td>\n",
       "      <td>Cao, Yanan; Li, Lin; Feng, Zhimin; Wan, Shengqing; Huang, Peide; Sun, Xiaohui; Wen, Fang; Huang, Xuanlin; Ning, Guang; Wang, Weiqing</td>\n",
       "      <td>Cell Discovery</td>\n",
       "      <td>3.003431e+09</td>\n",
       "      <td>#1861</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>210a892deb1c61577f6fba58505fd65356ce6636</td>\n",
       "      <td>CZI</td>\n",
       "      <td>Incubation Period and Other Epidemiological Characteristics of 2019 Novel Coronavirus Infections with Right Truncation: A Statistical Analysis of Publicly Available Case Data</td>\n",
       "      <td>http://doi.org/10.3390/jcm9020538</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cc-by</td>\n",
       "      <td>The geographic spread of 2019 novel coronavirus (COVID-19) infections from the epicenter of Wuhan, China, has provided an opportunity to study the natural history of the recently emerged virus. Us...</td>\n",
       "      <td>2020</td>\n",
       "      <td>Linton, M. Natalie; Kobayashi, Tetsuro; Yang, Yichi; Hayashi, Katsuma; Akhmetzhanov, R. Andrei; Jung, Sung-mok; Yuan, Baoyin; Kinoshita, Ryo; Nishiura, Hiroshi</td>\n",
       "      <td>Journal of Clinical Medicine</td>\n",
       "      <td>3.006065e+09</td>\n",
       "      <td>#1043</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>e3b40cc8e0e137c416b4a2273a4dca94ae8178cc</td>\n",
       "      <td>CZI</td>\n",
       "      <td>Characteristics of and Public Health Responses to the Coronavirus Disease 2019 Outbreak in China</td>\n",
       "      <td>http://doi.org/10.3390/jcm9020575</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32093211.0</td>\n",
       "      <td>cc-by</td>\n",
       "      <td>In December 2019, cases of unidentified pneumonia with a history of exposure in the Huanan Seafood Market were reported in Wuhan, Hubei Province. A novel coronavirus, SARS-CoV-2, was identified to...</td>\n",
       "      <td>2020</td>\n",
       "      <td>Deng, Sheng-Qun; Peng, Hong-Juan</td>\n",
       "      <td>J Clin Med</td>\n",
       "      <td>1.776631e+08</td>\n",
       "      <td>#1999</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>92c2c9839304b4f2bc1276d41b1aa885d8b364fd</td>\n",
       "      <td>CZI</td>\n",
       "      <td>Imaging changes in severe COVID-19 pneumonia</td>\n",
       "      <td>http://doi.org/10.1007/s00134-020-05976-w</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32125453.0</td>\n",
       "      <td>cc-by-nc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020</td>\n",
       "      <td>Zhang, Wei</td>\n",
       "      <td>Intensive Care Med</td>\n",
       "      <td>3.006643e+09</td>\n",
       "      <td>#3242</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        sha source_x  \\\n",
       "0  c630ebcdf30652f0422c3ec12a00b50241dc9bd9      CZI   \n",
       "1  53eccda7977a31e3d0f565c884da036b1e85438e      CZI   \n",
       "2  210a892deb1c61577f6fba58505fd65356ce6636      CZI   \n",
       "3  e3b40cc8e0e137c416b4a2273a4dca94ae8178cc      CZI   \n",
       "4  92c2c9839304b4f2bc1276d41b1aa885d8b364fd      CZI   \n",
       "\n",
       "                                                                                                                                                                            title  \\\n",
       "0                                                          Angiotensin-converting enzyme 2 (ACE2) as a SARS-CoV-2 receptor: molecular mechanisms and potential therapeutic target   \n",
       "1                                                             Comparative genetic analysis of the novel coronavirus (2019-nCoV/SARS-CoV-2) receptor ACE2 in different populations   \n",
       "2  Incubation Period and Other Epidemiological Characteristics of 2019 Novel Coronavirus Infections with Right Truncation: A Statistical Analysis of Publicly Available Case Data   \n",
       "3                                                                                Characteristics of and Public Health Responses to the Coronavirus Disease 2019 Outbreak in China   \n",
       "4                                                                                                                                    Imaging changes in severe COVID-19 pneumonia   \n",
       "\n",
       "                                         doi pmcid   pubmed_id   license  \\\n",
       "0  http://doi.org/10.1007/s00134-020-05985-9   NaN  32125455.0  cc-by-nc   \n",
       "1   http://doi.org/10.1038/s41421-020-0147-1   NaN         NaN     cc-by   \n",
       "2          http://doi.org/10.3390/jcm9020538   NaN         NaN     cc-by   \n",
       "3          http://doi.org/10.3390/jcm9020575   NaN  32093211.0     cc-by   \n",
       "4  http://doi.org/10.1007/s00134-020-05976-w   NaN  32125453.0  cc-by-nc   \n",
       "\n",
       "                                                                                                                                                                                                  abstract  \\\n",
       "0                                                                                                                                                                                                      NaN   \n",
       "1                                                                                                                                                                                                      NaN   \n",
       "2  The geographic spread of 2019 novel coronavirus (COVID-19) infections from the epicenter of Wuhan, China, has provided an opportunity to study the natural history of the recently emerged virus. Us...   \n",
       "3  In December 2019, cases of unidentified pneumonia with a history of exposure in the Huanan Seafood Market were reported in Wuhan, Hubei Province. A novel coronavirus, SARS-CoV-2, was identified to...   \n",
       "4                                                                                                                                                                                                      NaN   \n",
       "\n",
       "  publish_time  \\\n",
       "0         2020   \n",
       "1         2020   \n",
       "2         2020   \n",
       "3         2020   \n",
       "4         2020   \n",
       "\n",
       "                                                                                                                                                           authors  \\\n",
       "0                                                                                 Zhang, Haibo; Penninger, Josef M.; Li, Yimin; Zhong, Nanshan; Slutsky, Arthur S.   \n",
       "1                             Cao, Yanan; Li, Lin; Feng, Zhimin; Wan, Shengqing; Huang, Peide; Sun, Xiaohui; Wen, Fang; Huang, Xuanlin; Ning, Guang; Wang, Weiqing   \n",
       "2  Linton, M. Natalie; Kobayashi, Tetsuro; Yang, Yichi; Hayashi, Katsuma; Akhmetzhanov, R. Andrei; Jung, Sung-mok; Yuan, Baoyin; Kinoshita, Ryo; Nishiura, Hiroshi   \n",
       "3                                                                                                                                 Deng, Sheng-Qun; Peng, Hong-Juan   \n",
       "4                                                                                                                                                       Zhang, Wei   \n",
       "\n",
       "                        journal  Microsoft Academic Paper ID WHO #Covidence  \\\n",
       "0            Intensive Care Med                 2.002765e+09          #3252   \n",
       "1                Cell Discovery                 3.003431e+09          #1861   \n",
       "2  Journal of Clinical Medicine                 3.006065e+09          #1043   \n",
       "3                    J Clin Med                 1.776631e+08          #1999   \n",
       "4            Intensive Care Med                 3.006643e+09          #3242   \n",
       "\n",
       "  has_full_text  \n",
       "0          True  \n",
       "1          True  \n",
       "2          True  \n",
       "3          True  \n",
       "4         False  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_data.doi = meta_data.doi.fillna(' ').apply(doi_url)\n",
    "meta_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sha                            11836\n",
       "source_x                           0\n",
       "title                              1\n",
       "doi                                0\n",
       "pmcid                           1732\n",
       "pubmed_id                      12094\n",
       "license                        11566\n",
       "abstract                        2290\n",
       "publish_time                   10647\n",
       "authors                          381\n",
       "journal                        11060\n",
       "Microsoft Academic Paper ID    23533\n",
       "WHO #Covidence                 23431\n",
       "has_full_text                  11836\n",
       "dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_data.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#Joining author names to include middle name \n",
    "def format_name(author):\n",
    "    middle_name = \" \".join(author['middle'])\n",
    "    \n",
    "    if author['middle']:\n",
    "        return \" \".join([author['first'], middle_name, author['last']])\n",
    "    else:\n",
    "        return \" \".join([author['first'], author['last']])\n",
    "\n",
    "def format_affiliation(affiliation):\n",
    "    text = []\n",
    "    location = affiliation.get('location')\n",
    "    if location:\n",
    "        text.extend(list(affiliation['location'].values()))\n",
    "    \n",
    "    institution = affiliation.get('institution')\n",
    "    if institution:\n",
    "        text = [institution] + text\n",
    "    return \", \".join(text)\n",
    "\n",
    "def format_authors(authors, with_affiliation=False):\n",
    "    name_ls = []\n",
    "    \n",
    "    for author in authors:\n",
    "        name = format_name(author)\n",
    "        if with_affiliation:\n",
    "            affiliation = format_affiliation(author['affiliation'])\n",
    "            if affiliation:\n",
    "                name_ls.append(f\"{name} ({affiliation})\")\n",
    "            else:\n",
    "                name_ls.append(name)\n",
    "        else:\n",
    "            name_ls.append(name)\n",
    "    \n",
    "    return \", \".join(name_ls)\n",
    "\n",
    "def format_body(body_text):\n",
    "    texts = [(di['section'], di['text']) for di in body_text]\n",
    "    texts_di = {di['section']: \"\" for di in body_text}\n",
    "    \n",
    "    for section, text in texts:\n",
    "        texts_di[section] += text\n",
    "\n",
    "    body = \"\"\n",
    "\n",
    "    for section, text in texts_di.items():\n",
    "        body += section\n",
    "        body += \"\\n\\n\"\n",
    "        body += text\n",
    "        body += \"\\n\\n\"\n",
    "    \n",
    "    return body\n",
    "\n",
    "def format_bib(bibs):\n",
    "    if type(bibs) == dict:\n",
    "        bibs = list(bibs.values())\n",
    "    bibs = deepcopy(bibs)\n",
    "    formatted = []\n",
    "    \n",
    "    for bib in bibs:\n",
    "        bib['authors'] = format_authors(\n",
    "            bib['authors'], \n",
    "            with_affiliation=False\n",
    "        )\n",
    "        formatted_ls = [str(bib[k]) for k in ['title', 'authors', 'venue', 'year']]\n",
    "        formatted.append(\", \".join(formatted_ls))\n",
    "\n",
    "    return \"; \".join(formatted)\n",
    "\n",
    "def load_files(dirname):\n",
    "    filenames = os.listdir(dirname)\n",
    "    raw_files = []\n",
    "\n",
    "    for filename in tqdm(filenames):\n",
    "        filename = dirname + filename\n",
    "        file = json.load(open(filename, 'rb'))\n",
    "        raw_files.append(file)\n",
    "    \n",
    "    return raw_files\n",
    "\n",
    "def generate_clean_df(all_files):\n",
    "    cleaned_files = []\n",
    "    \n",
    "    for file in tqdm(all_files):\n",
    "        features = [\n",
    "            file['paper_id'],\n",
    "            file['metadata']['title'],\n",
    "            format_authors(file['metadata']['authors']),\n",
    "            format_authors(file['metadata']['authors'], \n",
    "                           with_affiliation=True),\n",
    "            format_body(file['abstract']),\n",
    "            format_body(file['body_text']),\n",
    "            format_bib(file['bib_entries']),\n",
    "            file['metadata']['authors'],\n",
    "            file['bib_entries']\n",
    "        ]\n",
    "\n",
    "        cleaned_files.append(features)\n",
    "\n",
    "    col_names = ['paper_id', 'title', 'authors',\n",
    "                 'affiliations', 'abstract', 'text', \n",
    "                 'bibliography','raw_authors','raw_bibliography']\n",
    "\n",
    "    clean_df = pd.DataFrame(cleaned_files, columns=col_names)\n",
    "    clean_df.head()\n",
    "    \n",
    "    return clean_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of articles retrieved from biorxiv: 803\n"
     ]
    }
   ],
   "source": [
    "biorxiv_div = 'biorxiv_medrxiv/biorxiv_medrxiv/'\n",
    "filenames = os.listdir(biorxiv_div)\n",
    "print('Number of articles retrieved from biorxiv:', len(filenames))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_files = []\n",
    "for filename in filenames:\n",
    "    filename = biorxiv_div + filename\n",
    "    file = json.load(open(filename, 'rb'))\n",
    "    all_files.append(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dicitonary keys: dict_keys(['paper_id', 'metadata', 'abstract', 'body_text', 'bib_entries', 'ref_entries', 'back_matter'])\n"
     ]
    }
   ],
   "source": [
    "file = all_files[2]\n",
    "print('Dicitonary keys:', file.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'cite_spans': [],\n",
      "  'ref_spans': [],\n",
      "  'section': 'Abstract',\n",
      "  'text': 'Infectious bronchitis (IB) causes significant economic losses in '\n",
      "          'the global poultry industry. Control of infectious bronchitis is '\n",
      "          'hindered by the genetic diversity of the causative agent, '\n",
      "          'infectious bronchitis virus (IBV), which has led to the emergence '\n",
      "          'of several serotypes that lack complete serologic cross-protection. '\n",
      "          'While serotyping by definition requires immunologic '\n",
      "          'characterization, genotyping is an efficient means to identify IBVs '\n",
      "          'detected in samples. Sanger sequencing of the S1 subunit of the '\n",
      "          'spike gene is currently used to genotype IBV; however, the '\n",
      "          'universal S1 PCR was created to work from cultured IBV and it is '\n",
      "          'inefficient at detecting mixed isolates. This paper describes a '\n",
      "          'MinION-based AmpSeq method that genetically typed IBV from clinical '\n",
      "          'samples, including samples with multiple isolates. Total RNA was '\n",
      "          'extracted from fifteen tracheal scrapings and choanal cleft swab '\n",
      "          'samples, randomly reverse transcribed, and PCR amplified using '\n",
      "          'modified S1-targeted primers. Amplicons were barcoded to allow for '\n",
      "          \"pooling of samples, processed per manufacturer's instructions into \"\n",
      "          'a 1D MinION sequencing library, and sequenced on the MinION. The '\n",
      "          'AmpSeq method detected IBV in 13 of 14 IBV-positive samples.'},\n",
      " {'cite_spans': [],\n",
      "  'ref_spans': [],\n",
      "  'section': 'Abstract',\n",
      "  'text': 'AmpSeq accurately detected and genotyped both IBV lineages in three '\n",
      "          'of five samples containing two IBV lineages. Additionally, one '\n",
      "          'sample contained three IBV lineages, and AmpSeq accurately detected '\n",
      "          'two of the three. Strain identification, including detection of '\n",
      "          'different strains from the same lineage, was also possible with '\n",
      "          'this AmpSeq method. The results demonstrate the feasibility of '\n",
      "          'using MinION-based AmpSeq for rapid and accurate identification and '\n",
      "          'lineage typing of IBV from oral swab samples.'},\n",
      " {'cite_spans': [{'end': 416, 'ref_id': 'BIBREF7', 'start': 413, 'text': '14,'},\n",
      "                 {'end': 419,\n",
      "                  'ref_id': 'BIBREF25',\n",
      "                  'start': 417,\n",
      "                  'text': '38'}],\n",
      "  'ref_spans': [],\n",
      "  'section': 'Abstract',\n",
      "  'text': 'Infectious bronchitis virus is an enveloped, pleomorphic '\n",
      "          'gammacoronavirus with an unsegmented, single-stranded, '\n",
      "          'positive-sense, 26-27.8 Kb, RNA genome that encodes the '\n",
      "          'nonstructural polyprotein,1a and 1b, and several structural '\n",
      "          'proteins: spike (S), envelop (E), membrane (M) and nucleocapsid '\n",
      "          '(N). 38,41 In addition, two accessory genes, expressing 3a, 3b and '\n",
      "          '5a and 5b, respectively, have also been described. 6, 14, 38 The S '\n",
      "          'protein is highly glycosylated and post-translational cleavage '\n",
      "          'leads to two subunits: S1 and S2. 10,48 Besides acting as the viral '\n",
      "          'attachment protein, the S1 protein is a major target of '\n",
      "          'neutralizing antibodies. 7 As with many attachment proteins that '\n",
      "          'are targets of virus-neutralizing antibodies, the S1 subunit is '\n",
      "          'highly diverse with almost 50% of the amino acids differing among '\n",
      "          'IBV serotypes. 2,21,38 Such variation leads to important biological '\n",
      "          'differences between IBV serotypes and the emergence of novel '\n",
      "          'variants. More than 60 serotypes of IBV have been reported, but the '\n",
      "          'most common serotypes of All rights reserved. No reuse allowed '\n",
      "          'without permission.'},\n",
      " {'cite_spans': [{'end': 17, 'ref_id': 'BIBREF9', 'start': 15, 'text': '16'}],\n",
      "  'ref_spans': [],\n",
      "  'section': 'Abstract',\n",
      "  'text': 'Massachusetts. 16 This genetic diversity leads to emergence of new '\n",
      "          'serotypes and a lack of complete cross-serotype protection by '\n",
      "          'vaccines. 29 The correlation of IBV genotypes and serotypes of IBV '\n",
      "          'has been reported 45 ; therefore, accurate genotypic identification '\n",
      "          'of IBV is an important step to diagnose IBV in clinical respiratory '\n",
      "          'cases, ensure selection of proper IBV vaccines for use in '\n",
      "          'vaccination programs, and to better understand the epidemiology of '\n",
      "          'this global virus.'},\n",
      " {'cite_spans': [],\n",
      "  'ref_spans': [],\n",
      "  'section': 'Abstract',\n",
      "  'text': 'One of the recent comprehensive classification schemes for IBV uses '\n",
      "          'S1 gene sequencebased phylogeny of IBV and identified six genotypes '\n",
      "          '(I-VI), 32 sub-genotypic lineages, and a number of inter-lineage '\n",
      "          'recombinants in global strains of IBV. 44 Among the six genotypes, '\n",
      "          'genotype I (G-I) is the most diverse group of viruses, with 27 '\n",
      "          'unique lineages. 44 As such, sequencing of the S1 subunit provides '\n",
      "          'important information regarding the classification of IBV within '\n",
      "          'samples.'},\n",
      " {'cite_spans': [{'end': 511, 'ref_id': 'BIBREF0', 'start': 509, 'text': '1,'},\n",
      "                 {'end': 515,\n",
      "                  'ref_id': 'BIBREF10',\n",
      "                  'start': 512,\n",
      "                  'text': '17,'},\n",
      "                 {'end': 519,\n",
      "                  'ref_id': 'BIBREF13',\n",
      "                  'start': 516,\n",
      "                  'text': '22,'},\n",
      "                 {'end': 523,\n",
      "                  'ref_id': 'BIBREF14',\n",
      "                  'start': 520,\n",
      "                  'text': '23,'},\n",
      "                 {'end': 527,\n",
      "                  'ref_id': 'BIBREF16',\n",
      "                  'start': 524,\n",
      "                  'text': '25,'},\n",
      "                 {'end': 530,\n",
      "                  'ref_id': 'BIBREF20',\n",
      "                  'start': 528,\n",
      "                  'text': '33'}],\n",
      "  'ref_spans': [],\n",
      "  'section': 'Abstract',\n",
      "  'text': 'The genetic classification of IBV has relied on genotype-specific '\n",
      "          'RT-qPCR assays, serotype-specific S1 RT-qPCR 32,40 and/or pan-IBV '\n",
      "          'S1 RT-PCR assays coupled with Sanger sequencing. 8,31,49 '\n",
      "          'Genotype-specific RT-qPCRs are limited to short fragments, which '\n",
      "          'may miss important changes in the S1 (~1.6 kb) 44 outside the short '\n",
      "          'target. Additionally, since the target is short, the primers lie '\n",
      "          'within the variable regions of the S1 and may require a different '\n",
      "          'assay for each genotype. Lastly, the pan-IBV S1 primers 1, 17, 22, '\n",
      "          '23, 25, 33 , while only requiring one reaction, have relatively low '\n",
      "          'sensitivity and are typically used only on egg-cultured virus, '\n",
      "          'which adds an additional, time consuming step and many diagnostic '\n",
      "          'labs do not have SPF embryos readily available. Lastly, cloning of '\n",
      "          'PCR products has been used to detect multiple IBV sub-populations '\n",
      "          'in the sample 30 ; however, this is inefficient when potentially '\n",
      "          'dealing with multiple (three or four) All rights reserved. No reuse '\n",
      "          'allowed without permission.'},\n",
      " {'cite_spans': [],\n",
      "  'ref_spans': [],\n",
      "  'section': 'Abstract',\n",
      "  'text': 'The copyright holder for this preprint (which was not '\n",
      "          'peer-reviewed) is the author/funder. . '\n",
      "          'https://doi.org/10.1101/634600 doi: bioRxiv preprint IBV types. As '\n",
      "          'such, the development of a pan-IBV sequencing method to rapidly '\n",
      "          'determine the genotype(s) in samples would aid in lineage typing '\n",
      "          'and tracking of IBV genotypes circulating in poultry flocks.'},\n",
      " {'cite_spans': [{'end': 141,\n",
      "                  'ref_id': 'BIBREF21',\n",
      "                  'start': 139,\n",
      "                  'text': '34'}],\n",
      "  'ref_spans': [],\n",
      "  'section': 'Abstract',\n",
      "  'text': 'Recently, third generation sequencing technology has been used for '\n",
      "          'detection of viral nucleotides and sequencing ultra-long DNA '\n",
      "          'molecules. 34 The MinION Nanopore sequencer, a new DNA sequencing '\n",
      "          'technology that allows for rapid, in-house, real-time technique '\n",
      "          'detection and differentiation of IBV lineages, may be cost '\n",
      "          'effective and useful in the field. 19 Ampliconbased sequencing has '\n",
      "          'also been used to amplify specific regions of Newcastle disease '\n",
      "          '3,13 , infectious laryngotracheitis virus 39 Zika 34 , Ebola 35 and '\n",
      "          'avian influenza viruses 46 , by simple RT-PCR and then sequencing '\n",
      "          'on the MinION device. Real-time data analysis, the lack of '\n",
      "          'significant start-up cost investment or maintenance expenses, '\n",
      "          'simultaneous and sequential multiplexing unique to MinION, and the '\n",
      "          'ability to sequence long DNA molecules so that primers are in '\n",
      "          'conserved regions while the product contains the variable region '\n",
      "          'are the features that make this technology highly feasible to be '\n",
      "          'used for disease diagnostics.'},\n",
      " {'cite_spans': [],\n",
      "  'ref_spans': [],\n",
      "  'section': 'Abstract',\n",
      "  'text': 'Accurate identification of IBV genotypes from samples, including '\n",
      "          'detection of multiple isolates from a single sample, is crucial to '\n",
      "          'respiratory disease diagnosis, selection of appropriate IBV '\n",
      "          'vaccines, and epidemiologic studies. Therefore, the aim of the '\n",
      "          'current study is to create a single, amplicon-based protocol to '\n",
      "          'sequence IBV-S1 gene and develop a sequence analysis workflow to '\n",
      "          'identify IBV isolates from clinical swab samples. This method '\n",
      "          'provides a useful assay for IBV and a model for the development of '\n",
      "          'future amplicon sequencing (AmpSeq) based assays.'},\n",
      " {'cite_spans': [],\n",
      "  'ref_spans': [],\n",
      "  'section': 'Abstract',\n",
      "  'text': 'All rights reserved. No reuse allowed without permission.'},\n",
      " {'cite_spans': [],\n",
      "  'ref_spans': [{'end': 118, 'ref_id': None, 'start': 111, 'text': 'Table 2'},\n",
      "                {'end': 805, 'ref_id': None, 'start': 798, 'text': 'Table 2'}],\n",
      "  'section': 'Abstract',\n",
      "  'text': 'A universal S1 primer set, 27 tailed with the MinION universal '\n",
      "          'adapter sequence of 22 nucleotides (underlined, Table 2 ) to allow '\n",
      "          'barcoding of amplicons, was used for targeted amplification of the '\n",
      "          'IBV S1 gene for IBV. The PCR reaction mixture (Expand High Fidelity '\n",
      "          'PCR system, Roche Diagnostics, Mannheim, Germany) was composed of '\n",
      "          '10 µl of cDNA, 1 µl of 10 µM forward primer, 1 µl of 10 µM reverse '\n",
      "          'primer, 2.5 µl of 10× Expand high fidelity buffer, 1 µl of Expand '\n",
      "          'High Fidelity Enzyme Mix, and 1 µl of 10mM dNTPs, to a final volume '\n",
      "          'to 25 µl with 9 µl of nuclease-free water. The following '\n",
      "          'thermocycling conditions were used for amplicon synthesis: '\n",
      "          'denaturation at 95 °C for 2 min; 30 cycles of 94 °C for 1 min, 45 '\n",
      "          '°C for 2 min, and 72 °C for 2 minutes, and final extension at 72 °C '\n",
      "          'for 10 min. See Table 2 for details about all three primer sets '\n",
      "          'used in this study. PCR products were electrophoresed in 1.5% '\n",
      "          'agarose with SYBR Safe DNA gel stain (Invitrogen) for visual '\n",
      "          'evaluation. Amplified DNA was purified by Agencourt AMPure XP beads '\n",
      "          '(Beckman Coulter, USA) at 1.6:1 (volumetric bead:DNA) and '\n",
      "          'quantified using the dsDNA High Sensitivity Assay Kit on a Qubit® '\n",
      "          'Fluorimeter 3.0.'},\n",
      " {'cite_spans': [],\n",
      "  'ref_spans': [],\n",
      "  'section': 'Abstract',\n",
      "  'text': 'The amplicons obtained from the tailed S1 primer set were then used '\n",
      "          'to prepare MinIONcompatible DNA libraries. Briefly, each of the '\n",
      "          'amplicons were diluted to 0.5 nM and amplified using barcoding '\n",
      "          'primers (1D PCR Barcoding Amplicon Kit, Oxford Nanopore '\n",
      "          'Technologies, [ONT], UK) and LongAmp Taq 2X Master Mix (New England '\n",
      "          'Biolabs, MA, USA) with the following conditions: 95 °C for 3 min; '\n",
      "          '15 cycles of 95 °C for 15 seconds, 62 °C for 15 seconds, 65 °C for '\n",
      "          '80 seconds; and 65 °C for 80 seconds. The barcoded amplicons were '\n",
      "          'bead purified (1:1.4; bead:solution), pooled (Run1; pool of 10 '\n",
      "          'samples, Run2; pool of 5 samples) into a single All rights '\n",
      "          'reserved. No reuse allowed without permission.'},\n",
      " {'cite_spans': [],\n",
      "  'ref_spans': [],\n",
      "  'section': 'Abstract',\n",
      "  'text': 'Centrifuge 24 , Kraken 50 , QIIME 5 and Mothur 37 have been used to '\n",
      "          'identify and profile microbial species; however, the high error '\n",
      "          'rate 20 of Nanopore reads translates to poor classification '\n",
      "          'accuracy for many of these tools. Alternatively, de novo 3,26 or '\n",
      "          'reference-based 18 assembly methods have been used in MinION and '\n",
      "          'other deep sequencing platforms. Using a strategy similar to other '\n",
      "          'unbiased laboratory tests (e.g., standard bacterial cultures), an '\n",
      "          'approach was developed to maximize usage of reads (i.e., reads are '\n",
      "          'not discarded based on pre-set length or abundance requirements, '\n",
      "          'similar to how a single colony may be interpreted as a significant '\n",
      "          'result). This approach uses read-based classification against a '\n",
      "          'database containing an equal number of representative sequences per '\n",
      "          'lineage to detect and classify the reads based on their lineage, '\n",
      "          'before conducting read-based classification to the isolate level by '\n",
      "          'using all available IBV sequences in a separate database. Finally, '\n",
      "          'lineage-clustered isolate-level read assignments are interpreted by '\n",
      "          'a veterinary diagnostician to result in reads available for '\n",
      "          'consensus building.'},\n",
      " {'cite_spans': [],\n",
      "  'ref_spans': [],\n",
      "  'section': 'Abstract',\n",
      "  'text': 'The use of a final consensus alignment helps to overcome the '\n",
      "          'individual error rate of MinION sequencing. 3 Similarly, the '\n",
      "          'absence of predetermined metrics used in de novo assembly allows '\n",
      "          'for the informed decision as to how many consensus sequences to '\n",
      "          'build, a bioinformatics problem when dealing with clinical cases '\n",
      "          'that can contain more than one isolate (e.g., similar to how there '\n",
      "          'is not a predetermined number of significant bacterial colonies). '\n",
      "          'Confirmatory followup tests (e.g., RT-PCR) may be needed when '\n",
      "          'dealing with low number of reads in clinical samples; however, '\n",
      "          'future testing of this new technology will allow for creating '\n",
      "          'standards for such confirmatory testing.'},\n",
      " {'cite_spans': [],\n",
      "  'ref_spans': [],\n",
      "  'section': 'Abstract',\n",
      "  'text': 'Taken together, these results suggest the application of '\n",
      "          'MinION-based AmpSeq, specifically for detecting IBV isolates from '\n",
      "          'clinical samples within a few days, compared to several days to '\n",
      "          'weeks and multiple diagnostic assays to culture and detect multiple '\n",
      "          'IBV All rights reserved. No reuse allowed without permission.'}]\n"
     ]
    }
   ],
   "source": [
    "pprint(file['abstract'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "body_text type: <class 'list'>\n",
      "body_)text length: 48\n",
      "body_text keys: dict_keys(['text', 'cite_spans', 'ref_spans', 'section'])\n"
     ]
    }
   ],
   "source": [
    "print('body_text type:', type(file['body_text']))\n",
    "print('body_)text length:', len(file['body_text']))\n",
    "print('body_text keys:', file['body_text'][0].keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "body_text content:\n",
      "[{'cite_spans': [{...}, {...}, {...}, {...}, {...}],\n",
      "  'ref_spans': [],\n",
      "  'section': 'Introduction',\n",
      "  'text': 'Infectious bronchitis (IB), which is caused by infectious '\n",
      "          'bronchitis virus (IBV), is one of the most important diseases of '\n",
      "          'poultry, causing severe economic losses worldwide. 8 Clinical signs '\n",
      "          'of disease are diverse and include respiratory distress, severe '\n",
      "          'ocular discharge, poor body weight gain, decreased egg production, '\n",
      "          'flushing (renal disease), and occasionally mortality in chickens. 7 '\n",
      "          'IB is often complicated by secondary bacterial (e.g., E. coli, '\n",
      "          'Mycoplasma sp.) and viral infections (e.g., avian influenza virus, '\n",
      "          'Newcastle disease virus, and infectious laryngotracheitis virus). '\n",
      "          '43 Lack of cross protection among IBV serotypes is a challenge to '\n",
      "          'controlling IB 15 ; therefore, control of IB relies heavily on '\n",
      "          'serotype-specific live attenuated vaccines. 8 Collectively, the '\n",
      "          'presence of multiple IBVs in a single sample, emergence of variant '\n",
      "          'IBVs, and the high genetic diversity of IBV can complicate the '\n",
      "          'diagnosis of IB and illustrate the need for enhanced diagnostics. '\n",
      "          '15 '},\n",
      " {'cite_spans': [],\n",
      "  'ref_spans': [{...}],\n",
      "  'section': 'Samples',\n",
      "  'text': 'Clinical swab samples (n = 15) were obtained from samples submitted '\n",
      "          'to the Poultry Diagnostic & Research Center, University of Georgia. '\n",
      "          'Detailed information about the samples is provided in Table 1 .'}]\n"
     ]
    }
   ],
   "source": [
    "print('body_text content:')\n",
    "pprint(file['body_text'][:2], depth=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Introduction',\n",
      " 'Samples',\n",
      " 'IBV RT-qPCR assay',\n",
      " 'MinION cDNA synthesis',\n",
      " 'Building customized BLAST databases for AmpSeq analysis',\n",
      " 'AmpSeq IBV lineage and isolate identification',\n",
      " 'Sanger sequencing amplicon synthesis',\n",
      " 'Sanger sequencing analysis',\n",
      " 'Sanger and MinION sequence pairwise identity',\n",
      " 'RT-qPCR assays',\n",
      " 'Sanger sequencing',\n",
      " 'MinION sequencing and lineage identification (Run 1)',\n",
      " 'MinION consensus sequence evaluation for isolate identification',\n",
      " 'MinION sequencing and lineage identification (Run 2)',\n",
      " 'Pairwise sequence identity between Sanger and MinION consensus sequences',\n",
      " 'Discussion']\n",
      "{'AmpSeq IBV lineage and isolate identification': 'A schematic diagram of the '\n",
      "                                                  'workflow of MinION data '\n",
      "                                                  'analysis is presented in '\n",
      "                                                  'Fig. 1 . For Nanopore '\n",
      "                                                  'sequencing data, '\n",
      "                                                  'pre-processing steps were '\n",
      "                                                  'performed to prepare data '\n",
      "                                                  'for downstream analysis. '\n",
      "                                                  'Briefly, Nanopore reads '\n",
      "                                                  '(FAST5) were basecalled '\n",
      "                                                  'using Albacore v 2.02 '\n",
      "                                                  '(https://github.com/Albacore/albacore) '\n",
      "                                                  'with following parameters '\n",
      "                                                  '(read_fast5_basecaller.py '\n",
      "                                                  '-i /Input_reads.fast5 '\n",
      "                                                  '--recursive -t 4 -s '\n",
      "                                                  '/Output_files --flowcell '\n",
      "                                                  'FLO-MIN107 --kit '\n",
      "                                                  'SQK-LSK108o fastq). '\n",
      "                                                  'Porechop '\n",
      "                                                  '(https://github.com/rrwick/Porechop) '\n",
      "                                                  'was used for adapter '\n",
      "                                                  'trimming, (default '\n",
      "                                                  'setting), barcode-based '\n",
      "                                                  'demultiplexing (default '\n",
      "                                                  'settings), and to trim an '\n",
      "                                                  'additional 21 nucleotides, '\n",
      "                                                  'representing the S1 primer '\n",
      "                                                  'sequences (porechop -i '\n",
      "                                                  'Input_file.fastq '\n",
      "                                                  '--extra_end_trim 21 -b '\n",
      "                                                  './output_demultiplexed '\n",
      "                                                  'files.fastq). After barcode '\n",
      "                                                  'and adapter removal, reads '\n",
      "                                                  'were analyzed with a '\n",
      "                                                  'script-based, 2-step data '\n",
      "                                                  'analysis protocol, which '\n",
      "                                                  'includes Centrifuge-kreport '\n",
      "                                                  'as taxonomic read '\n",
      "                                                  'classifier 24 using the '\n",
      "                                                  'sequences in the above '\n",
      "                                                  'mentioned BLAST '\n",
      "                                                  'databases.Briefly, '\n",
      "                                                  'basecalled reads (FASTQ) '\n",
      "                                                  'from individual barcoded '\n",
      "                                                  'samples were used as an '\n",
      "                                                  'input. All rights reserved. '\n",
      "                                                  'No reuse allowed without '\n",
      "                                                  'permission.The copyright '\n",
      "                                                  'holder for this preprint '\n",
      "                                                  '(which was not '\n",
      "                                                  'peer-reviewed) is the '\n",
      "                                                  'author/funder. . '\n",
      "                                                  'https://doi.org/10.1101/634600 '\n",
      "                                                  'doi: bioRxiv preprint '\n",
      "                                                  'First, the basecalled reads '\n",
      "                                                  'were aligned to the '\n",
      "                                                  'IBV-lineage-typing database '\n",
      "                                                  'using BLASTn and reads were '\n",
      "                                                  'clustered based on the read '\n",
      "                                                  'sequence alignment to the '\n",
      "                                                  'respective prototype '\n",
      "                                                  'sequence of IBV lineage. '\n",
      "                                                  'These read clusters were '\n",
      "                                                  'used to interpret presence '\n",
      "                                                  'of IBV genotypes and '\n",
      "                                                  'lineages in the samples. '\n",
      "                                                  'For isolate level '\n",
      "                                                  'identification of IBV, the '\n",
      "                                                  'lineage-based read clusters '\n",
      "                                                  'were individually aligned '\n",
      "                                                  'to the All-IBV database, '\n",
      "                                                  'which produced sub-clusters '\n",
      "                                                  'of reads. Each of these '\n",
      "                                                  'read sub-clusters '\n",
      "                                                  'potentially represents a '\n",
      "                                                  'different isolate and was '\n",
      "                                                  'further used for '\n",
      "                                                  'interpretation. Knowing '\n",
      "                                                  'that MinION sequencing has '\n",
      "                                                  'a high sequencing error '\n",
      "                                                  'rate in individual reads, '\n",
      "                                                  'further steps were added in '\n",
      "                                                  'the data analysis algorithm '\n",
      "                                                  'to obtain a more accurate '\n",
      "                                                  'consensus sequence. '\n",
      "                                                  'Therefore, these read '\n",
      "                                                  'sub-clusters were mapped, '\n",
      "                                                  'using Geneious mapper in '\n",
      "                                                  'Geneious software v11.1.3. '\n",
      "                                                  '(Biomatters, Ltd, Auckland, '\n",
      "                                                  'New Zealand), to the '\n",
      "                                                  'IBV-lineagetyping database '\n",
      "                                                  '(a reference FASTA file) to '\n",
      "                                                  'obtain consensus sequences '\n",
      "                                                  'from each sub-cluster.A '\n",
      "                                                  'minimum threshold for '\n",
      "                                                  'number of reads per '\n",
      "                                                  'consensus sequence was not '\n",
      "                                                  'set as in two samples '\n",
      "                                                  '(sample #4 in Run1 and '\n",
      "                                                  'sample #12 in Run 2), only '\n",
      "                                                  '1 and 5 IBV reads were '\n",
      "                                                  'obtained, respectively; '\n",
      "                                                  'therefore, consensus '\n",
      "                                                  'sequences were built from '\n",
      "                                                  'all the available IBV reads '\n",
      "                                                  'per subcluster. The '\n",
      "                                                  'consensus sequences were '\n",
      "                                                  'compared to GenBank using '\n",
      "                                                  'BLASTn. To select the \"top '\n",
      "                                                  'hit\" from BLASTn output, '\n",
      "                                                  'sequence search results '\n",
      "                                                  'were ordered by \"sequence '\n",
      "                                                  'identity\" and then sequence '\n",
      "                                                  'alignments were evaluated '\n",
      "                                                  'for \"minimum mismatches\" '\n",
      "                                                  'and \"coverage\" of query or '\n",
      "                                                  'subject sequence. All '\n",
      "                                                  'isolates with the highest '\n",
      "                                                  'query or subject coverage '\n",
      "                                                  'and the fewest mismatches '\n",
      "                                                  'were used as \"top hit\" in '\n",
      "                                                  'the final results.',\n",
      " 'Building customized BLAST databases for AmpSeq analysis': 'First, a '\n",
      "                                                            'lineage-typing '\n",
      "                                                            'database, '\n",
      "                                                            'containing 32 IBV '\n",
      "                                                            'S1 gene sequences '\n",
      "                                                            '(one sequence '\n",
      "                                                            'from each of 32 '\n",
      "                                                            'lineages in the 6 '\n",
      "                                                            'genotypes 44 ) '\n",
      "                                                            'and the chicken '\n",
      "                                                            'genome '\n",
      "                                                            '(GCF_000002315.4_Gallus_gallus-5.0), '\n",
      "                                                            'was constructed '\n",
      "                                                            '(referred to as '\n",
      "                                                            '\"IBV-lineage-typing '\n",
      "                                                            'database\" in this '\n",
      "                                                            'study). A second '\n",
      "                                                            'database was '\n",
      "                                                            'constructed with '\n",
      "                                                            'all the Avian '\n",
      "                                                            'Coronaviruses S1 '\n",
      "                                                            'sequences (n = '\n",
      "                                                            '7328) available '\n",
      "                                                            'in NCBI (as of: '\n",
      "                                                            '09-08-2017) and '\n",
      "                                                            'the chicken '\n",
      "                                                            'genome (referred '\n",
      "                                                            'to as \"All-IBV '\n",
      "                                                            'database\" in this '\n",
      "                                                            'study). Prior to '\n",
      "                                                            'database '\n",
      "                                                            'construction, all '\n",
      "                                                            'sequences were '\n",
      "                                                            'dustmasked (NCBI '\n",
      "                                                            'C++ ToolKit), and '\n",
      "                                                            'each isolate was '\n",
      "                                                            'assigned a unique '\n",
      "                                                            'taxonomy ID as a '\n",
      "                                                            'species '\n",
      "                                                            'hierarchically '\n",
      "                                                            'under the genus '\n",
      "                                                            'Gammacoronavirus '\n",
      "                                                            'to allow easy '\n",
      "                                                            'sorting. The '\n",
      "                                                            'local BLAST '\n",
      "                                                            'databases were '\n",
      "                                                            'compiled using '\n",
      "                                                            'default settings. '\n",
      "                                                            'The '\n",
      "                                                            'IBV-lineage-typing '\n",
      "                                                            'database was All '\n",
      "                                                            'rights reserved. '\n",
      "                                                            'No reuse allowed '\n",
      "                                                            'without '\n",
      "                                                            'permission.The '\n",
      "                                                            'copyright holder '\n",
      "                                                            'for this preprint '\n",
      "                                                            '(which was not '\n",
      "                                                            'peer-reviewed) is '\n",
      "                                                            'the '\n",
      "                                                            'author/funder. . '\n",
      "                                                            'https://doi.org/10.1101/634600 '\n",
      "                                                            'doi: bioRxiv '\n",
      "                                                            'preprint used to '\n",
      "                                                            'assign IBV '\n",
      "                                                            'lineages to each '\n",
      "                                                            'read, as '\n",
      "                                                            'appropriate. The '\n",
      "                                                            'All-IBV database '\n",
      "                                                            'was used to '\n",
      "                                                            'assign IBV '\n",
      "                                                            'isolates '\n",
      "                                                            'taxonomic ID to '\n",
      "                                                            'each of IBV reads '\n",
      "                                                            'from within IBV '\n",
      "                                                            'lineage read '\n",
      "                                                            'clusters.The '\n",
      "                                                            'IBV-lineage-typing '\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                            'database, which '\n",
      "                                                            'contains only one '\n",
      "                                                            'sequence per '\n",
      "                                                            'lineage, is '\n",
      "                                                            'required as '\n",
      "                                                            'centrifuge '\n",
      "                                                            'divides the score '\n",
      "                                                            'for any given '\n",
      "                                                            'read by the '\n",
      "                                                            'number of hits '\n",
      "                                                            'that have an '\n",
      "                                                            'equal score.Since '\n",
      "                                                            'some lineages are '\n",
      "                                                            'overrepresented '\n",
      "                                                            'in GenBank (e.g., '\n",
      "                                                            'Mass strains), '\n",
      "                                                            'this results in '\n",
      "                                                            'the scores of '\n",
      "                                                            'those reads '\n",
      "                                                            '(e.g., Mass '\n",
      "                                                            'reads) being '\n",
      "                                                            'divided by a '\n",
      "                                                            'large number, '\n",
      "                                                            'effectively '\n",
      "                                                            'reducing the '\n",
      "                                                            'score for any '\n",
      "                                                            'single alignment. '\n",
      "                                                            'However, because '\n",
      "                                                            'there is only one '\n",
      "                                                            'read per lineage '\n",
      "                                                            'in the '\n",
      "                                                            'IBVlineage-typing '\n",
      "                                                            'database, it has '\n",
      "                                                            'insufficient '\n",
      "                                                            'diversity to '\n",
      "                                                            'determine if '\n",
      "                                                            'there is more '\n",
      "                                                            'than one strain '\n",
      "                                                            'of the same '\n",
      "                                                            'lineage present '\n",
      "                                                            'in a sample. '\n",
      "                                                            'Thus, the All-IBV '\n",
      "                                                            'database provides '\n",
      "                                                            'the diversity '\n",
      "                                                            'required for that '\n",
      "                                                            'analysis.',\n",
      " 'Discussion': 'This study describes a sequencing approach for rapid and '\n",
      "               'accurate identification of IBV genotypes/lineages utilizing '\n",
      "               'the MinION Nanopore sequencer. Additionally, a sequence '\n",
      "               'analysis workflow is described here. The workflow analysis is '\n",
      "               'designed to deal with the multiple, highly similar sequences '\n",
      "               'contained within GenBank that are due to numerous sequencing '\n",
      "               'experiments of IBV (especially vaccine viruses). This protocol '\n",
      "               'demonstrates the feasibility of real-time-based sequencing to '\n",
      "               'provide isolate-level resolution of IBV directly from clinical '\n",
      "               'swab samples, including samples that contain multiple '\n",
      "               'isolates.The proper diagnosis of IBV as the cause of clinical '\n",
      "               'respiratory disease is contingent on virus typing and '\n",
      "               'differentiating live vaccine viruses from field strains. 11 '\n",
      "               'Previously, it was reported that IBV genotypes correlate well '\n",
      "               'with the serotypes of IBV 45 ; therefore, accurate genotypic '\n",
      "               'identification of IBV will be useful to diagnose vaccine and '\n",
      "               'variant viruses in clinical samples. The rapid pan-IBV and '\n",
      "               'serotype specific RT-qPCR assays 4,23,36 have been used for '\n",
      "               'serotype identification; however, positive results from '\n",
      "               'RT-qPCR is insufficient to determine the IBV genotype and '\n",
      "               'obtain isolate-level resolution of IBV; thus, sequence '\n",
      "               'analysis of IBV-S1 gene is required. Previously, use of '\n",
      "               'partial S1 gene sequences (450 bp) to type IBV has been '\n",
      "               'described. 29 However, increasing the length of sequenced S1 '\n",
      "               'gene (~1620 bp) results in more data to be used for genotyping '\n",
      "               'as more of the hypervariable region is covered. 17, 27, 45 For '\n",
      "               'example, the AmpSeq protocol was able to differentiate two '\n",
      "               'genetically very similar (99.5% at S1 gene of IBV genome) but '\n",
      "               'serotypically different GIV-L1 isolates, DE072 and GA98, '\n",
      "               'within a single sample. Thus, the detection of highly diverse '\n",
      "               'IBV genotypes, via the S1 subunit, with a single and rapid '\n",
      "               'sequencing protocol is desirable and the ability to detect '\n",
      "               'multiple isolates in a single sample would improve IBV '\n",
      "               'detection in clinical samples. All rights reserved. No reuse '\n",
      "               'allowed without permission.The copyright holder for this '\n",
      "               'preprint (which was not peer-reviewed) is the author/funder. . '\n",
      "               'https://doi.org/10.1101/634600 doi: bioRxiv preprint '\n",
      "               'Currently, pan-IBV RT-qPCR is used to rapidly detect IBV from '\n",
      "               'clinical samples for screening purposes, 4 but this requires '\n",
      "               'additional serotype-specific RT-qPCRs to genotype positive '\n",
      "               'samples, including samples containing more than one genotype. '\n",
      "               'Additionally, if unidentified IBVs are in the sample, then '\n",
      "               'current detection and characterization may also require egg '\n",
      "               'culture followed by various PCRs. 8 The AmpSeq method '\n",
      "               'described in this report detected IBV in 13 of 14 IBV-positive '\n",
      "               'and detected all the mixed isolates in four of seven samples '\n",
      "               'containing two or more isolates. While mixed isolates were not '\n",
      "               '100% identified (0/2 in #1, 2/4 in #2, and ½ in #3) in three '\n",
      "               'samples with the AmpSeq method, detection of those mixed '\n",
      "               'isolates required several non-AmpSeq assays. Although, there '\n",
      "               'is room for improving this new AmpSeq method, it represents a '\n",
      "               'promising, single-step assay that can be used without egg '\n",
      "               'culture. Thus, it is another tool for the detection of IBV, '\n",
      "               'especially in cases when multiple isolates may be present or '\n",
      "               'when genotyping is especially important.One area that is '\n",
      "               'problematic for a test such as AmpSeq is to detect all the '\n",
      "               'isolates that may be present. While AmpSeq did detect all '\n",
      "               'isolates in 4 samples, in 3 cases not all of the isolates were '\n",
      "               'detected. A potential explanation to partial detection of '\n",
      "               'multiple IBV genotypes could be the relative abundance of IBV '\n",
      "               'genotypes in these clinical samples. It could also be that '\n",
      "               'amplification of these IBV genotypes by serotype-specific '\n",
      "               'RT-qPCR assays is more efficient due to small targeted '\n",
      "               'fragment size (e.g., 120 bp for PCR and ~1600 bp for AmpSeq) '\n",
      "               'and better primer alignment to the target (e.g., degenerate '\n",
      "               'bases are used in the S1 primers used for AmpSeq). One '\n",
      "               'complicating factor of the current AmpSeq protocol is that the '\n",
      "               'IBV target sequence 27 was not originally designed for high '\n",
      "               'specificity, especially from clinical samples. As such, a high '\n",
      "               'proportion of total reads were non-IBV reads (e.g., often '\n",
      "               'mapping to the chicken genome, data not shown), consistent '\n",
      "               'with the extra bands visible in the original report for these '\n",
      "               'All rights reserved. No reuse allowed without permission.The '\n",
      "               'copyright holder for this preprint (which was not '\n",
      "               'peer-reviewed) is the author/funder. . '\n",
      "               'https://doi.org/10.1101/634600 doi: bioRxiv preprint target '\n",
      "               'sequences. 28 Additionally, it is possible that certain '\n",
      "               'genotypes are better complemented to the S1 primers than '\n",
      "               'others and may outcompete those isolates for amplification in '\n",
      "               'the AmpSeq protocol. Lastly, increasing the total number of '\n",
      "               'reads collected by AmpSeq may improve the ability to detect '\n",
      "               'all genotypes in sample by AmpSeq. It is possible to allow the '\n",
      "               'sequencing to continue longer to obtain more reads per sample. '\n",
      "               'Overall, AmpSeq is a feasible test for IBV characterization, '\n",
      "               'and work is ongoing to improve this new type of assay.For '\n",
      "               'example, cost and time efficiency of a sequencing protocol can '\n",
      "               'be improved through multiplexing of more samples in a single '\n",
      "               'sequencing run. 3, 47 In this study, samples (Run 1 = 10, Run '\n",
      "               '2 = 5) were simultaneously multiplexed (i.e., pooled and then '\n",
      "               'sequenced in one run) while maintaining IBV genotyping from '\n",
      "               'data collected. Since the MinION flow cells were not '\n",
      "               'exhausted, and can be washed and re-used, the AmpSeq method '\n",
      "               'also has the potential for sequential multiplexing. This would '\n",
      "               'decrease the need to hold samples for weeks while waiting for '\n",
      "               'the cost-optimal number of samples for simultaneous '\n",
      "               'multiplexing. Alternatively, if detection of all isolates '\n",
      "               'within a given sample is preferred, the sequencing can be run '\n",
      "               'longer to increase sensitivity. The single protocol nature of '\n",
      "               'AmpSeq, the ability to obtain S1 gene sequence results, '\n",
      "               'real-time data analysis, and flexibility of testing design '\n",
      "               'makes MinION-based AmpSeq a viable sequencing protocol for '\n",
      "               'typing of IBV isolates.The advent of real-time, in-house, '\n",
      "               'third generation sequencing represents a transformative '\n",
      "               'opportunity for diagnostic laboratories, by offering the '\n",
      "               'ability to more fully characterize PCR reactions beyond '\n",
      "               'confirming amplicon size (e.g., routine electrophoresis), '\n",
      "               'Sanger sequencing RT-PCR products, or by confirming a partial '\n",
      "               'sequence through probe hybridization (e.g., probe-based qPCR). '\n",
      "               'However, the interpretation of such large data sets represents '\n",
      "               'a challenge to veterinary diagnosticians. Read-based '\n",
      "               'classification software such as genotypes from single sample. '\n",
      "               'Thus, the MinION-based AmpSeq coupled with the data analysis '\n",
      "               'workflow for identification, differentiation, and accurate '\n",
      "               'prediction of IBV genotypes from clinical swab samples can be '\n",
      "               'used as an adjunct to other established rapid diagnostic '\n",
      "               'assays 9,12 until extensive testing of this protocol is done '\n",
      "               'to improve and validate AmpSeq for IBV diagnostics. '\n",
      "               'Furthermore, AmpSeq-based assays can be, and are being applied '\n",
      "               'to other viral pathogens, 3,37 demonstrating the power and '\n",
      "               'utility of this method in the age of molecular diagnostics. '\n",
      "               'The copyright holder for this preprint (which was not '\n",
      "               'peer-reviewed) is the author/funder. . '\n",
      "               'https://doi.org/10.1101/634600 doi: bioRxiv preprint ',\n",
      " 'IBV RT-qPCR assay': 'Total RNA was extracted from each of the swab samples '\n",
      "                      'using QIAamp Viral RNA Mini kit (Qiagen, Hilden, '\n",
      "                      \"Germany) as per manufacturer's instructions, aliquoted, \"\n",
      "                      'and stored at −80 ºC until further use. These samples '\n",
      "                      'were previously tested for IBV and GA08 serotype of '\n",
      "                      'IBV.Briefly, a pan IBV RT-qPCR assay 4 was used to '\n",
      "                      'detect IBV in general and then IBV strain specific '\n",
      "                      'RT-qPCR assay 36 was used to detect GA08 in the samples '\n",
      "                      'using the AgPath ID, One step RT-PCR kit (Thermo Fisher '\n",
      "                      'Scientific, Waltham, MA, USA) on the Applied Biosystems '\n",
      "                      '7500 FAST following the previously described protocols.',\n",
      " 'Introduction': 'Infectious bronchitis (IB), which is caused by infectious '\n",
      "                 'bronchitis virus (IBV), is one of the most important '\n",
      "                 'diseases of poultry, causing severe economic losses '\n",
      "                 'worldwide. 8 Clinical signs of disease are diverse and '\n",
      "                 'include respiratory distress, severe ocular discharge, poor '\n",
      "                 'body weight gain, decreased egg production, flushing (renal '\n",
      "                 'disease), and occasionally mortality in chickens. 7 IB is '\n",
      "                 'often complicated by secondary bacterial (e.g., E. coli, '\n",
      "                 'Mycoplasma sp.) and viral infections (e.g., avian influenza '\n",
      "                 'virus, Newcastle disease virus, and infectious '\n",
      "                 'laryngotracheitis virus). 43 Lack of cross protection among '\n",
      "                 'IBV serotypes is a challenge to controlling IB 15 ; '\n",
      "                 'therefore, control of IB relies heavily on serotype-specific '\n",
      "                 'live attenuated vaccines. 8 Collectively, the presence of '\n",
      "                 'multiple IBVs in a single sample, emergence of variant IBVs, '\n",
      "                 'and the high genetic diversity of IBV can complicate the '\n",
      "                 'diagnosis of IB and illustrate the need for enhanced '\n",
      "                 'diagnostics. 15 ',\n",
      " 'MinION cDNA synthesis': 'For MinION cDNA synthesis, RNA was extracted from '\n",
      "                          '500 µl of each clinical sample using Trizol-LS '\n",
      "                          '(Thermo Fisher Scientific, Waltham, MA, USA), per '\n",
      "                          \"manufacturer's directions.A reaction mixture of 8 \"\n",
      "                          'µl of total RNA, 1 µl of random primers and 1 µl of '\n",
      "                          'dNTPs was incubated at 65 °C for 5 min, chilled on '\n",
      "                          'ice for at least 1 min followed by addition of 10 '\n",
      "                          'µl of cDNA synthesis mix including SuperScript III '\n",
      "                          '(Thermo Fisher Scientific, Waltham, MA, USA) '\n",
      "                          \"according to the manufacture's instruction. The \"\n",
      "                          'reaction was incubated at 25 °C for 10 min, then at '\n",
      "                          '50 °C for 50 min for cDNA synthesis. The reaction '\n",
      "                          'was terminated at 85 °C for 5 min, and then chilled '\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                          'on ice. To remove residual RNA, the cDNA solution '\n",
      "                          'was incubated with RNase H at 37 °C for 20 min. All '\n",
      "                          'rights reserved. No reuse allowed without '\n",
      "                          'permission.The copyright holder for this preprint '\n",
      "                          '(which was not peer-reviewed) is the author/funder. '\n",
      "                          'sequencing protocol was initiated using the MinKNOW '\n",
      "                          'software v.5.12. Sequencing was allowed to continue '\n",
      "                          'for 2 hours until 42,940 reads were obtained for '\n",
      "                          'Run1 and 156,000 reads were obtained for Run 2 in 6 '\n",
      "                          'hours.',\n",
      " 'MinION consensus sequence evaluation for isolate identification': 'It was '\n",
      "                                                                    'observed '\n",
      "                                                                    'that the '\n",
      "                                                                    'single '\n",
      "                                                                    'read '\n",
      "                                                                    'cluster '\n",
      "                                                                    'composed '\n",
      "                                                                    'of less '\n",
      "                                                                    'than 5 '\n",
      "                                                                    'reads '\n",
      "                                                                    '(Run1, '\n",
      "                                                                    'sample '\n",
      "                                                                    '#4, GA98) '\n",
      "                                                                    'yielded a '\n",
      "                                                                    'poor-quality '\n",
      "                                                                    'consensus '\n",
      "                                                                    'sequence, '\n",
      "                                                                    'consistent '\n",
      "                                                                    'with the '\n",
      "                                                                    'known '\n",
      "                                                                    'individual '\n",
      "                                                                    'read '\n",
      "                                                                    'error '\n",
      "                                                                    'rate of '\n",
      "                                                                    'MinION '\n",
      "                                                                    'sequencing. '\n",
      "                                                                    'High-quality '\n",
      "                                                                    'consensus '\n",
      "                                                                    'sequences '\n",
      "                                                                    'were '\n",
      "                                                                    'obtained '\n",
      "                                                                    'from the '\n",
      "                                                                    'other '\n",
      "                                                                    'read '\n",
      "                                                                    'clusters '\n",
      "                                                                    '(>5 reads '\n",
      "                                                                    'per '\n",
      "                                                                    'cluster). '\n",
      "                                                                    'Each of '\n",
      "                                                                    'the '\n",
      "                                                                    'obtained '\n",
      "                                                                    'consensus '\n",
      "                                                                    'sequences '\n",
      "                                                                    'were '\n",
      "                                                                    'compared '\n",
      "                                                                    'to '\n",
      "                                                                    'GenBank '\n",
      "                                                                    'databases '\n",
      "                                                                    'using '\n",
      "                                                                    'BLAST, '\n",
      "                                                                    'which '\n",
      "                                                                    'revealed '\n",
      "                                                                    '>99% '\n",
      "                                                                    'sequence '\n",
      "                                                                    'identity '\n",
      "                                                                    'to '\n",
      "                                                                    'respective '\n",
      "                                                                    'IBV '\n",
      "                                                                    'isolates '\n",
      "                                                                    'Table 4 '\n",
      "                                                                    '). As '\n",
      "                                                                    'described '\n",
      "                                                                    'in the '\n",
      "                                                                    'Sanger '\n",
      "                                                                    'Sequencing '\n",
      "                                                                    'section, '\n",
      "                                                                    'two '\n",
      "                                                                    'samples '\n",
      "                                                                    '(#8 and '\n",
      "                                                                    '#9) '\n",
      "                                                                    'contained '\n",
      "                                                                    'only one '\n",
      "                                                                    'IBV '\n",
      "                                                                    'isolate '\n",
      "                                                                    'per the '\n",
      "                                                                    'non-MinION '\n",
      "                                                                    'assays, '\n",
      "                                                                    'and the '\n",
      "                                                                    'AmpSeq '\n",
      "                                                                    'results '\n",
      "                                                                    'were '\n",
      "                                                                    'consistent '\n",
      "                                                                    'with '\n",
      "                                                                    'these '\n",
      "                                                                    'findings. '\n",
      "                                                                    'Non-MinION '\n",
      "                                                                    'assays '\n",
      "                                                                    'showed '\n",
      "                                                                    'that '\n",
      "                                                                    'seven '\n",
      "                                                                    'samples '\n",
      "                                                                    'contained '\n",
      "                                                                    'two or '\n",
      "                                                                    'more '\n",
      "                                                                    'isolates '\n",
      "                                                                    'per '\n",
      "                                                                    'sample. '\n",
      "                                                                    'Six '\n",
      "                                                                    'samples '\n",
      "                                                                    '(#1, #3, '\n",
      "                                                                    '#4, #6, '\n",
      "                                                                    '#7, and '\n",
      "                                                                    '#10) '\n",
      "                                                                    'contained '\n",
      "                                                                    'two '\n",
      "                                                                    'isolates '\n",
      "                                                                    'per '\n",
      "                                                                    'sample '\n",
      "                                                                    '(Table 4 '\n",
      "                                                                    '). Of '\n",
      "                                                                    'those six '\n",
      "                                                                    'samples, '\n",
      "                                                                    'AmpSeq '\n",
      "                                                                    'detected '\n",
      "                                                                    'both '\n",
      "                                                                    'isolates '\n",
      "                                                                    'in four '\n",
      "                                                                    'samples '\n",
      "                                                                    '(#4, #6, '\n",
      "                                                                    '#7, and '\n",
      "                                                                    '#10), one '\n",
      "                                                                    'isolate '\n",
      "                                                                    'in one '\n",
      "                                                                    'sample '\n",
      "                                                                    '(#3), and '\n",
      "                                                                    'no IBV in '\n",
      "                                                                    'one '\n",
      "                                                                    'sample '\n",
      "                                                                    '(#1; as '\n",
      "                                                                    'mentioned '\n",
      "                                                                    'above, '\n",
      "                                                                    'sample #1 '\n",
      "                                                                    'was the '\n",
      "                                                                    'sample '\n",
      "                                                                    'with the '\n",
      "                                                                    'highest '\n",
      "                                                                    'Cq '\n",
      "                                                                    'value). '\n",
      "                                                                    'Of note, '\n",
      "                                                                    'sample '\n",
      "                                                                    '#10 '\n",
      "                                                                    'contained '\n",
      "                                                                    'two '\n",
      "                                                                    'isolates '\n",
      "                                                                    'from the '\n",
      "                                                                    'same '\n",
      "                                                                    'lineage '\n",
      "                                                                    'and '\n",
      "                                                                    'AmpSeq '\n",
      "                                                                    'was able '\n",
      "                                                                    'to '\n",
      "                                                                    'identify '\n",
      "                                                                    'both '\n",
      "                                                                    'isolates '\n",
      "                                                                    'within '\n",
      "                                                                    'this '\n",
      "                                                                    'sample.Finally, '\n",
      "                                                                    'the '\n",
      "                                                                    'seventh '\n",
      "                                                                    'multi-isolate '\n",
      "                                                                    'sample '\n",
      "                                                                    '(#2) '\n",
      "                                                                    'contained '\n",
      "                                                                    'four '\n",
      "                                                                    'isolates '\n",
      "                                                                    'per the '\n",
      "                                                                    'non-MinION '\n",
      "                                                                    'results, '\n",
      "                                                                    'and the '\n",
      "                                                                    'AmpSeq '\n",
      "                                                                    'detected '\n",
      "                                                                    'two of '\n",
      "                                                                    'those '\n",
      "                                                                    'four '\n",
      "                                                                    'isolates. '\n",
      "                                                                    'These '\n",
      "                                                                    'data show '\n",
      "                                                                    'that a '\n",
      "                                                                    'co-infection '\n",
      "                                                                    'of '\n",
      "                                                                    'multiple '\n",
      "                                                                    'IBV '\n",
      "                                                                    'lineages '\n",
      "                                                                    'existed '\n",
      "                                                                    'in the '\n",
      "                                                                    'above-mentioned '\n",
      "                                                                    'samples '\n",
      "                                                                    'but a '\n",
      "                                                                    'single '\n",
      "                                                                    'RT-qPCR, '\n",
      "                                                                    'or Sanger '\n",
      "                                                                    'sequencing '\n",
      "                                                                    'of a '\n",
      "                                                                    'single '\n",
      "                                                                    'clone may '\n",
      "                                                                    'not have '\n",
      "                                                                    'detected '\n",
      "                                                                    'these '\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                                    'co-infecting '\n",
      "                                                                    'IBV '\n",
      "                                                                    'lineages '\n",
      "                                                                    'and '\n",
      "                                                                    'required '\n",
      "                                                                    'multiple '\n",
      "                                                                    'assays to '\n",
      "                                                                    'detect '\n",
      "                                                                    'all the '\n",
      "                                                                    'IBV '\n",
      "                                                                    'types. '\n",
      "                                                                    'However, '\n",
      "                                                                    'the '\n",
      "                                                                    'AmpSeq '\n",
      "                                                                    'protocol '\n",
      "                                                                    'accurately '\n",
      "                                                                    'detected '\n",
      "                                                                    'multiple '\n",
      "                                                                    'IBV '\n",
      "                                                                    'lineages '\n",
      "                                                                    'in 4 of 7 '\n",
      "                                                                    'samples, '\n",
      "                                                                    'with '\n",
      "                                                                    'partial '\n",
      "                                                                    'detection '\n",
      "                                                                    'in two of '\n",
      "                                                                    'the '\n",
      "                                                                    'remaining '\n",
      "                                                                    'three '\n",
      "                                                                    'samples '\n",
      "                                                                    '(Table 4 '\n",
      "                                                                    '). In '\n",
      "                                                                    'samples '\n",
      "                                                                    'which had '\n",
      "                                                                    'the same '\n",
      "                                                                    'IBV type, '\n",
      "                                                                    'Blast '\n",
      "                                                                    'search of '\n",
      "                                                                    'consensus '\n",
      "                                                                    'All '\n",
      "                                                                    'rights '\n",
      "                                                                    'reserved. '\n",
      "                                                                    'No reuse '\n",
      "                                                                    'allowed '\n",
      "                                                                    'without '\n",
      "                                                                    'permission.The '\n",
      "                                                                    'copyright '\n",
      "                                                                    'holder '\n",
      "                                                                    'for this '\n",
      "                                                                    'preprint '\n",
      "                                                                    '(which '\n",
      "                                                                    'was not '\n",
      "                                                                    'peer-reviewed) '\n",
      "                                                                    'is the '\n",
      "                                                                    'author/funder. '\n",
      "                                                                    '. '\n",
      "                                                                    'https://doi.org/10.1101/634600 '\n",
      "                                                                    'doi: '\n",
      "                                                                    'bioRxiv '\n",
      "                                                                    'preprint '\n",
      "                                                                    'sequences '\n",
      "                                                                    'obtained '\n",
      "                                                                    'from '\n",
      "                                                                    'AmpSeq '\n",
      "                                                                    'and '\n",
      "                                                                    'Sanger '\n",
      "                                                                    'sequencing '\n",
      "                                                                    'identified '\n",
      "                                                                    'the same '\n",
      "                                                                    '(or '\n",
      "                                                                    'highly '\n",
      "                                                                    'related '\n",
      "                                                                    'isolate '\n",
      "                                                                    'for '\n",
      "                                                                    'sample #6 '\n",
      "                                                                    '[GA98] '\n",
      "                                                                    'and '\n",
      "                                                                    'sample '\n",
      "                                                                    '#9) in '\n",
      "                                                                    'the NCBI '\n",
      "                                                                    'nt '\n",
      "                                                                    'database '\n",
      "                                                                    'as per '\n",
      "                                                                    'parameters '\n",
      "                                                                    'described '\n",
      "                                                                    'in the '\n",
      "                                                                    'methods '\n",
      "                                                                    'section.',\n",
      " 'MinION sequencing and lineage identification (Run 1)': 'PCR amplicons '\n",
      "                                                         'obtained directly '\n",
      "                                                         'from 10 clinical '\n",
      "                                                         'swab samples with '\n",
      "                                                         'vaccine IBV '\n",
      "                                                         'serotypes were '\n",
      "                                                         'barcoded, pooled '\n",
      "                                                         'together and '\n",
      "                                                         'sequenced on MinION '\n",
      "                                                         'device (Run 1). A '\n",
      "                                                         'total of 38,661 '\n",
      "                                                         'reads were '\n",
      "                                                         'successfully '\n",
      "                                                         'basecalled from the '\n",
      "                                                         'entire sequencing '\n",
      "                                                         'run (total reads = '\n",
      "                                                         '42,940).After '\n",
      "                                                         'demultiplexing, '\n",
      "                                                         'reads per barcode '\n",
      "                                                         'ranged from 831 to '\n",
      "                                                         '4114. A total of '\n",
      "                                                         '14,845 reads were '\n",
      "                                                         'not assigned to any '\n",
      "                                                         'of the used barcodes '\n",
      "                                                         'and 128 reads were '\n",
      "                                                         'discarded due to '\n",
      "                                                         'middle adapters.The '\n",
      "                                                         'Nanopore reads were '\n",
      "                                                         'queried against '\n",
      "                                                         'IBV-lineage-typing '\n",
      "                                                         'database to '\n",
      "                                                         'determine if the '\n",
      "                                                         'samples contained '\n",
      "                                                         'IBV (Figure 1 ). '\n",
      "                                                         'This AmpSeq protocol '\n",
      "                                                         'detected IBV reads '\n",
      "                                                         'in 8 of 10 '\n",
      "                                                         'samples.The number '\n",
      "                                                         'of IBV reads per '\n",
      "                                                         'sample (IBV positive '\n",
      "                                                         'sample) ranged from '\n",
      "                                                         '56-944. The '\n",
      "                                                         'MinIONnegative '\n",
      "                                                         'samples included '\n",
      "                                                         'sample #5, which was '\n",
      "                                                         'consistently '\n",
      "                                                         'negative with '\n",
      "                                                         'RT-qPCR and RT-PCR '\n",
      "                                                         'assays (and thus '\n",
      "                                                         'interpreted as IBV '\n",
      "                                                         'negative), and '\n",
      "                                                         'sample #1, which had '\n",
      "                                                         'the highest Cq '\n",
      "                                                         'values in the '\n",
      "                                                         'pan-IBV RT-qPCR '\n",
      "                                                         '(Table 1) '\n",
      "                                                         '.Sequencing data was '\n",
      "                                                         'further analyzed to '\n",
      "                                                         'determine if '\n",
      "                                                         'multiple IBV '\n",
      "                                                         'genotypes or '\n",
      "                                                         'lineages could be '\n",
      "                                                         'detected. Lineage-1 '\n",
      "                                                         'and lineage-27 from '\n",
      "                                                         'Genotype I (GI-L1, '\n",
      "                                                         'GI-L27), and '\n",
      "                                                         'lineage-1 from '\n",
      "                                                         'Genotype IV (GIV-L1) '\n",
      "                                                         'were the detected '\n",
      "                                                         'lineages in the '\n",
      "                                                         'samples. Results '\n",
      "                                                         'from individual '\n",
      "                                                         'RT-qPCR assay for '\n",
      "                                                         'GA08 (GI-L27), and '\n",
      "                                                         'RT-PCR assays for '\n",
      "                                                         'GIV-L1and GI-L1 '\n",
      "                                                         'confirmed the '\n",
      "                                                         'presence of multiple '\n",
      "                                                         'IBV genotypes and '\n",
      "                                                         'lineages in 6 of the '\n",
      "                                                         '9 IBV-positive '\n",
      "                                                         'samples (#1, #2, #3, '\n",
      "                                                         '#4, #6, #7) and '\n",
      "                                                         'AmpSeq results '\n",
      "                                                         'confirmed multiple '\n",
      "                                                         'genotypes and '\n",
      "                                                         'lineages in 4 (#2, '\n",
      "                                                         '#4, #6, #7) of those '\n",
      "                                                         '6 samples (AmpSeq '\n",
      "                                                         'failed to detect any '\n",
      "                                                         'IBV in #1, and '\n",
      "                                                         'failed to detect '\n",
      "                                                         'GIV-L1 in #3). All '\n",
      "                                                         'rights reserved. No '\n",
      "                                                         'reuse allowed '\n",
      "                                                         'without '\n",
      "                                                         'permission.The '\n",
      "                                                         'copyright holder for '\n",
      "                                                         'this preprint (which '\n",
      "                                                         'was not '\n",
      "                                                         'peer-reviewed) is '\n",
      "                                                         'the author/funder. . '\n",
      "                                                         'https://doi.org/10.1101/634600 '\n",
      "                                                         'doi: bioRxiv '\n",
      "                                                         'preprintThe presence '\n",
      "                                                         'of a single lineage '\n",
      "                                                         'was confirmed by '\n",
      "                                                         'PCR-based assays and '\n",
      "                                                         'AmpSeq in three '\n",
      "                                                         'samples (#8, # 9 and '\n",
      "                                                         '#10) ( Table 3) .',\n",
      " 'MinION sequencing and lineage identification (Run 2)': 'To evaluate the '\n",
      "                                                         'utility of this '\n",
      "                                                         'protocol on clinical '\n",
      "                                                         'swab samples with '\n",
      "                                                         'pathogenic IBV '\n",
      "                                                         'variants, IBV-S1 '\n",
      "                                                         'gene was amplified '\n",
      "                                                         'directly from five '\n",
      "                                                         'clinical tracheal '\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                         'scrapings and the '\n",
      "                                                         'PCR amplicons were '\n",
      "                                                         'used to create '\n",
      "                                                         'MinION libraries. A '\n",
      "                                                         'total of 146,540 '\n",
      "                                                         'reads were '\n",
      "                                                         'successfully '\n",
      "                                                         'basecalled from the '\n",
      "                                                         'entire sequencing '\n",
      "                                                         'run (total reads = '\n",
      "                                                         '156,000). After '\n",
      "                                                         'demultiplexing, '\n",
      "                                                         'reads per barcode '\n",
      "                                                         'ranged from 4,285 to '\n",
      "                                                         '41,131. A total of '\n",
      "                                                         '24,297 reads were '\n",
      "                                                         'not assigned to any '\n",
      "                                                         'of the used barcodes '\n",
      "                                                         'and 8,109 reads were '\n",
      "                                                         'discarded due to '\n",
      "                                                         'middle adapters in '\n",
      "                                                         'the basecalled '\n",
      "                                                         'reads.Real-time '\n",
      "                                                         'analysis of MinION '\n",
      "                                                         'data, which was '\n",
      "                                                         'obtained within 10 '\n",
      "                                                         'minutes of the '\n",
      "                                                         'sequencing, was '\n",
      "                                                         'sufficient for '\n",
      "                                                         'detection of IBV. '\n",
      "                                                         'However, the '\n",
      "                                                         'sequencing data '\n",
      "                                                         'obtained from entire '\n",
      "                                                         'sequencing run was '\n",
      "                                                         'processed with the '\n",
      "                                                         'same protocol as '\n",
      "                                                         'described above. '\n",
      "                                                         'This AmpSeq protocol '\n",
      "                                                         'detected IBV reads '\n",
      "                                                         'in all of the 5 '\n",
      "                                                         'tested samples using '\n",
      "                                                         'all basecalled '\n",
      "                                                         'reads. The number of '\n",
      "                                                         'IBV reads per sample '\n",
      "                                                         'ranged from 5-4,956. '\n",
      "                                                         'Additionally, the '\n",
      "                                                         'sequencing data '\n",
      "                                                         'analysis showed that '\n",
      "                                                         'the IBV reads '\n",
      "                                                         'belonged to GI-17 '\n",
      "                                                         'lineage and typed '\n",
      "                                                         'the isolate as '\n",
      "                                                         'DMV1639. After '\n",
      "                                                         'MinION sequencing '\n",
      "                                                         'results, these '\n",
      "                                                         'samples were later '\n",
      "                                                         'tested to confirm '\n",
      "                                                         'the presence of IBV '\n",
      "                                                         'variant by '\n",
      "                                                         'MDL_DMV1639 IBV '\n",
      "                                                         'variant specific '\n",
      "                                                         'RT-qPCR assay. All 5 '\n",
      "                                                         'samples were '\n",
      "                                                         'positive for IBV '\n",
      "                                                         'MDL_DMV1639 variant '\n",
      "                                                         'of IBV (Table 5) .',\n",
      " 'Pairwise sequence identity between Sanger and MinION consensus sequences': 'In '\n",
      "                                                                             'all '\n",
      "                                                                             'of '\n",
      "                                                                             'the '\n",
      "                                                                             'samples '\n",
      "                                                                             'where '\n",
      "                                                                             'a '\n",
      "                                                                             'matching '\n",
      "                                                                             'IBV '\n",
      "                                                                             'genotype/lineage '\n",
      "                                                                             'was '\n",
      "                                                                             'identified '\n",
      "                                                                             'by '\n",
      "                                                                             'both '\n",
      "                                                                             'AmpSeq '\n",
      "                                                                             'and '\n",
      "                                                                             'Sanger '\n",
      "                                                                             'sequencing, '\n",
      "                                                                             'the '\n",
      "                                                                             'AmpSeq '\n",
      "                                                                             'data '\n",
      "                                                                             'showed '\n",
      "                                                                             'high '\n",
      "                                                                             'concordance '\n",
      "                                                                             'with '\n",
      "                                                                             'Sanger '\n",
      "                                                                             'sequencing '\n",
      "                                                                             'data, '\n",
      "                                                                             'In '\n",
      "                                                                             'one '\n",
      "                                                                             'sample '\n",
      "                                                                             '(#4), '\n",
      "                                                                             'only '\n",
      "                                                                             'a '\n",
      "                                                                             'single '\n",
      "                                                                             'MinION '\n",
      "                                                                             'read '\n",
      "                                                                             'of '\n",
      "                                                                             'GA98 '\n",
      "                                                                             'was '\n",
      "                                                                             'detected '\n",
      "                                                                             'by '\n",
      "                                                                             'All '\n",
      "                                                                             'rights '\n",
      "                                                                             'reserved. '\n",
      "                                                                             'No '\n",
      "                                                                             'reuse '\n",
      "                                                                             'allowed '\n",
      "                                                                             'without '\n",
      "                                                                             'permission.The '\n",
      "                                                                             'copyright '\n",
      "                                                                             'holder '\n",
      "                                                                             'for '\n",
      "                                                                             'this '\n",
      "                                                                             'preprint '\n",
      "                                                                             '(which '\n",
      "                                                                             'was '\n",
      "                                                                             'not '\n",
      "                                                                             'peer-reviewed) '\n",
      "                                                                             'is '\n",
      "                                                                             'the '\n",
      "                                                                             'author/funder. '\n",
      "                                                                             '. '\n",
      "                                                                             'https://doi.org/10.1101/634600 '\n",
      "                                                                             'doi: '\n",
      "                                                                             'bioRxiv '\n",
      "                                                                             'preprint '\n",
      "                                                                             'sequencing, '\n",
      "                                                                             'consistent '\n",
      "                                                                             'with '\n",
      "                                                                             'the '\n",
      "                                                                             'reported '\n",
      "                                                                             'single-read '\n",
      "                                                                             'accuracy '\n",
      "                                                                             'of '\n",
      "                                                                             'MinION '\n",
      "                                                                             'sequencing. '\n",
      "                                                                             'All '\n",
      "                                                                             'other '\n",
      "                                                                             'samples '\n",
      "                                                                             'showed '\n",
      "                                                                             '99-100% '\n",
      "                                                                             'pairwise '\n",
      "                                                                             'identity '\n",
      "                                                                             'across '\n",
      "                                                                             'the '\n",
      "                                                                             'partial '\n",
      "                                                                             'S1 '\n",
      "                                                                             'fragment '\n",
      "                                                                             'generated '\n",
      "                                                                             'by '\n",
      "                                                                             'Sanger '\n",
      "                                                                             'Sequencing. '\n",
      "                                                                             '(Table '\n",
      "                                                                             '4 '\n",
      "                                                                             ').All '\n",
      "                                                                             'rights '\n",
      "                                                                             'reserved. '\n",
      "                                                                             'No '\n",
      "                                                                             'reuse '\n",
      "                                                                             'allowed '\n",
      "                                                                             'without '\n",
      "                                                                             'permission.The '\n",
      "                                                                             'copyright '\n",
      "                                                                             'holder '\n",
      "                                                                             'for '\n",
      "                                                                             'this '\n",
      "                                                                             'preprint '\n",
      "                                                                             '(which '\n",
      "                                                                             'was '\n",
      "                                                                             'not '\n",
      "                                                                             'peer-reviewed) '\n",
      "                                                                             'is '\n",
      "                                                                             'the '\n",
      "                                                                             'author/funder. '\n",
      "                                                                             '. '\n",
      "                                                                             'https://doi.org/10.1101/634600 '\n",
      "                                                                             'doi: '\n",
      "                                                                             'bioRxiv '\n",
      "                                                                             'preprint',\n",
      " 'RT-qPCR assays': 'Five samples from Run 1 were tested with a pan-IBV RT-qPCR '\n",
      "                   'assay 4 , and four of those were positive for IBV. Samples '\n",
      "                   'with pathogenic IBV strains (Run 2, n = 5) were tested '\n",
      "                   'with the pan-IBV RT-qPCR and all five samples were IBV '\n",
      "                   'positive (Table 1) . Next, a GA08 serotype specific primer '\n",
      "                   'set (in house validated set of primers used for '\n",
      "                   'diagnostics) was used on all samples from Run 1 (including '\n",
      "                   'five samples that were previously tested by the pan-IBV '\n",
      "                   'PCR) and four of 10 samples were positive. Sample #3 and '\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                   '#4, which were positive with the pan-IBV RT-qPCR assay, '\n",
      "                   'were negative with GA08 specific RT-qPCR assay indicating '\n",
      "                   'that these samples contained IBV serotypes other than GA08 '\n",
      "                   'and required further testing. Sample #5 tested negative '\n",
      "                   'with both assays (Table 1) .',\n",
      " 'Samples': 'Clinical swab samples (n = 15) were obtained from samples '\n",
      "            'submitted to the Poultry Diagnostic & Research Center, University '\n",
      "            'of Georgia. Detailed information about the samples is provided in '\n",
      "            'Table 1 .',\n",
      " 'Sanger and MinION sequence pairwise identity': 'A pairwise nucleotide '\n",
      "                                                 'identity comparison on the '\n",
      "                                                 'partial S1 sequences of IBVs '\n",
      "                                                 'from each of the samples '\n",
      "                                                 'sequenced with Sanger '\n",
      "                                                 'sequencing and MinION '\n",
      "                                                 'sequencing was '\n",
      "                                                 'performed.Briefly, the final '\n",
      "                                                 'AmpSeq consensus sequences '\n",
      "                                                 'from identical IBV types in '\n",
      "                                                 'each of the sample were '\n",
      "                                                 'aligned using ClustalW in '\n",
      "                                                 'MEGA6 42 and this alignment '\n",
      "                                                 'was used for pairwise '\n",
      "                                                 'nucleotide identity using '\n",
      "                                                 'MEGA6. 42 All rights '\n",
      "                                                 'reserved. No reuse allowed '\n",
      "                                                 'without permission.The '\n",
      "                                                 'copyright holder for this '\n",
      "                                                 'preprint (which was not '\n",
      "                                                 'peer-reviewed) is the '\n",
      "                                                 'author/funder. . '\n",
      "                                                 'https://doi.org/10.1101/634600 '\n",
      "                                                 'doi: bioRxiv preprint',\n",
      " 'Sanger sequencing': 'A total of 10 clinical swab samples (Run 1) were '\n",
      "                      'processed for Sanger sequencing to confirm the presence '\n",
      "                      'of IBV lineages. Based on the MinION results, primers '\n",
      "                      'targeting GIV-L1 and GI-L1 (GA98 and Conn and primer '\n",
      "                      'sets, respectively) were created. The PCR products were '\n",
      "                      'cloned and multiple (6-24) clones from each sample were '\n",
      "                      'submitted for Sanger sequencing and results are '\n",
      "                      'summarized in Table 3 . Using GIV-L1 serotype primers '\n",
      "                      'showed that 9 out of 10 samples were positive with 1 '\n",
      "                      'sample (#5) negative for IBV (which is consistent with '\n",
      "                      'the pan-IBV RT-qPCR results for this sample). Consensus '\n",
      "                      'sequences from multiple clones obtained by using GIV-L1 '\n",
      "                      'primers showed the presence of GA98 (GA/A9dvaccinated) '\n",
      "                      'in 7 of 9 IBV-positive samples. DE072 was detected in '\n",
      "                      'samples #1 and #9 as the lone GIV-1 isolate but was '\n",
      "                      'also detected in three samples (#2, and #10) that also '\n",
      "                      'contained GA98 (Table 4 ). By using the GI-L1 primer '\n",
      "                      'set, which amplified Conn and Mass type IBVs, 3 of 9 '\n",
      "                      'IBV-positive samples were All rights reserved. No reuse '\n",
      "                      'allowed without permission.The copyright holder for '\n",
      "                      'this preprint (which was not peer-reviewed) is the '\n",
      "                      'author/funder. (Table 4 ). Based on the two '\n",
      "                      'conventional RT-PCRs and the GA08 RT-qPCR, it was '\n",
      "                      'determined that only samples #8 and #9 contained a '\n",
      "                      'single IBV isolate, while the other 7 samples were '\n",
      "                      'positive for 2-4 types of IBV (Table 3) .',\n",
      " 'Sanger sequencing amplicon synthesis': 'For Sanger sequencing amplicons '\n",
      "                                         'synthesis, cDNA from MinION library '\n",
      "                                         'preparation was amplified using the '\n",
      "                                         'following primer sets (Table 2) . '\n",
      "                                         'Firstly, a primer set based on '\n",
      "                                         'Connecticut sequence NCBI FJ904716.1 '\n",
      "                                         'was used to amplify Genotype '\n",
      "                                         'I-lineage 1 viruses (e.g., '\n",
      "                                         'Connecticut All rights reserved. No '\n",
      "                                         'reuse allowed without permission.The '\n",
      "                                         'copyright holder for this preprint '\n",
      "                                         '(which was not peer-reviewed) is the '\n",
      "                                         'author/funder. . '\n",
      "                                         'https://doi.org/10.1101/634600 doi: '\n",
      "                                         'bioRxiv preprint and Massachusetts '\n",
      "                                         'serotypes). Secondly, a primer set '\n",
      "                                         'based on GA98 sequence NCBI '\n",
      "                                         'AF274439.1 was used to amplify '\n",
      "                                         'Genotype IV-lineage 1 viruses (e.g., '\n",
      "                                         'Georgia 1998 and Delaware 072 '\n",
      "                                         'serotypes). Primer sets were '\n",
      "                                         'designed using NCBI Primer-BLAST. 51 '\n",
      "                                         'The PCR reaction mixture is composed '\n",
      "                                         'of 10 µl of cDNA, 0.5 µl of 10 µM '\n",
      "                                         'forward primer, 0.5 µl of 10 µM '\n",
      "                                         'reverse primer, 2.5 µl of 10× '\n",
      "                                         'buffer, 0.5 µl of enzyme mix, and '\n",
      "                                         '0.5 µl of 10mM dNTPs, then made the '\n",
      "                                         'final volume to 25 µl with '\n",
      "                                         'nuclease-free water. The following '\n",
      "                                         'thermocycling conditions were used '\n",
      "                                         'for amplicon synthesis: denaturation '\n",
      "                                         'at 95 °C for 1 min, followed by ',\n",
      " 'Sanger sequencing analysis': 'For Sanger sequencing data, the chromatogram '\n",
      "                               'from each of the samples was manually checked '\n",
      "                               'and primer sequences were trimmed in MEGA 6.0. '\n",
      "                               '42 Forward and reverse sequences from multiple '\n",
      "                               'clones were aligned using MEGA 6.0 software '\n",
      "                               'and consensus sequences were All rights '\n",
      "                               'reserved. No reuse allowed without '\n",
      "                               'permission.The copyright holder for this '\n",
      "                               'preprint (which was not peer-reviewed) is the '\n",
      "                               'author/funder. . '\n",
      "                               'https://doi.org/10.1101/634600 doi: bioRxiv '\n",
      "                               'preprint compared to GenBank using Basic Local '\n",
      "                               'Alignment Tool (BLAST) '\n",
      "                               '(https://blast.ncbi.nlm.nih.gov/) (as of: '\n",
      "                               '12-03-2018) and the top hit from BLASTn output '\n",
      "                               'was selected, as described above for MinION '\n",
      "                               'sequencing data, to identify the IBV isolate '\n",
      "                               'in samples.'}\n"
     ]
    }
   ],
   "source": [
    "texts = [(di['section'], di['text']) for di in file['body_text']]\n",
    "texts_di = {di['section']: '' for di in file['body_text']}\n",
    "for section, text in texts: \n",
    "    texts_di[section] += text\n",
    "\n",
    "pprint(list(texts_di.keys()))\n",
    "pprint(texts_di)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Introduction\n",
      "\n",
      "Infectious bronchitis (IB), which is caused by infectious bronchitis virus (IBV), is one of the most important diseases of poultry, causing severe economic losses worldwide. 8 Clinical signs of disease are diverse and include respiratory distress, severe ocular discharge, poor body weight gain, decreased egg production, flushing (renal disease), and occasionally mortality in chickens. 7 IB is often complicated by secondary bacterial (e.g., E. coli, Mycoplasma sp.) and viral infections (e.g., avian influenza virus, Newcastle disease virus, and infectious laryngotracheitis virus). 43 Lack of cross protection among IBV serotypes is a challenge to controlling IB 15 ; therefore, control of IB relies heavily on serotype-specific live attenuated vaccines. 8 Collectively, the presence of multiple IBVs in a single sample, emergence of variant IBVs, and the high genetic diversity of IBV can complicate the diagnosis of IB and illustrate the need for enhanced diagnostics. 15 \n",
      "\n",
      "Samples\n",
      "\n",
      "Clinical swab samples (n = 15) were obtained from samples submitted to the Poultry Diagnostic & Research Center, University of Georgia. Detailed information about the samples is provided in Table 1 .\n",
      "\n",
      "IBV RT-qPCR assay\n",
      "\n",
      "Total RNA was extracted from each of the swab samples using QIAamp Viral RNA Mini kit (Qiagen, Hilden, Germany) as per manufacturer's instructions, aliquoted, and stored at −80 ºC until further use. These samples were previously tested for IBV and GA08 serotype of IBV.Briefly, a pan IBV RT-qPCR assay 4 was used to detect IBV in general and then IBV strain specific RT-qPCR assay 36 was used to detect GA08 in the samples using the AgPath ID, One step RT-PCR kit (Thermo Fisher Scientific, Waltham, MA, USA) on the Applied Biosystems 7500 FAST following the previously described protocols.\n",
      "\n",
      "MinION cDNA synthesis\n",
      "\n",
      "For MinION cDNA synthesis, RNA was extracted from 500 µl of each clinical sample using Trizol-LS (Thermo Fisher Scientific, Waltham, MA, USA), per manufacturer's directions.A reaction mixture of 8 µl of total RNA, 1 µl of random primers and 1 µl of dNTPs was incubated at 65 °C for 5 min, chilled on ice for at least 1 min followed by addition of 10 µl of cDNA synthesis mix including SuperScript III (Thermo Fisher Scientific, Waltham, MA, USA) according to the manufacture's instruction. The reaction was incubated at 25 °C for 10 min, then at 50 °C for 50 min for cDNA synthesis. The reaction was terminated at 85 °C for 5 min, and then chilled on ice. To remove residual RNA, the cDNA solution was incubated with RNase H at 37 °C for 20 min. All rights reserved. No reuse allowed without permission.The copyright holder for this preprint (which was not peer-reviewed) is the author/funder. sequencing protocol was initiated using the MinKNOW software v.5.12. Sequencing was allowed to continue for 2 hours until 42,940 reads were obtained for Run1 and 156,000 reads were obtained for Run 2 in 6 hours.\n",
      "\n",
      "Building customized BLAST databases for AmpSeq analysis\n"
     ]
    }
   ],
   "source": [
    "body = ''\n",
    "\n",
    "#for section, text in texts_di.items():\n",
    "#    body += section\n",
    "#    body += '\\n\\n'\n",
    "#    body += text\n",
    "#    body += '\\n\\n'\n",
    "\n",
    "#print(body[:3000])\n",
    "\n",
    "print(format_body(file['body_text'])[:3000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['title', 'authors'])\n"
     ]
    }
   ],
   "source": [
    "print(all_files[0]['metadata'].keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The RNA pseudoknots in foot-and-mouth disease virus are dispensable for genome replication but essential for the production of infectious virus. 2 3\n"
     ]
    }
   ],
   "source": [
    "print(all_files[0]['metadata']['title'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'affiliation': {'institution': 'City University of Hong Kong',\n",
      "                  'laboratory': '',\n",
      "                  'location': {'country': 'China', 'settlement': 'Hong Kong'}},\n",
      "  'email': '',\n",
      "  'first': 'Hanchu',\n",
      "  'last': 'Zhou',\n",
      "  'middle': [],\n",
      "  'suffix': ''},\n",
      " {'affiliation': {'institution': 'City University of Hong Kong',\n",
      "                  'laboratory': '',\n",
      "                  'location': {'country': 'China', 'settlement': 'Hong Kong'}},\n",
      "  'email': '',\n",
      "  'first': 'Jiannan',\n",
      "  'last': 'Yang',\n",
      "  'middle': [],\n",
      "  'suffix': ''},\n",
      " {'affiliation': {'institution': 'City University of Hong Kong',\n",
      "                  'laboratory': '',\n",
      "                  'location': {'country': 'China', 'settlement': 'Hong Kong'}},\n",
      "  'email': '',\n",
      "  'first': 'Kaicheng',\n",
      "  'last': 'Tang',\n",
      "  'middle': [],\n",
      "  'suffix': ''}]\n"
     ]
    }
   ],
   "source": [
    "authors = all_files[1]['metadata']['authors']\n",
    "pprint(authors[:3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name: Hanchu Zhou\n",
      "Affiliation: City University of Hong Kong, Hong Kong, China\n",
      "\n",
      "Name: Jiannan Yang\n",
      "Affiliation: City University of Hong Kong, Hong Kong, China\n",
      "\n",
      "Name: Kaicheng Tang\n",
      "Affiliation: City University of Hong Kong, Hong Kong, China\n",
      "\n",
      "Name: † \n",
      "Affiliation: \n",
      "\n",
      "Name: Qingpeng Zhang\n",
      "Affiliation: City University of Hong Kong, Hong Kong, China\n",
      "\n",
      "Name: Zhidong Cao\n",
      "Affiliation: Chinese Academy of Sciences, Beijing, China\n",
      "\n",
      "Name: Dirk Pfeiffer\n",
      "Affiliation: City University of Hong Kong, Hong Kong, China\n",
      "\n",
      "Name: Daniel Dajun Zeng\n",
      "Affiliation: Chinese Academy of Sciences, Beijing, China\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for author in authors:\n",
    "    print('Name:', format_name(author))\n",
    "    print('Affiliation:', format_affiliation(author['affiliation']))\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'authors': [{'affiliation': {'institution': 'Chinese Academy of Sciences',\n",
      "                              'laboratory': '',\n",
      "                              'location': {...}},\n",
      "              'email': '',\n",
      "              'first': 'Xiaoyang',\n",
      "              'last': 'Ji',\n",
      "              'middle': [],\n",
      "              'suffix': ''},\n",
      "             {'affiliation': {'institution': 'Chinese Academy of Sciences',\n",
      "                              'laboratory': '',\n",
      "                              'location': {...}},\n",
      "              'email': '',\n",
      "              'first': 'Chunming',\n",
      "              'last': 'Zhang',\n",
      "              'middle': [],\n",
      "              'suffix': ''},\n",
      "             {'affiliation': {'institution': 'Chinese Academy of Sciences',\n",
      "                              'laboratory': 'State Key Laboratory of Computer '\n",
      "                                            'Architecture',\n",
      "                              'location': {...}},\n",
      "              'email': '',\n",
      "              'first': 'Yubo',\n",
      "              'last': 'Zhai',\n",
      "              'middle': [],\n",
      "              'suffix': ''},\n",
      "             {'affiliation': {},\n",
      "              'email': '',\n",
      "              'first': 'Zhonghai',\n",
      "              'last': 'Zhang',\n",
      "              'middle': [],\n",
      "              'suffix': ''},\n",
      "             {'affiliation': {'institution': 'Phil Rivers Technology',\n",
      "                              'laboratory': '',\n",
      "                              'location': {...}},\n",
      "              'email': '',\n",
      "              'first': 'Chunli',\n",
      "              'last': 'Zhang',\n",
      "              'middle': [],\n",
      "              'suffix': ''},\n",
      "             {'affiliation': {'institution': 'Chinese Academy of Sciences',\n",
      "                              'laboratory': '',\n",
      "                              'location': {...}},\n",
      "              'email': '',\n",
      "              'first': 'Yiqing',\n",
      "              'last': 'Xue',\n",
      "              'middle': [],\n",
      "              'suffix': ''},\n",
      "             {'affiliation': {'institution': 'Chinese Academy of Sciences',\n",
      "                              'laboratory': 'State Key Laboratory of Computer '\n",
      "                                            'Architecture',\n",
      "                              'location': {...}},\n",
      "              'email': '',\n",
      "              'first': 'Guangming',\n",
      "              'last': 'Tan',\n",
      "              'middle': [],\n",
      "              'suffix': ''},\n",
      "             {'affiliation': {'institution': 'Chinese Academy of Sciences',\n",
      "                              'laboratory': '',\n",
      "                              'location': {...}},\n",
      "              'email': 'g.niu@philrivers.com',\n",
      "              'first': 'Gang',\n",
      "              'last': 'Niu',\n",
      "              'middle': [],\n",
      "              'suffix': ''}],\n",
      " 'title': 'TWIRLS, an automated topic-wise inference method based on massive '\n",
      "          'literature, suggests a possible mechanism via ACE2 for the '\n",
      "          'pathological changes in the human host after coronavirus infection'}\n"
     ]
    }
   ],
   "source": [
    "pprint(all_files[4]['metadata'], depth=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formatting without affiliation:\n",
      "Xiaoyang Ji, Chunming Zhang, Yubo Zhai, Zhonghai Zhang, Chunli Zhang, Yiqing Xue, Guangming Tan, Gang Niu\n",
      "\n",
      "Formatting with affiliation:\n",
      "Xiaoyang Ji (Chinese Academy of Sciences, Beijing, China), Chunming Zhang (Chinese Academy of Sciences, Beijing, China), Yubo Zhai (Chinese Academy of Sciences, Beijing, China), Zhonghai Zhang, Chunli Zhang (Phil Rivers Technology, Beijing, China), Yiqing Xue (Chinese Academy of Sciences, Beijing, China), Guangming Tan (Chinese Academy of Sciences, Beijing, China), Gang Niu (Chinese Academy of Sciences, Beijing, China)\n"
     ]
    }
   ],
   "source": [
    "authors = all_files[4]['metadata']['authors']\n",
    "print('Formatting without affiliation:')\n",
    "print(format_authors(authors, with_affiliation=False))\n",
    "print('\\nFormatting with affiliation:')\n",
    "print(format_authors(authors, with_affiliation=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'authors': [{'first': 'S', 'last': 'Abro', 'middle': [...], 'suffix': ''}],\n",
      "  'issn': '',\n",
      "  'other_ids': {},\n",
      "  'pages': '237--246',\n",
      "  'ref_id': 'b0',\n",
      "  'title': 'Emergence of novel strains of avian infectious bronchitis virus in '\n",
      "           'Sweden',\n",
      "  'venue': 'Vet Microbiol',\n",
      "  'volume': '155',\n",
      "  'year': 2012},\n",
      " {'authors': [{'first': 'Y', 'last': 'Bochkov', 'middle': [...], 'suffix': ''}],\n",
      "  'issn': '',\n",
      "  'other_ids': {},\n",
      "  'pages': '65--71',\n",
      "  'ref_id': 'b1',\n",
      "  'title': 'Phylogenetic analysis of partial S1 and N gene sequences of '\n",
      "           'infectious bronchitis virus isolates from Italy revealed genetic '\n",
      "           'diversity and recombination',\n",
      "  'venue': 'Virus Genes',\n",
      "  'volume': '35',\n",
      "  'year': 2007}]\n"
     ]
    }
   ],
   "source": [
    "bibs = list(file['bib_entries'].values())\n",
    "pprint(bibs[:2], depth=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Y A Bochkov'"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "format_authors(bibs[1]['authors'], with_affiliation=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Emergence of novel strains of avian infectious bronchitis virus in Sweden, S H Abro, Vet Microbiol, 2012; Phylogenetic analysis of partial S1 and N gene sequences of infectious bronchitis virus isolates from Italy revealed genetic diversity and recombination, Y A Bochkov, Virus Genes, 2007; Rapid virulence prediction and identification of Newcastle disease virus genotypes using third-generation sequencing, S L Butt, Virol J, 2018; Development and evaluation of a real-time Taqman RT-PCR assay for the detection of infectious bronchitis virus from infected chickens, S A Callison, J Virol Methods, 2006; Infectious bronchitis virus infections of chickens in Belgium: an epidemiological survey, P De Herdt, Vlaams Diergeneeskundig Tijdschrift, 2016\n"
     ]
    }
   ],
   "source": [
    "bib_formatted = format_bib(bibs[:5])\n",
    "print(bib_formatted)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6cc8aa1e5a784acc930cc986083cc2c9",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=803), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "cleaned_files = []\n",
    "\n",
    "for file in tqdm(all_files):\n",
    "    features = [\n",
    "        file['paper_id'],\n",
    "        file['metadata']['title'],\n",
    "        format_authors(file['metadata']['authors']),\n",
    "        format_authors(file['metadata']['authors'],\n",
    "                      with_affiliation=True),\n",
    "        format_body(file['abstract']),\n",
    "        format_body(file['body_text']),\n",
    "        format_bib(file['bib_entries']),\n",
    "        file['metadata']['authors'],\n",
    "        file['bib_entries']\n",
    "    ]\n",
    "    \n",
    "    cleaned_files.append(features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paper_id</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>affiliations</th>\n",
       "      <th>abstract</th>\n",
       "      <th>text</th>\n",
       "      <th>bibliography</th>\n",
       "      <th>raw_authors</th>\n",
       "      <th>raw_bibliography</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0015023cc06b5362d332b3baf348d11567ca2fbb</td>\n",
       "      <td>The RNA pseudoknots in foot-and-mouth disease ...</td>\n",
       "      <td>Joseph C Ward, Lidia Lasecka-Dykes, Chris Neil...</td>\n",
       "      <td>Joseph C Ward, Lidia Lasecka-Dykes, Chris Neil...</td>\n",
       "      <td>Abstract\\n\\nword count: 194 22 Text word count...</td>\n",
       "      <td>\\n\\nVP3, and VP0 (which is further processed t...</td>\n",
       "      <td>Genetic economy in 598 picornaviruses: Foot-an...</td>\n",
       "      <td>[{'first': 'Joseph', 'middle': ['C'], 'last': ...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'Genetic...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>004f0f8bb66cf446678dc13cf2701feec4f36d76</td>\n",
       "      <td>Healthcare-resource-adjusted vulnerabilities t...</td>\n",
       "      <td>Hanchu Zhou, Jiannan Yang, Kaicheng Tang, † , ...</td>\n",
       "      <td>Hanchu Zhou (City University of Hong Kong, Hon...</td>\n",
       "      <td></td>\n",
       "      <td>Introduction\\n\\nThe 2019-nCoV epidemic has spr...</td>\n",
       "      <td>World Health Organizations. Novel Coronavirus ...</td>\n",
       "      <td>[{'first': 'Hanchu', 'middle': [], 'last': 'Zh...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'World H...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>00d16927588fb04d4be0e6b269fc02f0d3c2aa7b</td>\n",
       "      <td>Real-time, MinION-based, amplicon sequencing f...</td>\n",
       "      <td>Salman L Butt, Eric C Erwood, Jian Zhang, Holl...</td>\n",
       "      <td>Salman L Butt (University of Georgia, 30602, A...</td>\n",
       "      <td>Abstract\\n\\nInfectious bronchitis (IB) causes ...</td>\n",
       "      <td>Introduction\\n\\nInfectious bronchitis (IB), wh...</td>\n",
       "      <td>Emergence of novel strains of avian infectious...</td>\n",
       "      <td>[{'first': 'Salman', 'middle': ['L'], 'last': ...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'Emergen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>013d9d1cba8a54d5d3718c229b812d7cf91b6c89</td>\n",
       "      <td>Assessing spread risk of Wuhan novel coronavir...</td>\n",
       "      <td>Shengjie Lai, Isaac I Bogoch, Nick W Ruktanonc...</td>\n",
       "      <td>Shengjie Lai (University of Southampton, UK), ...</td>\n",
       "      <td>Abstract\\n\\nBackground: A novel coronavirus (2...</td>\n",
       "      <td>Introduction\\n\\nIn December 2019, a cluster of...</td>\n",
       "      <td>A Novel Coronavirus Genome Identified in a Clu...</td>\n",
       "      <td>[{'first': 'Shengjie', 'middle': [], 'last': '...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'A Novel...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>01d162d7fae6aaba8e6e60e563ef4c2fca7b0e18</td>\n",
       "      <td>TWIRLS, an automated topic-wise inference meth...</td>\n",
       "      <td>Xiaoyang Ji, Chunming Zhang, Yubo Zhai, Zhongh...</td>\n",
       "      <td>Xiaoyang Ji (Chinese Academy of Sciences, Beij...</td>\n",
       "      <td>Abstract\\n\\nFaced with the current large-scale...</td>\n",
       "      <td>Introduction\\n\\nThe sudden outbreak of the new...</td>\n",
       "      <td>A pneumonia outbreak associated with a new cor...</td>\n",
       "      <td>[{'first': 'Xiaoyang', 'middle': [], 'last': '...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'A pneum...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   paper_id  \\\n",
       "0  0015023cc06b5362d332b3baf348d11567ca2fbb   \n",
       "1  004f0f8bb66cf446678dc13cf2701feec4f36d76   \n",
       "2  00d16927588fb04d4be0e6b269fc02f0d3c2aa7b   \n",
       "3  013d9d1cba8a54d5d3718c229b812d7cf91b6c89   \n",
       "4  01d162d7fae6aaba8e6e60e563ef4c2fca7b0e18   \n",
       "\n",
       "                                               title  \\\n",
       "0  The RNA pseudoknots in foot-and-mouth disease ...   \n",
       "1  Healthcare-resource-adjusted vulnerabilities t...   \n",
       "2  Real-time, MinION-based, amplicon sequencing f...   \n",
       "3  Assessing spread risk of Wuhan novel coronavir...   \n",
       "4  TWIRLS, an automated topic-wise inference meth...   \n",
       "\n",
       "                                             authors  \\\n",
       "0  Joseph C Ward, Lidia Lasecka-Dykes, Chris Neil...   \n",
       "1  Hanchu Zhou, Jiannan Yang, Kaicheng Tang, † , ...   \n",
       "2  Salman L Butt, Eric C Erwood, Jian Zhang, Holl...   \n",
       "3  Shengjie Lai, Isaac I Bogoch, Nick W Ruktanonc...   \n",
       "4  Xiaoyang Ji, Chunming Zhang, Yubo Zhai, Zhongh...   \n",
       "\n",
       "                                        affiliations  \\\n",
       "0  Joseph C Ward, Lidia Lasecka-Dykes, Chris Neil...   \n",
       "1  Hanchu Zhou (City University of Hong Kong, Hon...   \n",
       "2  Salman L Butt (University of Georgia, 30602, A...   \n",
       "3  Shengjie Lai (University of Southampton, UK), ...   \n",
       "4  Xiaoyang Ji (Chinese Academy of Sciences, Beij...   \n",
       "\n",
       "                                            abstract  \\\n",
       "0  Abstract\\n\\nword count: 194 22 Text word count...   \n",
       "1                                                      \n",
       "2  Abstract\\n\\nInfectious bronchitis (IB) causes ...   \n",
       "3  Abstract\\n\\nBackground: A novel coronavirus (2...   \n",
       "4  Abstract\\n\\nFaced with the current large-scale...   \n",
       "\n",
       "                                                text  \\\n",
       "0  \\n\\nVP3, and VP0 (which is further processed t...   \n",
       "1  Introduction\\n\\nThe 2019-nCoV epidemic has spr...   \n",
       "2  Introduction\\n\\nInfectious bronchitis (IB), wh...   \n",
       "3  Introduction\\n\\nIn December 2019, a cluster of...   \n",
       "4  Introduction\\n\\nThe sudden outbreak of the new...   \n",
       "\n",
       "                                        bibliography  \\\n",
       "0  Genetic economy in 598 picornaviruses: Foot-an...   \n",
       "1  World Health Organizations. Novel Coronavirus ...   \n",
       "2  Emergence of novel strains of avian infectious...   \n",
       "3  A Novel Coronavirus Genome Identified in a Clu...   \n",
       "4  A pneumonia outbreak associated with a new cor...   \n",
       "\n",
       "                                         raw_authors  \\\n",
       "0  [{'first': 'Joseph', 'middle': ['C'], 'last': ...   \n",
       "1  [{'first': 'Hanchu', 'middle': [], 'last': 'Zh...   \n",
       "2  [{'first': 'Salman', 'middle': ['L'], 'last': ...   \n",
       "3  [{'first': 'Shengjie', 'middle': [], 'last': '...   \n",
       "4  [{'first': 'Xiaoyang', 'middle': [], 'last': '...   \n",
       "\n",
       "                                    raw_bibliography  \n",
       "0  {'BIBREF0': {'ref_id': 'b0', 'title': 'Genetic...  \n",
       "1  {'BIBREF0': {'ref_id': 'b0', 'title': 'World H...  \n",
       "2  {'BIBREF0': {'ref_id': 'b0', 'title': 'Emergen...  \n",
       "3  {'BIBREF0': {'ref_id': 'b0', 'title': 'A Novel...  \n",
       "4  {'BIBREF0': {'ref_id': 'b0', 'title': 'A pneum...  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "col_names = [\n",
    "    'paper_id',\n",
    "    'title', \n",
    "    'authors',\n",
    "    'affiliations',\n",
    "    'abstract', \n",
    "    'text', \n",
    "    'bibliography',\n",
    "    'raw_authors',\n",
    "    'raw_bibliography'\n",
    "]\n",
    "\n",
    "clean_df = pd.DataFrame(cleaned_files, columns =col_names)\n",
    "clean_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "clean_df.to_csv('biorxiv_clean.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8bda7be4611542a4a2793cf234696dff",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=1426), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "531d11f2e4f04bfc8c64162cb3d856d6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=1426), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paper_id</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>affiliations</th>\n",
       "      <th>abstract</th>\n",
       "      <th>text</th>\n",
       "      <th>bibliography</th>\n",
       "      <th>raw_authors</th>\n",
       "      <th>raw_bibliography</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>002f09dfc9a1323a15bf72e349d8b733ac97a2aa</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>\\n\\nT he modern word camel is derived from the...</td>\n",
       "      <td>Ancient and modern DNA reveal dynamics of dome...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'Ancient...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>0036e8891c93ae63611bde179ada1e03e8577dea</td>\n",
       "      <td>Stable Occupancy of the Crimean-Congo Hemorrha...</td>\n",
       "      <td>Florine E M Scholte, Brian L Hua, Jessica R Sp...</td>\n",
       "      <td>Florine E M Scholte (National Center for Emerg...</td>\n",
       "      <td>Abstract\\n\\nCrimean-Congo hemorrhagic fever vi...</td>\n",
       "      <td>\\n\\nto Western Europe with the assistance of m...</td>\n",
       "      <td>Seroepidemiological studies of Crimean-Congo h...</td>\n",
       "      <td>[{'first': 'Florine', 'middle': ['E M'], 'last...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'Seroepi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>00573277e6be50669016f770bc28ec2da0639a8f</td>\n",
       "      <td>Asymptomatic Severe Acute Respiratory Syndrome...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>\\n\\nWe identified a nurse who was asymptomatic...</td>\n",
       "      <td>Coronavirus as a possible cause of severe acut...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'Coronav...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>00683d59d56123ae85e080d00ef1b3edd3f7405d</td>\n",
       "      <td>A Rift Valley fever (RVF) epidemic affecting a...</td>\n",
       "      <td>Raphaëlle Métras, Marc Baguelin, W John Edmund...</td>\n",
       "      <td>Raphaëlle Métras, Marc Baguelin, W John Edmund...</td>\n",
       "      <td>Abstract\\n\\nThe first cases occurred after hea...</td>\n",
       "      <td>R ift Valley fever (RVF) is a zoonotic arbovir...</td>\n",
       "      <td>Rift Valley fever virus (Bunyaviridae: Phlebov...</td>\n",
       "      <td>[{'first': 'Raphaëlle', 'middle': [], 'last': ...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'Rift Va...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>0104f6ceccf92ae8567a0102f89cbb976969a774</td>\n",
       "      <td>BMC Medical Genetics Association of HLA class ...</td>\n",
       "      <td>Marie Lin, Hsiang-Kuang Tseng, Jean A Trejaut,...</td>\n",
       "      <td>Marie Lin (Mackay Memorial Hospital, Taipei, T...</td>\n",
       "      <td>Abstract\\n\\nThe human leukocyte antigen (HLA) ...</td>\n",
       "      <td>\\n\\npatient group, a further significant incre...</td>\n",
       "      <td>for surveillance of severe acute respiratory s...</td>\n",
       "      <td>[{'first': 'Marie', 'middle': [], 'last': 'Lin...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'for sur...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   paper_id  \\\n",
       "0  002f09dfc9a1323a15bf72e349d8b733ac97a2aa   \n",
       "1  0036e8891c93ae63611bde179ada1e03e8577dea   \n",
       "2  00573277e6be50669016f770bc28ec2da0639a8f   \n",
       "3  00683d59d56123ae85e080d00ef1b3edd3f7405d   \n",
       "4  0104f6ceccf92ae8567a0102f89cbb976969a774   \n",
       "\n",
       "                                               title  \\\n",
       "0                                                      \n",
       "1  Stable Occupancy of the Crimean-Congo Hemorrha...   \n",
       "2  Asymptomatic Severe Acute Respiratory Syndrome...   \n",
       "3  A Rift Valley fever (RVF) epidemic affecting a...   \n",
       "4  BMC Medical Genetics Association of HLA class ...   \n",
       "\n",
       "                                             authors  \\\n",
       "0                                                      \n",
       "1  Florine E M Scholte, Brian L Hua, Jessica R Sp...   \n",
       "2                                                      \n",
       "3  Raphaëlle Métras, Marc Baguelin, W John Edmund...   \n",
       "4  Marie Lin, Hsiang-Kuang Tseng, Jean A Trejaut,...   \n",
       "\n",
       "                                        affiliations  \\\n",
       "0                                                      \n",
       "1  Florine E M Scholte (National Center for Emerg...   \n",
       "2                                                      \n",
       "3  Raphaëlle Métras, Marc Baguelin, W John Edmund...   \n",
       "4  Marie Lin (Mackay Memorial Hospital, Taipei, T...   \n",
       "\n",
       "                                            abstract  \\\n",
       "0                                                      \n",
       "1  Abstract\\n\\nCrimean-Congo hemorrhagic fever vi...   \n",
       "2                                                      \n",
       "3  Abstract\\n\\nThe first cases occurred after hea...   \n",
       "4  Abstract\\n\\nThe human leukocyte antigen (HLA) ...   \n",
       "\n",
       "                                                text  \\\n",
       "0  \\n\\nT he modern word camel is derived from the...   \n",
       "1  \\n\\nto Western Europe with the assistance of m...   \n",
       "2  \\n\\nWe identified a nurse who was asymptomatic...   \n",
       "3  R ift Valley fever (RVF) is a zoonotic arbovir...   \n",
       "4  \\n\\npatient group, a further significant incre...   \n",
       "\n",
       "                                        bibliography  \\\n",
       "0  Ancient and modern DNA reveal dynamics of dome...   \n",
       "1  Seroepidemiological studies of Crimean-Congo h...   \n",
       "2  Coronavirus as a possible cause of severe acut...   \n",
       "3  Rift Valley fever virus (Bunyaviridae: Phlebov...   \n",
       "4  for surveillance of severe acute respiratory s...   \n",
       "\n",
       "                                         raw_authors  \\\n",
       "0                                                 []   \n",
       "1  [{'first': 'Florine', 'middle': ['E M'], 'last...   \n",
       "2                                                 []   \n",
       "3  [{'first': 'Raphaëlle', 'middle': [], 'last': ...   \n",
       "4  [{'first': 'Marie', 'middle': [], 'last': 'Lin...   \n",
       "\n",
       "                                    raw_bibliography  \n",
       "0  {'BIBREF0': {'ref_id': 'b0', 'title': 'Ancient...  \n",
       "1  {'BIBREF0': {'ref_id': 'b0', 'title': 'Seroepi...  \n",
       "2  {'BIBREF0': {'ref_id': 'b0', 'title': 'Coronav...  \n",
       "3  {'BIBREF0': {'ref_id': 'b0', 'title': 'Rift Va...  \n",
       "4  {'BIBREF0': {'ref_id': 'b0', 'title': 'for sur...  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pmc_dir = 'pmc_custom_license/pmc_custom_license/'\n",
    "pmc_files = load_files(pmc_dir)\n",
    "pmc_df = generate_clean_df(pmc_files)\n",
    "pmc_df.to_csv('clean_pmc.csv', index=False)\n",
    "pmc_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "26166e76d5884d6bb2bfcd9b66f27585",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=9000), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1a7db3ea9b074b0e8289c6a0072c80ed",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=9000), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paper_id</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>affiliations</th>\n",
       "      <th>abstract</th>\n",
       "      <th>text</th>\n",
       "      <th>bibliography</th>\n",
       "      <th>raw_authors</th>\n",
       "      <th>raw_bibliography</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>000b7d1517ceebb34e1e3e817695b6de03e2fa78</td>\n",
       "      <td>Supplementary Information An eco-epidemiologic...</td>\n",
       "      <td>Julien Mélade, Nicolas Wieseke 4#, Beza Ramazi...</td>\n",
       "      <td>Julien Mélade (2 rue Maxime Rivière, 97490 Sai...</td>\n",
       "      <td></td>\n",
       "      <td>\\n\\n- Figure S1 : Phylogeny of all sequences b...</td>\n",
       "      <td>NDV/HQ266603/Chicken/1992, , , None; MuV/FJ375...</td>\n",
       "      <td>[{'first': 'Julien', 'middle': [], 'last': 'Mé...</td>\n",
       "      <td>{'BIBREF32': {'ref_id': 'b32', 'title': 'NDV/H...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>00142f93c18b07350be89e96372d240372437ed9</td>\n",
       "      <td>immunity to pathogens taught by specialized hu...</td>\n",
       "      <td>Elisabetta Padovan, Marina Cella, Shahram Sale...</td>\n",
       "      <td>Elisabetta Padovan, Marina Cella, Shahram Sale...</td>\n",
       "      <td>Abstract\\n\\nDendritic cells (DCs) are speciali...</td>\n",
       "      <td>\\n\\niNTRODUCTiON Human beings are constantly e...</td>\n",
       "      <td>The dendritic cell system and its role in immu...</td>\n",
       "      <td>[{'first': 'Elisabetta', 'middle': [], 'last':...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'The den...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>0022796bb2112abd2e6423ba2d57751db06049fb</td>\n",
       "      <td>Public Health Responses to and Challenges for ...</td>\n",
       "      <td>Elvina Viennet, Scott A Ritchie, Craig R Willi...</td>\n",
       "      <td>Elvina Viennet (The Australian National Univer...</td>\n",
       "      <td>Abstract\\n\\nDengue has a negative impact in lo...</td>\n",
       "      <td>Introduction\\n\\nPathogens and vectors can now ...</td>\n",
       "      <td>The global distribution and burden of dengue, ...</td>\n",
       "      <td>[{'first': 'Elvina', 'middle': [], 'last': 'Vi...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'The glo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>00326efcca0852dc6e39dc6b7786267e1bc4f194</td>\n",
       "      <td>a section of the journal Frontiers in Pediatri...</td>\n",
       "      <td>Jan Hau Lee, Oguz Dursun, Phuc Huu Phan, Yek K...</td>\n",
       "      <td>Jan Hau Lee, Oguz Dursun, Phuc Huu Phan, Yek K...</td>\n",
       "      <td>Abstract\\n\\nFifteen years ago, United Nations ...</td>\n",
       "      <td>\\n\\nIn addition to preventative care and nutri...</td>\n",
       "      <td>Global, regional, and national levels of neona...</td>\n",
       "      <td>[{'first': 'Jan', 'middle': ['Hau'], 'last': '...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'Global,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>00352a58c8766861effed18a4b079d1683fec2ec</td>\n",
       "      <td>MINI REVIEW Function of the Deubiquitinating E...</td>\n",
       "      <td>Molly Hodul, Caroline L Dahlberg, Peter Juo, C...</td>\n",
       "      <td>Molly Hodul (Tufts University, Boston, MA, Uni...</td>\n",
       "      <td>Abstract\\n\\nPosttranslational modification of ...</td>\n",
       "      <td>INTRODUCTION\\n\\nUbiquitination is a widely use...</td>\n",
       "      <td>Regulation of AMPA receptor trafficking and sy...</td>\n",
       "      <td>[{'first': 'Molly', 'middle': [], 'last': 'Hod...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'Regulat...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   paper_id  \\\n",
       "0  000b7d1517ceebb34e1e3e817695b6de03e2fa78   \n",
       "1  00142f93c18b07350be89e96372d240372437ed9   \n",
       "2  0022796bb2112abd2e6423ba2d57751db06049fb   \n",
       "3  00326efcca0852dc6e39dc6b7786267e1bc4f194   \n",
       "4  00352a58c8766861effed18a4b079d1683fec2ec   \n",
       "\n",
       "                                               title  \\\n",
       "0  Supplementary Information An eco-epidemiologic...   \n",
       "1  immunity to pathogens taught by specialized hu...   \n",
       "2  Public Health Responses to and Challenges for ...   \n",
       "3  a section of the journal Frontiers in Pediatri...   \n",
       "4  MINI REVIEW Function of the Deubiquitinating E...   \n",
       "\n",
       "                                             authors  \\\n",
       "0  Julien Mélade, Nicolas Wieseke 4#, Beza Ramazi...   \n",
       "1  Elisabetta Padovan, Marina Cella, Shahram Sale...   \n",
       "2  Elvina Viennet, Scott A Ritchie, Craig R Willi...   \n",
       "3  Jan Hau Lee, Oguz Dursun, Phuc Huu Phan, Yek K...   \n",
       "4  Molly Hodul, Caroline L Dahlberg, Peter Juo, C...   \n",
       "\n",
       "                                        affiliations  \\\n",
       "0  Julien Mélade (2 rue Maxime Rivière, 97490 Sai...   \n",
       "1  Elisabetta Padovan, Marina Cella, Shahram Sale...   \n",
       "2  Elvina Viennet (The Australian National Univer...   \n",
       "3  Jan Hau Lee, Oguz Dursun, Phuc Huu Phan, Yek K...   \n",
       "4  Molly Hodul (Tufts University, Boston, MA, Uni...   \n",
       "\n",
       "                                            abstract  \\\n",
       "0                                                      \n",
       "1  Abstract\\n\\nDendritic cells (DCs) are speciali...   \n",
       "2  Abstract\\n\\nDengue has a negative impact in lo...   \n",
       "3  Abstract\\n\\nFifteen years ago, United Nations ...   \n",
       "4  Abstract\\n\\nPosttranslational modification of ...   \n",
       "\n",
       "                                                text  \\\n",
       "0  \\n\\n- Figure S1 : Phylogeny of all sequences b...   \n",
       "1  \\n\\niNTRODUCTiON Human beings are constantly e...   \n",
       "2  Introduction\\n\\nPathogens and vectors can now ...   \n",
       "3  \\n\\nIn addition to preventative care and nutri...   \n",
       "4  INTRODUCTION\\n\\nUbiquitination is a widely use...   \n",
       "\n",
       "                                        bibliography  \\\n",
       "0  NDV/HQ266603/Chicken/1992, , , None; MuV/FJ375...   \n",
       "1  The dendritic cell system and its role in immu...   \n",
       "2  The global distribution and burden of dengue, ...   \n",
       "3  Global, regional, and national levels of neona...   \n",
       "4  Regulation of AMPA receptor trafficking and sy...   \n",
       "\n",
       "                                         raw_authors  \\\n",
       "0  [{'first': 'Julien', 'middle': [], 'last': 'Mé...   \n",
       "1  [{'first': 'Elisabetta', 'middle': [], 'last':...   \n",
       "2  [{'first': 'Elvina', 'middle': [], 'last': 'Vi...   \n",
       "3  [{'first': 'Jan', 'middle': ['Hau'], 'last': '...   \n",
       "4  [{'first': 'Molly', 'middle': [], 'last': 'Hod...   \n",
       "\n",
       "                                    raw_bibliography  \n",
       "0  {'BIBREF32': {'ref_id': 'b32', 'title': 'NDV/H...  \n",
       "1  {'BIBREF0': {'ref_id': 'b0', 'title': 'The den...  \n",
       "2  {'BIBREF0': {'ref_id': 'b0', 'title': 'The glo...  \n",
       "3  {'BIBREF0': {'ref_id': 'b0', 'title': 'Global,...  \n",
       "4  {'BIBREF0': {'ref_id': 'b0', 'title': 'Regulat...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comm_dir = 'comm_use_subset/comm_use_subset/'\n",
    "comm_files = load_files(comm_dir)\n",
    "comm_df = generate_clean_df(comm_files)\n",
    "comm_df.to_csv('clean_comm_use.csv', index=False)\n",
    "comm_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6178728d7b1c428c90d044c2470e2c3a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=1973), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "eadbd2763ce14c4286db3cdcad9b6da1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=1973), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paper_id</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>affiliations</th>\n",
       "      <th>abstract</th>\n",
       "      <th>text</th>\n",
       "      <th>bibliography</th>\n",
       "      <th>raw_authors</th>\n",
       "      <th>raw_bibliography</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0036b28fddf7e93da0970303672934ea2f9944e7</td>\n",
       "      <td>The fecal microbiota and unconjugated fecal bi...</td>\n",
       "      <td>B Ruggerone, A C Manchester, F Del Baldo, F Fr...</td>\n",
       "      <td>B Ruggerone (University of Thessaly, Karditsa,...</td>\n",
       "      <td>Abstract\\n\\nand Blautia (P = 0.008) significan...</td>\n",
       "      <td>\\n\\nhuman type 1 DM. The aim of this study was...</td>\n",
       "      <td>Significant Feline Proteinuria: a retrospectiv...</td>\n",
       "      <td>[{'first': 'B', 'middle': [], 'last': 'Ruggero...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'Signifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>005c43980edf3fcc2a4d12ee7ad630ddb651ce6e</td>\n",
       "      <td>Development of a smartphone-based rapid dual f...</td>\n",
       "      <td>Seon-Ju Yeo, Homan Kang, Tung Duy Dao, Thi Bui...</td>\n",
       "      <td>Seon-Ju Yeo (Wonkwang University, 570-749, Iks...</td>\n",
       "      <td>Abstract\\n\\nAccurate and rapid diagnosis of hi...</td>\n",
       "      <td>Introduction\\n\\nAvian influenza (AI) viruses a...</td>\n",
       "      <td>Avian and other zoonotic influenza, , , 2018; ...</td>\n",
       "      <td>[{'first': 'Seon-Ju', 'middle': [], 'last': 'Y...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'Avian a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>006be99e337c84b8758591a54f0362353b24dfde</td>\n",
       "      <td>Regulated Export of a Secretory Protein from t...</td>\n",
       "      <td>Stephen S Macintyre</td>\n",
       "      <td>Stephen S Macintyre (Case Western Reserve Univ...</td>\n",
       "      <td>Abstract\\n\\n253 protein, C-reactive protein (C...</td>\n",
       "      <td>\\n\\nAbstract. The half-time for secretion of t...</td>\n",
       "      <td>Studies on the binding specificity of human C-...</td>\n",
       "      <td>[{'first': 'Stephen', 'middle': ['S'], 'last':...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'Studies...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>00a00d0edc750db4a0c299dd1ec0c6871f5a4f24</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Editorial\\n\\nThis is an Open Access article di...</td>\n",
       "      <td>Report of the Ebola Interim Assessment Panel, ...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'Report ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>00e5a723d44eb9f2698c38b518eff85c00f9753b</td>\n",
       "      <td></td>\n",
       "      <td>Vijay Harypursat, Yao-Kai Chen,  Yao-Kai,  Chen</td>\n",
       "      <td>Vijay Harypursat (Chongqing Public Health Medi...</td>\n",
       "      <td></td>\n",
       "      <td>\\n\\nCoronaviruses have in the past been known ...</td>\n",
       "      <td>Efficient replication of the novel human betac...</td>\n",
       "      <td>[{'first': 'Vijay', 'middle': [], 'last': 'Har...</td>\n",
       "      <td>{'BIBREF0': {'ref_id': 'b0', 'title': 'Efficie...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   paper_id  \\\n",
       "0  0036b28fddf7e93da0970303672934ea2f9944e7   \n",
       "1  005c43980edf3fcc2a4d12ee7ad630ddb651ce6e   \n",
       "2  006be99e337c84b8758591a54f0362353b24dfde   \n",
       "3  00a00d0edc750db4a0c299dd1ec0c6871f5a4f24   \n",
       "4  00e5a723d44eb9f2698c38b518eff85c00f9753b   \n",
       "\n",
       "                                               title  \\\n",
       "0  The fecal microbiota and unconjugated fecal bi...   \n",
       "1  Development of a smartphone-based rapid dual f...   \n",
       "2  Regulated Export of a Secretory Protein from t...   \n",
       "3                                                      \n",
       "4                                                      \n",
       "\n",
       "                                             authors  \\\n",
       "0  B Ruggerone, A C Manchester, F Del Baldo, F Fr...   \n",
       "1  Seon-Ju Yeo, Homan Kang, Tung Duy Dao, Thi Bui...   \n",
       "2                                Stephen S Macintyre   \n",
       "3                                                      \n",
       "4    Vijay Harypursat, Yao-Kai Chen,  Yao-Kai,  Chen   \n",
       "\n",
       "                                        affiliations  \\\n",
       "0  B Ruggerone (University of Thessaly, Karditsa,...   \n",
       "1  Seon-Ju Yeo (Wonkwang University, 570-749, Iks...   \n",
       "2  Stephen S Macintyre (Case Western Reserve Univ...   \n",
       "3                                                      \n",
       "4  Vijay Harypursat (Chongqing Public Health Medi...   \n",
       "\n",
       "                                            abstract  \\\n",
       "0  Abstract\\n\\nand Blautia (P = 0.008) significan...   \n",
       "1  Abstract\\n\\nAccurate and rapid diagnosis of hi...   \n",
       "2  Abstract\\n\\n253 protein, C-reactive protein (C...   \n",
       "3                                                      \n",
       "4                                                      \n",
       "\n",
       "                                                text  \\\n",
       "0  \\n\\nhuman type 1 DM. The aim of this study was...   \n",
       "1  Introduction\\n\\nAvian influenza (AI) viruses a...   \n",
       "2  \\n\\nAbstract. The half-time for secretion of t...   \n",
       "3  Editorial\\n\\nThis is an Open Access article di...   \n",
       "4  \\n\\nCoronaviruses have in the past been known ...   \n",
       "\n",
       "                                        bibliography  \\\n",
       "0  Significant Feline Proteinuria: a retrospectiv...   \n",
       "1  Avian and other zoonotic influenza, , , 2018; ...   \n",
       "2  Studies on the binding specificity of human C-...   \n",
       "3  Report of the Ebola Interim Assessment Panel, ...   \n",
       "4  Efficient replication of the novel human betac...   \n",
       "\n",
       "                                         raw_authors  \\\n",
       "0  [{'first': 'B', 'middle': [], 'last': 'Ruggero...   \n",
       "1  [{'first': 'Seon-Ju', 'middle': [], 'last': 'Y...   \n",
       "2  [{'first': 'Stephen', 'middle': ['S'], 'last':...   \n",
       "3                                                 []   \n",
       "4  [{'first': 'Vijay', 'middle': [], 'last': 'Har...   \n",
       "\n",
       "                                    raw_bibliography  \n",
       "0  {'BIBREF0': {'ref_id': 'b0', 'title': 'Signifi...  \n",
       "1  {'BIBREF0': {'ref_id': 'b0', 'title': 'Avian a...  \n",
       "2  {'BIBREF0': {'ref_id': 'b0', 'title': 'Studies...  \n",
       "3  {'BIBREF0': {'ref_id': 'b0', 'title': 'Report ...  \n",
       "4  {'BIBREF0': {'ref_id': 'b0', 'title': 'Efficie...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "noncomm_dir = 'noncomm_use_subset/noncomm_use_subset/'\n",
    "noncomm_files = load_files(noncomm_dir)\n",
    "noncomm_df = generate_clean_df(noncomm_files)\n",
    "noncomm_df.to_csv('clean_noncomm_use.csv', index=False)\n",
    "noncomm_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
